---
title: "Blocker Project: Differential Expression using DESeq2- Command line miRDeep2 count matrix used"
output: html_notebook
---

#Using DESeq2 to perform differential expression on miRNASeq data, comparing blocked vs no blocked groups

****Note: this notebook uses only the count matrix generated by the command line miRDeep2 analysis, with default settings. This entire analysis was also run on the count matrix generated by BaseSpace miRDeep2. See "FINAL_DifferentialExpression_BaseSpaceCountMatrix_BlockerProject.Rmd" for this analysis.

Import libraries:
```{r}
suppressPackageStartupMessages(library(DESeq2))
suppressPackageStartupMessages(library(pheatmap))
suppressPackageStartupMessages(library(viridis))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(reshape2))
suppressPackageStartupMessages(library(grid))
suppressPackageStartupMessages(library(gridExtra))
suppressPackageStartupMessages(library(metaseqR))
```


#Two different input sources: 

*1) count matrix from BaseSpace (miRDeep2, no mismatches, isomirs and mature together) --> Downsampled to lowest read count
    a) Isomirs and mature
    b) Mature only
2) count matrix generated from miRDeep2 at the command line- mature only --> Downsampled to lowest read count


*This analysis performed in separate file: "FINAL_DifferentialExpression_BaseSpaceCountMatrix_BlockerProject.Rmd"

#For #1b and #2 count matrices, 2 different DESeq2 analyses performed:

1) Standard- Pairwise, technical replicates collapsed, all mature species included
2) Excluding targets- same as standard, but only including NON TARGET mature species
    a) excluding just the actual targets
    b) excluding the actual targets + members of the same family

Matrix #1 used initially, but after determining that a mismatch of 1 should be allowed (i.e., miRDeep2 default settings), the miRDeep2 analysis was rerun at the command line (matrix #2) with default miRDeep2 parameters. Both analyses used the same input bam file, generated from BaseSpace (splice-aware STAR alignmenet)


#Notes on DESeq2 Analysis:

-Since there are 4 sample types (A/B/C/D) each with at least one blocked/unblocked pair, pairwise DESeq2 analysis is performed

-Samples A and B have 3 technical replicates, for both blocked and unblocked groups. These technical replicates are collapsed using an internal DESeq2 function. These technical replicates have varying RNA input amounts into library construction, but this was determined to have no effect on the resulting libraries (See ___.Rmd). This is additionally confirmed by PCA clustering here.

Read in data:
```{r}
wd<- "C:/Users/jjl71/Documents/BlockerProject/BlockerProject/"
CL_countsInput<- read.csv(paste(wd, "RawData/miRNA_CLmiRDeep2Results_AllSamples_DefaultmiRDeep2Settings.csv",sep=""))
  #Note: this dataset comes from "Combining_miRNAHits_CLmiRBaseCountmatrix.Rmd", where count matrices for individual samples were combined into the single file here
rownames(CL_countsInput)<-CL_countsInput$X
CL_countsInput<- CL_countsInput[,-1]

#Previously downsampled data: (in "FINAL_BlockersReduceTargetMapping.Rmd")
CL_countsInput_Downsampled<- read.csv(paste(wd, "RawData/Downsampled_miRNA_CLmiRDeep2Results_AllSamples_DefaultmiRDeep2Settings.csv",sep=""))
rownames(CL_countsInput_Downsampled)<-CL_countsInput_Downsampled$X
CL_countsInput_Downsampled<- CL_countsInput_Downsampled[,-1]
#Reorder 
CL_countsInput_Downsampled<- CL_countsInput_Downsampled[,c(1,3,5,7,9,11,13,15,2,4,6,8,10,12,14,16)]
#Convert to integer (removes decimal places)
CL_countsInput_Downsampled_Integer<- as.data.frame(lapply(CL_countsInput_Downsampled, as.integer))
rownames(CL_countsInput_Downsampled_Integer)<- rownames(CL_countsInput_Downsampled)
CL_Downsampled<- CL_countsInput_Downsampled_Integer
#This matrix used for downstream purposes
```

################################################################################
#Using Matrix #2 (generated by CL miRDeep2, 1 mismatch allowed and still considered mature):
################################################################################


Prep data for DESeq2, create metadata and DESeq2 object, then run DESeq2:
```{r}
#rename rows, removing input info
colnames(CL_Downsampled)<- c(rep("A.0", 3), rep("B.0", 3), "C.0", "D.0", rep("A.1", 3), rep("B.1", 3), "C.1", "D.1")

#Create metadata:

#set blocker status
sampleGroup<- as.factor(c(rep("NoBlocker", 8), rep("Blocker", 8)))
#this column is used to collapse technical replicates
sample<- colnames(CL_Downsampled)
#this column is to control for sample to sample differences
patient<- as.factor(c(rep("A", 3), rep("B", 3), "C", "D", rep("A", 3), rep("B", 3), "C", "D"))
meta<- data.frame(sample, sampleGroup, patient)

#create DESeq2 object- just based on samplegroup (blockerstatus), will add samples to model later (i.e., pairwise analysis)
dds<- DESeqDataSetFromMatrix(countData=CL_Downsampled, 
                             colData=meta, 
                             design=~sampleGroup)

#Collapse replicates- using internal DESeq2 function
ddsColl<- collapseReplicates(dds, dds$sample)

#Create model that takes into account sample differences
ddsMF<- ddsColl
design(ddsMF)<- formula(~patient + sampleGroup)

#run DESeq2
dds<- DESeq(ddsMF)
```

Visualize sample differences using PCA

Note: this function uses the dds object, but no information on sample group is input into the analysis (i.e., unsupervised clustering)
```{r}
#Normalized (variance stabilized transformation), then run PCA
vsd <- varianceStabilizingTransformation(ddsMF, blind=TRUE)
PCA<-plotPCA(vsd, intgroup="sampleGroup",returnData=TRUE)

p<- ggplot(PCA, aes(x=PC1, y=PC2, color=group))+
  geom_point(size=4)+
  scale_color_manual(values=c("orangered1","springgreen2"))+
  theme_classic()+
  ylim(-15,15)+
  ggtitle("PCA: Blocker vs no Blocker, Mature")+
  geom_text(aes(label=name),hjust=.5, vjust=-1.2, color="black")
  
p
```



Is there any possibility this is due to collapsing of technical replicates? If we didn't collapse technical replicates, would we see all 3 A.1/3 B.1, etc still all clustering together?
Testing this..
```{r}
TechNotCollapsed<-CL_Downsampled

#Rename samples
colnames(TechNotCollapsed)<- c(paste("A.0.", 1:3,sep=""), paste("B.0.", 1:3,sep=""), "C.0", "D.0", paste("A.1.", 1:3,sep=""), paste("B.1.", 1:3,sep=""), "C.1", "D.1")

#Create metadata:

#set blocker status
sampleGroup<- as.factor(c(rep("NoBlocker", 8), rep("Blocker", 8)))
#this column is to control for sample to sample differences
patient<- as.factor(c(paste("A", 1:3,sep=""), paste("B", 1:3,sep=""), "C", "D", paste("A", 1:3,sep=""), paste("B", 1:3,sep=""), "C", "D"))
meta<- data.frame(sampleGroup, patient)

#create DESeq2 object- just based on samplegroup (blockerstatus), will add samples to model later (i.e., pairwise analysis)
dds_TNC<- DESeqDataSetFromMatrix(countData=TechNotCollapsed, 
                             colData=meta, 
                             design=~sampleGroup)


#Create model that takes into account sample differences
ddsMF_TNC<- dds_TNC
design(ddsMF_TNC)<- formula(~patient + sampleGroup)

#run DESeq2
dds_TNC<- DESeq(ddsMF_TNC)

#Normalized (variance stabilized transformation), then run PCA
vsd_TNC <- varianceStabilizingTransformation(ddsMF_TNC, blind=TRUE)
PCA_TNC<-plotPCA(vsd_TNC, intgroup="sampleGroup",returnData=TRUE)

p<- ggplot(PCA_TNC, aes(x=PC1, y=PC2, color=group))+
  geom_point(size=4)+
  scale_color_manual(values=c("orangered1","springgreen2"))+
  theme_classic()+
  ylim(-20,25)+
  ggtitle("PCA: Blocker vs no Blocker, Technical Replicates not Collapsed: Mature + Isomirs")+
  geom_text(aes(label=name),hjust=.5, vjust=-1.2, color="black")
  
p

```
So the answer to that question is yes- samples still cluster in a similar pattern, even if you don't collapse by technical replicates. A/B and C/D really are more similar to one another.


#Back to using the dds object with technical replicates collapsed:


Differential expression results:
```{r}
res<- results(dds)
res<- na.omit(res)
resOrdered<- res[order(res$padj),]
nrow(resOrdered[resOrdered$padj<0.05,])
nrow(resOrdered[resOrdered$padj<0.01,])
#37 DE miRNAs- padj <0.05
#25 DE miRNAs- padj <0.01

#Save results for later
CL_DESeq2Results<-resOrdered

rownames(resOrdered[resOrdered$padj<0.01,])
```


Visualize DE results with heatmap:
```{r}
#normalize results
rld<- rlog(dds)

#Select top 75 sig DE miRNAs
topgenes<- rownames(resOrdered[1:50,])
mat<- assay(rld)[topgenes,]
mat<- mat - rowMeans(mat)

#Heatmap settings
subtractFromMin<- 0
subtractFromMax<- 1
lengthOut<- 75
mat_breaks<- seq(min(mat-subtractFromMin), max(mat-subtractFromMax), length.out=lengthOut)

#Create metadata- to be used to color code to show which group the sample belongs to
metadata<- data.frame(Sample_Group=PCA$sampleGroup, row.names=PCA$name)

#plot heatmap
pheatmap(mat, breaks = mat_breaks, color =inferno(length(mat_breaks-1)),show_rownames = FALSE, annotation_col = metadata, border_color= NA)
```

Blocked/unblocked samples cluster together, as expected. Samples A/B and C/D also cluster together, as seen in the PCA results


Write to csv
```{r}
write.csv(resOrdered, paste(wd, "DESeq2Results/DESeq2Results_CLDownsampledMatureIsomir_121619.csv", sep =""))
```


#DESeq2 Analysis 2: removing targeted miRNAs

Two variations
    a) excluding just the actual targets
    b) excluding the actual targets + members of the same family

#Variation A (only actual targets removed) using count matrix #1b (from BaseSpace, mature species only)

```{r}
#Get list of all targets
ActualTargets<- c('hsa-miR-486-5p','hsa-miR-92a-3p','hsa-miR-451a')

#subset counts (mature only) by non-target miRNAs
CL_AT<- CL_Downsampled[!(rownames(CL_Downsampled) %in% ActualTargets),]
```

Create dds object, run DESeq2
```{r}
#Create metadata:
#set blocker status
sampleGroup<- as.factor(c(rep("NoBlocker", 8), rep("Blocker", 8)))
#this column is used to collapse technical replicates
sample<- colnames(CL_Downsampled)
#this column is to control for sample to sample differences
patient<- as.factor(c(rep("A", 3), rep("B", 3), "C", "D", rep("A", 3), rep("B", 3), "C", "D"))
meta<- data.frame(sample, sampleGroup, patient)

#create DESeq2 object- just based on samplegroup (blockerstatus), will add samples to model later (i.e., pairwise analysis)
dds<- DESeqDataSetFromMatrix(countData=CL_AT, 
                             colData=meta, 
                             design=~sampleGroup)

#Collapse replicates- using internal DESeq2 function
ddsColl<- collapseReplicates(dds, dds$sample)

#Create model that takes into account sample differences
ddsMF<- ddsColl
design(ddsMF)<- formula(~patient + sampleGroup)

#run DESeq2
dds<- DESeq(ddsMF)
```


Visualize sample differences using PCA
```{r}
#Normalized (variance stabilized transformation), then run PCA
vsd <- varianceStabilizingTransformation(ddsMF, blind=TRUE)
PCA<-plotPCA(vsd, intgroup="sampleGroup",returnData=TRUE)

p<- ggplot(PCA, aes(x=PC1, y=PC2, color=group))+
  geom_point(size=4)+
  scale_color_manual(values=c("orangered1","springgreen2"))+
  theme_classic()+
  ylim(-15,15)+
  ggtitle("PCA: Blocker vs no Blocker, Mature Only, Actual Targets Removed")+
  geom_text(aes(label=name),hjust=.5, vjust=-1.2, color="black")
  
p
```

When you remove the targets from analysis, samples cluster completely even more tightly by SAMPLE TYPE! Exactly what we would expect. Those 3 targets are causing the blocker status clustering we see in all other PCA plots shown previously here. Suggests that there are no huge off target effects caused by the blocker that cause large expression differences between blocked/unblocked


Extract DESeq2 results and visualize with heatmap
```{r}
res<- results(dds)
res<- na.omit(res)
resOrdered<- res[order(res$padj),]
nrow(resOrdered[resOrdered$padj<0.01,])
nrow(resOrdered[resOrdered$padj<0.05,])
#21 DE miRNAs- padj <0.05
#31 DE miRNAs- padj <0.01

rownames(resOrdered[resOrdered$padj<0.05,])

#normalize results
rld<- rlog(dds)

#Select top sig DE miRNAs- fewer here than for mature/isomir analysis because there are a lot fewer sig DE miRNAs
topgenes<- rownames(resOrdered[1:75,])
mat<- assay(rld)[topgenes,]
mat<- mat - rowMeans(mat)

#Heatmap settings
subtractFromMin<- 0
subtractFromMax<- 0
lengthOut<- 75
mat_breaks<- seq(min(mat-subtractFromMin), max(mat-subtractFromMax), length.out=lengthOut)

#Create metadata- to be used to color code to show which group the sample belongs to
metadata<- data.frame(Sample_Group=PCA$sampleGroup, row.names=PCA$name)

#plot heatmap
pheatmap(mat, breaks = mat_breaks, color =inferno(length(mat_breaks-1)),show_rownames = FALSE,  border_color= NA,annotation_col = metadata)
```

In heatmap of DE results, for A/B, samples cluster by SAMPLE and not BLOCKER STATUS, as we would expect if there are no huge off target effects caused by the blockers. However- C/D still cluster by blocker status. Will look at removing targets+target families, next, see if C/D still cluster by blocker status



Write results to file:
```{r}
#write to file 
write.csv(resOrdered, paste(wd, "DESeq2Results/DESeq2Results_ActualTargetsRemoved_DownSampledCLCountMatrix_Pairwise_121619.csv", sep = "" ))
```



#Variation B (actual targets + species in same family removed) 

Select all species with 92, 451, or 486 in name. Also select 2 hsa-mir-25 species (same family as 92a)
```{r}
Target_92<- rownames(CL_AT[grep("miR-92", rownames(CL_AT))[1:4],])
Target_451<- rownames(CL_AT[grep("miR-451", rownames(CL_AT))[1],])
Target_486<- rownames(CL_AT[grep("miR-486", rownames(CL_AT)),])
Target_25<- rownames(CL_AT[grep("miR-25", rownames(CL_AT)),])
TargetFamilies<- c(Target_451,Target_486,Target_92, Target_25)

#Using BaseSpace Mature counts matrix that already has the actual targets removed, remove the 8 additional species that are in the same family as the 3 actual targets
CL_NoTF<- CL_AT[!(rownames(CL_AT)%in%TargetFamilies),]
```



Create dds object, run DESeq2, visualize with PCA
```{r}
#create DESeq2 object- just based on samplegroup (blockerstatus), will add samples to model later (i.e., pairwise analysis)
dds<- DESeqDataSetFromMatrix(countData=CL_NoTF, 
                             colData=meta, 
                             design=~sampleGroup)

#Collapse replicates- using internal DESeq2 function
ddsColl<- collapseReplicates(dds, dds$sample)

#Create model that takes into account sample differences
ddsMF<- ddsColl
design(ddsMF)<- formula(~patient + sampleGroup)

#run DESeq2
dds<- DESeq(ddsMF)

#Visualize sample differences using PCA

#Normalized (variance stabilized transformation), then run PCA
vsd <- varianceStabilizingTransformation(ddsMF, blind=TRUE)
PCA<-plotPCA(vsd, intgroup="sampleGroup",returnData=TRUE)

p<- ggplot(PCA, aes(x=PC1, y=PC2, color=group))+
  geom_point(size=4)+
  scale_color_manual(values=c("orangered1","springgreen2"))+
  theme_classic()+
  ylim(-15,15)+
  ggtitle("PCA: Blocker vs no Blocker, Mature Only, Targets + Target Families removed")+
  geom_text(aes(label=name),hjust=.5, vjust=-1.2, color="black")
  
p
```

Very similar to previous analysis (just actual targets removed), samples still cluster by SAMPLE TYPE not by blocker status

Extract DESeq2 results and visualize with heatmap
```{r}
res<- results(dds)
res<- na.omit(res)
resOrdered<- res[order(res$padj),]
nrow(resOrdered[resOrdered$padj<0.05,])
nrow(resOrdered[resOrdered$padj<0.01,])
#29 DE miRNAs- padj <0.05
#20 DE miRNAs- padj <0.01

rownames(resOrdered[resOrdered$padj<0.05,])

#normalize results
#rld<- rlog(dds)

#Select top 25 sig DE miRNAs- fewer here than for mature/isomir analysis because there are a lot fewer sig DE miRNAs
topgenes<- rownames(resOrdered[1:50,])
mat<- assay(rld)[topgenes,]
mat<- mat - rowMeans(mat)

#Heatmap settings
subtractFromMin<- 0
subtractFromMax<- 1
lengthOut<- 75
mat_breaks<- seq(min(mat-subtractFromMin), max(mat-subtractFromMax), length.out=lengthOut)

#Create metadata- to be used to color code to show which group the sample belongs to
metadata<- data.frame(Sample_Group=PCA$sampleGroup, row.names=PCA$name)

#plot heatmap
pheatmap(mat, breaks = mat_breaks, color =inferno(length(mat_breaks-1)),show_rownames = FALSE, annotation_col = metadata, border_color= NA)
```

Samples cluster fully by sample type, not blocker status!

IMPORTANT: when you remove not just the actual targets, but also the actual targets + species from the same family (as above), samples cluster FULLY by sample type, not blocker status- even for samples C/D

Compare this to when you remove just the actual targets: even if you select 100 miRNAs, samples still don't fully cluster by sample type.

This just further increases our confidence that there are very few off target effects caused by blockers (other than blocking members of the same family, which can't really be fully avoided)


Write results to file:
```{r}
#write to file 
write.csv(resOrdered, paste(wd, "DESeq2Results/DESeq2Results_ActualTargets+FamiliesRemoved_DownSampledCLCountMatrix_Pairwise_121619.csv", sep = "" ))
```


#Graphing results in a different way: Log of CPM for each miRNA, comparing blocked vs unblocked values

For coloring sig/not sig expressed species, using BaseSpace mature downsampled counts- isomirs not included

```{r}
#calculating CPM
cpmInput<- CL_Downsampled

#calculate factor to multiply all counts for to convert to CPM
totalCounts<- colSums(cpmInput)
TotalCountsPM<- 1000000/totalCounts

#Convert all counts to CPM
CPMFinal<-cpmInput*TotalCountsPM
#get rid of any miRNAs that have a 0 counts for any sample- can't take log2 of it
CPMFinal<- CPMFinal[apply(CPMFinal, 1, function(row) all(row !=0 )),]

#log2 of CPM
CPMLog<- log2(CPMFinal+1) #+1 optional here

#Get miRNA info. Need: whether or not it was targeted + whether or not it was DE
#Add in target info- add column for whether or not the species was targeted
targets<- c('hsa-miR-486-5p','hsa-miR-92a-3p','hsa-miR-451a')
CPMLog$target<- rownames(CPMLog) %in% targets

#Add in DE info- using DESeq2 results from targets included analysis. Import data here.
resOrdered<-CL_DESeq2Results
sig<- resOrdered[resOrdered$padj<0.05,]
CPMLog$Sig<- rownames(CPMLog) %in% rownames(sig)

#Add in third column, used to color points in graph. Add together target/sig columns. If it's a target and sig, it'll be 2, if it's just sig it'll be 1, if it's neither it'll be 0.
TargetSig<- CPMLog[,17:18]
CPMLog$Color<- as.factor(rowSums(TargetSig))
CPMLog$Color<- gsub(2, "Target", CPMLog$Color)
CPMLog$Color<- gsub(1, "DE", CPMLog$Color)
CPMLog$Color<- gsub(0, "Not DE", CPMLog$Color)
```


```{r}
#Looking at sample A- get average for all 3 Inputs
CPMLogA<- CPMLog[,c(1:3, 9:11,19)]
CPMLogA$A.NoBlocker<- rowMeans(CPMLogA[,1:3])
CPMLogA$A.Blocker<- rowMeans(CPMLogA[,4:6])
CPMLogA<- CPMLogA[,c(8,9,7)]

pA<- ggplot(CPMLogA, aes(x=A.NoBlocker, y=A.Blocker, color=Color))+
  geom_point()+
  geom_abline(slope=1, color="red")+
  scale_colour_manual(values=c("red","black", "blue"))+
  ggtitle("Log of Raw, Downsampled CPM of all Mature miRNAs: Sample A")+
  theme_classic()
 
pA
```

Most species follow the same general trend for both blocked and unblocked- i.e., overall similar expression pattern. EXCEPT for the targets (plus a few off target effects) 

```{r}
#Looking at sample B- get average for all 3 Inputs
CPMLogB<- CPMLog[,c(4:6, 12:14,19)]
CPMLogB$B.NoBlocker<- rowMeans(CPMLogB[,1:3])
CPMLogB$B.Blocker<- rowMeans(CPMLogB[,4:6])
CPMLogB<- CPMLogB[,c(8,9,7)]

pB<- ggplot(CPMLogB, aes(x=B.NoBlocker, y=B.Blocker, color=Color))+
  geom_point()+
  geom_abline(slope=1, color="red")+
  scale_colour_manual(values=c("red","black", "blue"))+
  ggtitle("Log of Raw, Downsampled CPM of all Mature miRNAs: Sample B")+
  theme_classic()
 
pB
```

```{r}
#For sample C
CPMLogC<- CPMLog[,c(7, 15,19)]
colnames(CPMLogC)<- c("C.NoBlocker", "C.Blocker", "Color")

pC<- ggplot(CPMLogC, aes(x=C.NoBlocker, y=C.Blocker, color=Color))+
  geom_point()+
  geom_abline(slope=1, color="red")+
  scale_colour_manual(values=c("red","black", "blue"))+
  ggtitle("Log of Raw, Downsampled CPM of all Mature miRNAs: Sample C")+
  theme_classic()
 
pC
```

```{r}
#For sample D
CPMLogD<- CPMLog[,c(8, 16,19)]
colnames(CPMLogD)<- c("D.NoBlocker", "D.Blocker", "Color")

pD<- ggplot(CPMLogD, aes(x=D.NoBlocker, y=D.Blocker, color=Color))+
  geom_point()+
  geom_abline(slope=1, color="red")+
  scale_colour_manual(values=c("red","black", "blue"))+
  ggtitle("Log of Raw, Downsampled CPM of all Mature miRNAs: Sample D")+
  theme_classic()+
  theme(plot.title = element_text(size = 10, face = "bold"))
  
 
pD
```


```{r}
#Averaging values across al 4 samples
CPMLogAll<- CPMLog
CPMLogAll$NoBlocker<- rowMeans(CPMLogAll[,1:8])
CPMLogAll$Blocker<- rowMeans(CPMLogAll[,9:16])
CPMLogAll<- CPMLogAll[,c(20,21,19)]

pMean<- ggplot(CPMLogAll, aes(x=NoBlocker, y=Blocker, color=Color))+
  geom_point()+
  geom_abline(slope=1, color="red")+
  scale_colour_manual(values=c("red","black", "blue"))+
  ggtitle("Log of Raw, Downsampled CPM of all Mature miRNAs: Mean Across All Samples")+
  theme_classic()
 
pMean
```

Putting A-D together on one plot:
```{r}
grid.arrange(pA+ggtitle(""), pB+ggtitle(""), pC+ggtitle(""), pD+ggtitle(""), nrow=2,top = textGrob("Log of Raw, Downsampled CPM for all mature miRNAs: All Samples",gp=gpar(fontsize=15,font=2)))
```

##########################################################################################
#Improvements with blocker: does using blockers allow for the discovery of more DE miRNAs?
###########################################################################################

General idea: compare the number of significantly DE species between A and B in both blocked and unblocked samples.

Two analyses:

1) The 3 A.0 samples vs the 3 B.0 samples
2) The 3 A.1 samples vs the 3 B.1 samples

Both use the mature, downsampled BaseSpace matrix

Analysis #1: DESeq2 using A0 vs B0 - here, technical replicates are NOT collapsed
```{r}
#rename rows, removing input info
A0_B0<- CL_Downsampled[,c(1:6)]
colnames(A0_B0)<- c("A.0.1", "A.0.2", "A.0.3", "B.0.1", "B.0.2", "B.0.3")

#create metadata
sampleGroup<- c(rep("NoBlocker", 3), rep("Blocker", 3))
sample<- colnames(A0_B0)
meta<- data.frame(sample, sampleGroup)

#create DESeq2 object
dds<- DESeqDataSetFromMatrix(countData=A0_B0, 
                             colData=meta, 
                             design=~sampleGroup)


dds<- DESeq(dds)

res<- results(dds)
res<- na.omit(res)
resOrdered<- res[order(res$padj),]
nrow(resOrdered[resOrdered$padj<0.05,])
#89 DE miRNAs
noblocker_AvB<- resOrdered
noblocker_sig_AvB<-noblocker_AvB[noblocker_AvB$padj<0.05,]
```

DESeq2 using A1 vs B1- here, technical replicates are NOT collapsed
```{r}
#rename rows, removing input info
A1_B1<- CL_Downsampled[,c(9:14)]
colnames(A0_B0)<- c("A.1.1", "A.1.2", "A.1.3", "B.1.1", "B.1.2", "B.1.3")

#create metadata
sampleGroup<- c(rep("NoBlocker", 3), rep("Blocker", 3))
sample<- colnames(A1_B1)
meta<- data.frame(sample, sampleGroup)

#create DESeq2 object
dds<- DESeqDataSetFromMatrix(countData=A1_B1, 
                             colData=meta, 
                             design=~sampleGroup)


dds<- DESeq(dds)

res<- results(dds)
res<- na.omit(res)
resOrdered<- res[order(res$padj),]
nrow(resOrdered[resOrdered$padj<0.05,])
#119 DE miRNAs
blocker_AvB<- resOrdered
blocker_sig_AvB<-blocker_AvB[blocker_AvB$padj<0.05,]
```


```{r}
#So that's a....
(119-89)/119
#25% increase in the number of DE genes!!

#What's the improvement if we consider padj<0.01 to be sig?
noblocker_sig.01_AvB<-noblocker_AvB[noblocker_AvB$padj<0.1,]
blocker_sig.01_AvB<-blocker_AvB[blocker_AvB$padj<0.1,]
(nrow(blocker_sig.01_AvB)-nrow(noblocker_sig.01_AvB))/nrow(blocker_sig.01_AvB)
#34% improvement
```

When using downsampled counts, you get an even greater number of DE miRNAs. Makes sense, right? The blocker gives you more species with really low abundance in the sample, but that improvement is partially masked bc there's more reads for the no blocker samples. When you get rid of that advantage, you can see the blocker improvement even more clearly.

Something that might be interesting to look at- where do these 2 sets of DE miRNAs differ? If the reasoning above is correct, most should be the same and a lot of the ones that differ should have pretty low overall counts

```{r}
DEInBoth<- rownames(blocker_sig.01_AvB) %in% rownames(noblocker_sig.01_AvB)
sum(DEInBoth)
#87 of the 114 unblocked DE species are also in the blocked DE species (75%)
#So of the 164 blocked DE species, 77 are unique to blocked
#And of the 114 unblocked DE species, 27 are unique to unblocked

#get DE species unique to blocked
OnlyBlockerDE<- blocker_sig.01_AvB[!DEInBoth,]

#looking at raw counts of these DE miRNAs
counts_onlyblockersDE<- CL_Downsampled[rownames(CL_Downsampled) %in% rownames(OnlyBlockerDE),]
counts_onlyblockersDE<- counts_onlyblockersDE[,c(1:6,9:14)]
counts_onlyblockersDE<-counts_onlyblockersDE[order(counts_onlyblockersDE$A.0, decreasing=TRUE),]
```

```{r}
#Median read count for ALL blocker DE species:
mean(rowMedians(as.matrix(CL_Downsampled[rownames(blocker_sig.01_AvB),])))
#14,035

mean(rowMedians(as.matrix(counts_onlyblockersDE)))
#15,620

#Excluding hsa-miR-92a (target, likely only DE due to differing efficiency in A vs B samples):
mean(rowMedians(as.matrix(counts_onlyblockersDE)[-1,]))
#1,197

```

So overall: if you exclude miR-92a, the blocker-specific DE miRNAs have a much lower raw read count than the entire set of DE miRNAs. This suggests that these miRNAs are DE due to the increased ability to detect very lowly expressed miRNAs. This idea will be looked at more closely in "FINAL_BlockerImprovements.Rmd"


Looking at unblocked-specific DE miRNAs:
```{r}
#get DE species unique to blocked
DEInBoth<- rownames(noblocker_sig.01_AvB) %in% rownames(blocker_sig.01_AvB)
OnlyNoBlockerDE<- noblocker_sig.01_AvB[!DEInBoth,]

#looking at raw counts of these DE miRNAs
counts_onlynoblockersDE<- CL_Downsampled[rownames(CL_Downsampled) %in% rownames(OnlyNoBlockerDE),]
counts_onlynoblockersDE<- counts_onlynoblockersDE[,c(1:6,9:14)]
counts_onlynoblockersDE<-counts_onlynoblockersDE[order(counts_onlynoblockersDE$A.0, decreasing=TRUE),]

#Median read count for ALL blocker DE species:
mean(rowMedians(as.matrix(CL_Downsampled[rownames(noblocker_sig_AvB),])))
#9730

mean(rowMedians(as.matrix(counts_onlynoblockersDE)))
#3648

#No targets in these 8 no blocker specific miRNAs
```
Overall, much higher total read counts for no blocker-specific sig DE miRNAs. Supports the idea that without blocker, can't get really good discrimination of lowly expressed miRNAs!
