---
title: "Blocker Project: Differential Expression using DESeq2- Command line miRDeep2 count matrix used"
output: html_notebook
---

#Using DESeq2 to perform differential expression on miRNASeq data, comparing blocked vs no blocked groups

****Note: this notebook uses only the count matrix generated by the command line miRDeep2 analysis, with default settings. This entire analysis was also run on the count matrix generated by BaseSpace miRDeep2. See "FINAL_DifferentialExpression_BaseSpaceCountMatrix_BlockerProject.Rmd" for this analysis.

Import libraries:
```{r}
suppressPackageStartupMessages(library(DESeq2))
suppressPackageStartupMessages(library(pheatmap))
suppressPackageStartupMessages(library(viridis))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(reshape2))
suppressPackageStartupMessages(library(grid))
suppressPackageStartupMessages(library(gridExtra))
suppressPackageStartupMessages(library(metaseqR))
suppressPackageStartupMessages(library(colorRamps))
suppressPackageStartupMessages(library(plyr))
```


#Functions used in analyses:
```{r}
#Create metadata and DESeq2 object, then run DESeq2

RunDESeq2<- function(counts,meta, collapseRep=TRUE) {
  #create DESeq2 object- just based on samplegroup (blockerstatus)
  #will add samples to model later (i.e., pairwise analysis)
  dds<- DESeqDataSetFromMatrix(countData=counts, 
                               colData=meta, 
                               design=~sampleGroup)
  #Collapse technical replicates using internal DESeq2 function
  if (collapseRep ==TRUE){
    dds<- collapseReplicates(dds, dds$sample)
  }
  
  #Create model that takes into account sample differences
  ddsMF<- dds
  design(ddsMF)<- formula(~Patient + sampleGroup)
  
  #run DESeq2
  dds<- DESeq(ddsMF)
  
  return(dds)
}

#Get DE results
GetDEResults<- function(dds){
  res<- results(dds)
  res<- na.omit(res)
  resOrdered<- res[order(res$padj),]
  print(nrow(resOrdered[resOrdered$padj<0.05,]))
  print(nrow(resOrdered[resOrdered$padj<0.01,]))
  return(resOrdered)
}
  

#Create PCA plot

MyPCA<- function(dds, metadata, PC1Variance, PC2Variance, Groupcolors,miny, maxy, minx, maxx) {
  #Normalized (variance stabilized transformation), then run PCA
  vsd <- varianceStabilizingTransformation(dds, blind=TRUE)
  PCA<-plotPCA(vsd, intgroup="sampleGroup",returnData=TRUE)
  PCA$Patient<- metadata$Patient
  
  #plot
  p<- ggplot(PCA, aes(x=PC1, y=PC2, color=group, shape=Patient))+
  geom_point(size=8)+
  scale_color_manual(values=Groupcolors)+
  scale_shape_manual(values=c(15,20,43,18))+
  theme_classic()+
  ylim(miny,maxy)+
  xlim(minx,maxx)+
  ylab(paste("PC2: ", PC2Variance, "% Variance", sep =""))+
  xlab(paste("PC1: ", PC1Variance, "% Variance", sep =""))+
  #geom_text(aes(label=Sample),hjust=.5, size= 5,vjust=-1.2, color="black")+
  theme(axis.text.y   = element_text(size=10, colour = "black"),
        axis.title.y  = element_text(size=14, colour = "black"),
        axis.title.x  = element_text(size=14, colour = "black"),
        axis.text.x  = element_text(size=10, colour = "black"),
        legend.background = element_rect(fill=NA, size=1, 
                                         linetype="solid", color="black"),
        legend.title=element_blank(),
        legend.key.size = unit(.4, "cm"),
        legend.spacing.y = unit(0, "mm"),
        legend.position = c(.9,.85),
        panel.border = element_rect(colour = "black", fill=NA, size=2))
  return(p)
}

MyPCA_TechNotCollapsed<- function(counts, meta,PC1, PC2, miny, maxy, minx, maxx){
  #Rename samples
  TechNotCollapsed<- counts
  colnames(TechNotCollapsed)<- c(paste("A.0.", 1:3,sep=""), paste("B.0.", 1:3,sep=""), "C.0", "D.0", paste("A.1.", 1:3,sep=""), paste("B.1.", 1:3,sep=""), "C.1", "D.1")


  #create DESeq2 object
  dds_TNC<- DESeqDataSetFromMatrix(countData=TechNotCollapsed, 
                             colData=meta, 
                             design=~sampleGroup)


  #Create model that takes into account sample differences
  ddsMF_TNC<- dds_TNC
  design(ddsMF_TNC)<- formula(~Patient + sampleGroup)

  #run DESeq2
  dds_TNC<- DESeq(ddsMF_TNC)
  
  #Run through MyPCA function to plot
  p<- MyPCA(dds_TNC, meta, PC1, PC2, Groupcolors, miny, maxy, minx, maxx)
  return(p)
}

#Plot heatmap: takes as input DE results (padj ordered lowest to highest, rlog transformed data, metadata (blocker status and Sample type), number of miRNA species to be included, and min and max subtract (used for heatmap aesthetics))

PlotHeatmap<- function(OrderedResults, rld, metadata, nspecies, MinSubtract, MaxSubtract, fontsize){
  #Select top sig DE miRNAs
  topgenes<- rownames(OrderedResults[1:nspecies,])
  mat<- assay(rld)[topgenes,]
  mat<- mat - rowMeans(mat)
  colnames(mat)<-paste(metadata$Sample," (",metadata$sampleGroup, ")",sep="")
  
  #Heatmap settings
  mat_breaks<- seq(min(mat-MinSubtract), max(mat-MaxSubtract), length.out=75)

  #Create metadata- to be used to color code to show which group the sample belongs to
  metadata<- data.frame(Group=metadata$sampleGroup, 
                        row.names=paste(metadata$Sample," (",metadata$sampleGroup, ")",sep=""),
                        Sample=metadata$Sample)

  #plot heatmap
  p<-pheatmap(mat, breaks = mat_breaks, 
            color =colorRampPalette( c("red", "black", "green"), space="rgb")(100),
            show_rownames = TRUE, show_colnames=FALSE,
            annotation = metadata,  
            annotation_colors= list(
              Group=c(Blocked="firebrick3", Unblocked="blue3"),
                      Sample=c(A="goldenrod1", 
                               B="violetred3", 
                               C="skyblue3", 
                               D="darkolivegreen4")),
            fontsize_row = fontsize,fontsize=12,treeheight_row=0,
            border_color= NA)
  return(p)
}


#Plot log of cpm of raw counts- coloring based on whether it's differentially expressed
CPMPlot<- function(CPM, Sample){
  #Get subset of CPM for labeling points- points that are both sig DE and are higher in unblocked than   blocked (beneath the slope line)
  CPM$HigherInUnblocked<- (CPM$Blocker-CPM$NoBlocker)<0
  SubsetToLabel<- CPM[CPM$HigherInUnblocked=="TRUE" &CPM$Color!="Not DE",]
  
  #Plot
  p<- ggplot(CPM, aes(x=NoBlocker, y=Blocker, color=Color))+
  geom_point(size=3)+
  geom_abline(slope=1, color="red")+
  scale_colour_manual(values=c("red","black", "blue"))+
  theme_classic()+
  xlim(0,6)+
  ylim(0,6)+
  xlab(paste("Log of Unblocked Counts Per Million: Sample ", Sample, sep =""))+
  ylab(paste("Log of Unblocked Counts Per Million: Sample ", Sample, sep =""))+
  geom_text_repel(data=SubsetToLabel, aes(x=NoBlocker, y=Blocker, label=rownames(SubsetToLabel)), color="black",  size=3.5)+
  theme(axis.text.y   = element_text(size=10, colour = "black"),
        axis.title.y  = element_text(size=14, colour = "black"),
        axis.title.x  = element_text(size=14, colour = "black"),
        axis.text.x  = element_text(size=10, colour = "black"),
        legend.background = element_rect(fill=NA, size=1, 
                                         linetype="solid", color="black"),
        legend.title=element_blank(),
        legend.key.size = unit(.4, "cm"),
        legend.spacing.y = unit(0, "mm"),
        legend.position = c(.15,.892),
        panel.border = element_rect(colour = "black", fill=NA, size=2))
  return(p)
  
}

#When comparing DE efficiency in blocked vs unblocked, this function used to find species unique to blocked/unblocked and find raw counts for those species. Used for plotting.

GetCounts_UniqueDEmiRNA<- function(BlockedSig, UnblockedSig, SigLevel, RawCounts){
  #Get data frame of sig miRNAs for blocked and unblocked at given significance level
  NBSig<-UnblockedSig[UnblockedSig$padj<SigLevel,]
  BSig<- BlockedSig[BlockedSig$padj<SigLevel,]
  
  #Get DE miRNA specific to blocked and unblocked
  DEInUB<- rownames(as.data.frame(BSig)) %in% rownames(as.data.frame(NBSig))
  DEInB<- rownames(as.data.frame(NBSig)) %in% rownames(as.data.frame(BSig))
  OnlyBlockedDE<- BSig[!DEInUB,]
  OnlyUnBlockedDE<- BSig[!DEInB,]
  
  #Get raw counts for these miRNAs
  counts_onlyblockedDE<- RawCounts[rownames(RawCounts) %in% rownames(OnlyBlockedDE),]
  counts_onlyblockedDE<- counts_onlyblockedDE[,c(1:6,9:14)]
  counts_onlyblockedDE<-counts_onlyblockedDE[order(counts_onlyblockedDE$A.0, decreasing=TRUE),]
  
  counts_onlyunblockedDE<- RawCounts[rownames(RawCounts) %in% rownames(OnlyUnBlockedDE),]
  counts_onlyunblockedDE<- counts_onlyunblockedDE[,c(1:6,9:14)]
  counts_onlyunblockedDE<-counts_onlyunblockedDE[order(counts_onlyunblockedDE$A.1, decreasing=TRUE),]
  
  return(list(counts_onlyblockedDE, counts_onlyunblockedDE))
  
}

```


#Two different input sources: 

*1) count matrix from BaseSpace (miRDeep2, no mismatches, isomirs and mature together) --> Downsampled to lowest read count
    a) Isomirs and mature
    b) Mature only
2) count matrix generated from miRDeep2 at the command line- mature only --> Downsampled to lowest read count


*This analysis performed in separate file: "FINAL_DifferentialExpression_BaseSpaceCountMatrix_BlockerProject.Rmd"

#For #1b and #2 count matrices, 2 different DESeq2 analyses performed:

1) Standard- Pairwise, technical replicates collapsed, all mature species included
2) Excluding targets- same as standard, but only including NON TARGET mature species
    a) excluding just the actual targets
    b) excluding the actual targets + members of the same family

Matrix #1 used initially, but after determining that a mismatch of 1 should be allowed (i.e., miRDeep2 default settings), the miRDeep2 analysis was rerun at the command line (matrix #2) with default miRDeep2 parameters. Both analyses used the same input bam file, generated from BaseSpace (splice-aware STAR alignmenet)


#Notes on DESeq2 Analysis:

-Since there are 4 sample types (A/B/C/D) each with at least one blocked/unblocked pair, pairwise DESeq2 analysis is performed

-Samples A and B have 3 technical replicates, for both blocked and unblocked groups. These technical replicates are collapsed using an internal DESeq2 function. These technical replicates have varying RNA input amounts into library construction, but this was determined to have no effect on the resulting libraries (See ___.Rmd). This is additionally confirmed by PCA clustering here.

#Read in count data, reformat:
```{r}
#Set working directory
wd<- "//Cifs2/rcdata$/Channing_miRNA_Blocker_Test/BlockerProject/BlockerProject/"

#Read in previously downsampled data: (in "FINAL_BlockersReduceTargetMapping.Rmd")
CL_countsInput_Downsampled<- read.csv(paste(wd, "RawData/Downsampled_miRNA_CLmiRDeep2Results_AllSamples_DefaultmiRDeep2Settings.csv",sep=""))
rownames(CL_countsInput_Downsampled)<-CL_countsInput_Downsampled$X
CL_countsInput_Downsampled<- CL_countsInput_Downsampled[,-1]

#Reorder 
CL_countsInput_Downsampled<- 
CL_countsInput_Downsampled[,c(1,3,5,7,9,11,13,15,2,4,6,8,10,12,14,16)]

#Convert to integer (removes decimal places)
CL_countsInput_Downsampled_Integer<- as.data.frame(lapply(CL_countsInput_Downsampled, as.integer))
rownames(CL_countsInput_Downsampled_Integer)<- rownames(CL_countsInput_Downsampled)
CL_Downsampled<- CL_countsInput_Downsampled_Integer
```


```{r}
#Need to detach DESeq (need to use DESeq2)- used for metaseqR
detach("package:metaseqR")
detach("package:DESeq")
```

General plotting variables
```{r}
Groupcolors<- c("firebrick3","blue3")
Samplecolors<- c("goldenrod1","violetred3", "skyblue3", "darkolivegreen4")

metadata<- data.frame(sampleGroup=rep(c("Unblocked", "Blocked"),4), Patient=c("A", "A", "B", "B", "C", "C", "D", "D"))
```


################################################################################
#Using Matrix #2 (generated by CL miRDeep2, 1 mismatch allowed and still considered mature):
################################################################################


Prep data for DESeq2, create metadata and DESeq2 object, then run DESeq2:
```{r}
#rename rows, removing input info
colnames(CL_Downsampled)<- c(rep("A.0", 3), rep("B.0", 3), "C.0", "D.0", rep("A.1", 3), rep("B.1", 3), "C.1", "D.1")

#Create metadata- will be used in all analyses- when collapsing technical replicates and when leaving separate:
sampleGroup<- as.factor(c(rep("Unblocked", 8), rep("Blocked", 8))) #set blocker status
sample<- colnames(CL_Downsampled) #this column is used to collapse technical replicates
Patient<- as.factor(c(rep("A", 3), rep("B", 3), "C", "D", rep("A", 3), rep("B", 3), "C", "D")) #this column is to control for sample to sample differences
meta<- data.frame(sample, sampleGroup, Patient)

#Run DESeq2- using custom function
#Technical replicates are collapsed and pairwise analysis performed

dds<- RunDESeq2(CL_Downsampled, meta, TRUE)
```

Visualize sample differences using PCA

Note: this function uses the dds object, but no information on sample group is input into the analysis (i.e., unsupervised clustering)


```{r}
#Plot PCA using DESeq2 functions, get % variance from here. Input for custom function.
plotPCA(varianceStabilizingTransformation(dds), intgroup="sampleGroup")

p<-MyPCA(dds, metadata, 43, 18, Groupcolors,-15,20, -20,25)
```



Is there any possibility this is due to collapsing of technical replicates? If we didn't collapse technical replicates, would we see all 3 A.1/3 B.1, etc still all clustering together?
Testing this..
```{r}
#Run deseq2 --> plot on PCA
p<- MyPCA_TechNotCollapsed(CL_Downsampled, meta, 43, 18, -20,27,-25,28)
```
So the answer to that question is yes- samples still cluster in a similar pattern, even if you don't collapse by technical replicates. A/B and C/D really are more similar to one another.


#Back to using the dds object with technical replicates collapsed:


Differential expression results:
```{r}
resOrdered<-GetDEResults(dds)
#DE: 37 <0.05 /25 <0.01

CL_DESeq2Results<-resOrdered

rownames(resOrdered[resOrdered$padj<0.01,])
```

```{r}
#Save off targets for later analysis
PotentialOffTargets<- resOrdered[resOrdered$padj<0.05,]
```

Visualize DE results with heatmap:
```{r}
#normalize results
rld<- rlog(dds)

p<-PlotHeatmap(resOrdered,rld, metadata,100, -.2,1,12)
```

Blocked/unblocked samples cluster together, as expected. Samples A/B and C/D also cluster together, as seen in the PCA results


Write to csv
```{r}
write.csv(resOrdered, paste(wd, "DESeq2Results/DESeq2Results_CLDownsampledMatureIsomir_121619.csv", sep =""))
```


#DESeq2 Analysis 2: removing targeted miRNAs

Two variations
    a) excluding just the actual targets
    b) excluding the actual targets + members of the same family

#Variation A (only actual targets removed) 

```{r}
#Get list of all targets
ActualTargets<- c('hsa-miR-486-5p','hsa-miR-92a-3p','hsa-miR-451a')

#subset counts (mature only) by non-target miRNAs
CL_AT<- CL_Downsampled[!(rownames(CL_Downsampled) %in% ActualTargets),]
```

Create dds object, run DESeq2
```{r}
#Run DESeq2: custom functions can be found in Functions.Rmd file
#Technical replicates are collapsed and pairwise analysis performed
dds<- RunDESeq2(CL_AT, meta, TRUE)
```


Visualize sample differences using PCA
```{r}
#Normalized (variance stabilized transformation), then run PCA
plotPCA(varianceStabilizingTransformation(dds, blind=TRUE), intgroup="sampleGroup")

p<- MyPCA(dds, metadata,45, 18, colors, -15,20, -20,25)

#Plot PCA for without collapsing technical replicates
p<- MyPCA_TechNotCollapsed(CL_AT, meta, 45,18, -20,20,-25,27)
```


Extract DESeq2 results and visualize with heatmap
```{r}
resOrdered<-GetDEResults(dds)
#31/21

rownames(resOrdered[resOrdered$padj<0.01,])

#normalize results
rld<- rlog(dds)
```

Plot heatmap
```{r}
p<- PlotHeatmap(resOrdered, rld, metadata,50, -.2,1, 10)
```

In heatmap of DE results, for A/B, samples cluster by SAMPLE and not BLOCKER STATUS, as we would expect if there are no huge off target effects caused by the blockers. However- C/D still cluster by blocker status. Will look at removing targets+target families, next, see if C/D still cluster by blocker status



Write results to file:
```{r}
#write to file 
write.csv(resOrdered, paste(wd, "DESeq2Results/DESeq2Results_ActualTargetsRemoved_DownSampledCLCountMatrix_Pairwise_121619.csv", sep = "" ))
```



#Variation B (actual targets + species in same family removed) 

Select all species with 92, 451, or 486 in name. Also select 2 hsa-mir-25 species (same family as 92a)
```{r}
Target_92<- rownames(CL_AT[grep("miR-92", rownames(CL_AT))[5:8],])
Target_451<- rownames(CL_AT[grep("miR-451", rownames(CL_AT))[11],])
Target_486<- rownames(CL_AT[grep("miR-486", rownames(CL_AT)),])
Target_25<- rownames(CL_AT[grep("miR-25", rownames(CL_AT)),])
TargetFamilies<- c(Target_451,Target_486,Target_92, Target_25)

#Remove the 8 additional species that are in the same family as the 3 actual targets
CL_NoTF<- CL_AT[!(rownames(CL_AT)%in%TargetFamilies),]
```



Create dds object, run DESeq2, visualize with PCA
```{r}
#Run DESeq2: custom functions can be found in Functions.Rmd file
#Technical replicates are collapsed and pairwise analysis performed
dds<- RunDESeq2(CL_NoTF,meta,TRUE)

#Plot PCA
plotPCA(varianceStabilizingTransformation(dds), intgroup="sampleGroup")
p<- MyPCA(dds, metadata,46,19,colors, -15,17, -20,25)

#Plot PCA without collapsing technical replicates
p<- MyPCA_TechNotCollapsed(CL_NoTF, meta, 46, 19, -18, 20, -25, 30)
```

Very similar to previous analysis (just actual targets removed), samples still cluster by SAMPLE TYPE not by blocker status

Extract DESeq2 results and visualize with heatmap
```{r}
resOrdered<- GetDEResults(dds)
#27 DE miRNAs- padj <0.05
#19 DE miRNAs- padj <0.01

rownames(resOrdered[resOrdered$padj<0.01,])

#normalize results
rld<- rlog(dds)
```

Plot heatmap
```{r}
p<-PlotHeatmap(resOrdered, rld, metadata, 20, -.2,1.2,14)

```

Samples cluster fully by sample type, not blocker status!

IMPORTANT: when you remove not just the actual targets, but also the actual targets + species from the same family (as above), samples cluster FULLY by sample type, not blocker status- even for samples C/D

Compare this to when you remove just the actual targets: even if you select 100 miRNAs, samples still don't fully cluster by sample type.

This just further increases our confidence that there are very few off target effects caused by blockers (other than blocking members of the same family, which can't really be fully avoided)


Write results to file:
```{r}
#write to file 
write.csv(resOrdered, paste(wd, "DESeq2Results/DESeq2Results_ActualTargets+FamiliesRemoved_DownSampledCLCountMatrix_Pairwise_121619.csv", sep = "" ))
```


#Graphing results in a different way: Log of CPM for each miRNA, comparing blocked vs unblocked values

For coloring sig/not sig expressed species, using CL mature downsampled counts- isomirs not included

```{r}
#calculating CPM
cpmInput<- CL_Downsampled

#calculate factor to multiply all counts for to convert to CPM
totalCounts<- colSums(cpmInput)
TotalCountsPM<- 1000000/totalCounts

#Convert all counts to CPM
CPMFinal<-cpmInput*TotalCountsPM
#get rid of any miRNAs that have a 0 counts for any sample- can't take log2 of it
CPMFinal<- CPMFinal[apply(CPMFinal, 1, function(row) all(row !=0 )),]

#log2 of CPM
CPMLog<- log10(CPMFinal+1) #+1 optional here
#CPMLog<-CPMFinal

#Get miRNA info. Need: whether or not it was targeted + whether or not it was DE
#Add in target info- add column for whether or not the species was targeted
targets<- c('hsa-miR-486-5p','hsa-miR-92a-3p','hsa-miR-451a')
CPMLog$target<- rownames(CPMLog) %in% targets

#Add in DE info- using DESeq2 results from targets included analysis. Import data here.
resOrdered<-CL_DESeq2Results
sig<- resOrdered[resOrdered$padj<0.01,]
CPMLog$Sig<- rownames(CPMLog) %in% rownames(sig)

#Add in third column, used to color points in graph. Add together target/sig columns. If it's a target and sig, it'll be 2, if it's just sig it'll be 1, if it's neither it'll be 0.
TargetSig<- CPMLog[,17:18]
CPMLog$Color<- as.factor(rowSums(TargetSig))
CPMLog$Color<- gsub(2, "Target", CPMLog$Color)
CPMLog$Color<- gsub(1, "DE", CPMLog$Color)
CPMLog$Color<- gsub(0, "Not DE", CPMLog$Color)
CPMLog[grep("hsa-miR-451a", rownames(CPMLog)),19]<- "Target"

#Add in a fourth column, used to add label to point if the species is a target or in a target family
CPMLog$TargetFamily<- rownames(CPMLog) %in% c(TargetFamilies,ActualTargets)
```


```{r}
#Looking at sample A- get average for all 3 Inputs
CPMLogA<- CPMLog[,c(1:3, 9:11,19,20)]
CPMLogA$NoBlocker<- rowMeans(CPMLogA[,1:3])
CPMLogA$Blocker<- rowMeans(CPMLogA[,4:6])
CPMLogA<- CPMLogA[,c(9,10,7,8)]

pA<- CPMPlot(CPMLogA, "A")
```

Most species follow the same general trend for both blocked and unblocked- i.e., overall similar expression pattern. EXCEPT for the targets (plus a few off target effects) 

```{r}
#Looking at sample B- get average for all 3 Inputs
CPMLogB<- CPMLog[,c(4:6, 12:14,19,20)]
CPMLogB$NoBlocker<- rowMeans(CPMLogB[,1:3])
CPMLogB$Blocker<- rowMeans(CPMLogB[,4:6])
CPMLogB<- CPMLogB[,c(8,9,7,10)]

pB<- CPMPlot(CPMLogB, "B")
```

```{r}
#For sample C
CPMLogC<- CPMLog[,c(7, 15,19,20)]
colnames(CPMLogC)<- c("NoBlocker", "Blocker", "Color")

pC<- CPMPlot(CPMLogC, "C")
```

```{r}
#For sample D
CPMLogD<- CPMLog[,c(8, 16,19,20)]
colnames(CPMLogD)<- c("NoBlocker", "Blocker", "Color")

pD<- CPMPlot(CPMLogD, "D")
```


```{r}
#Averaging values across al 4 samples
CPMLogAll<- CPMLog
CPMLogAll$NoBlocker<- rowMeans(CPMLogAll[,1:8])
CPMLogAll$Blocker<- rowMeans(CPMLogAll[,9:16])
CPMLogAll<- CPMLogAll[,c(20,21,19)]

pMean<- CPMPlot(CPMLogAll, "All")
```

Putting A-D together on one plot:
```{r}
grid.arrange(pA+ggtitle(""), pB+ggtitle(""), pC+ggtitle(""), pD+ggtitle(""), nrow=2)

```

##########################################################################################
#Improvements with blocker: does using blockers allow for the discovery of more DE miRNAs?
###########################################################################################

General idea: compare the number of significantly DE species between A and B in both blocked and unblocked samples.

Two analyses:

1) The 3 A.0 samples vs the 3 B.0 samples
2) The 3 A.1 samples vs the 3 B.1 samples

Both use the mature, downsampled BaseSpace matrix

Analysis #1: DESeq2 using A0 vs B0 - here, technical replicates are NOT collapsed
```{r}
#rename rows, removing input info
A0_B0<- CL_Downsampled[,c(1:6)]
colnames(A0_B0)<- c("A.0.1", "A.0.2", "A.0.3", "B.0.1", "B.0.2", "B.0.3")

#create metadata
sampleGroup<- c(rep("NoBlocker", 3), rep("Blocker", 3))
sample<- colnames(A0_B0)
meta<- data.frame(sample, sampleGroup)

#create DESeq2 object
dds<- DESeqDataSetFromMatrix(countData=A0_B0, 
                             colData=meta, 
                             design=~sampleGroup)


dds<- DESeq(dds)

res<- results(dds)
res<- na.omit(res)
resOrdered<- res[order(res$padj),]
nrow(resOrdered[resOrdered$padj<0.05,])
#89 DE miRNAs
noblocker_AvB<- resOrdered
noblocker_sig_AvB<-noblocker_AvB[noblocker_AvB$padj<0.05,]
```

DESeq2 using A1 vs B1- here, technical replicates are NOT collapsed
```{r}
#rename rows, removing input info
A1_B1<- CL_Downsampled[,c(9:14)]
colnames(A0_B0)<- c("A.1.1", "A.1.2", "A.1.3", "B.1.1", "B.1.2", "B.1.3")

#create metadata
sampleGroup<- c(rep("NoBlocker", 3), rep("Blocker", 3))
sample<- colnames(A1_B1)
meta<- data.frame(sample, sampleGroup)

#create DESeq2 object
dds<- DESeqDataSetFromMatrix(countData=A1_B1, 
                             colData=meta, 
                             design=~sampleGroup)


dds<- DESeq(dds)

res<- results(dds)
res<- na.omit(res)
resOrdered<- res[order(res$padj),]
nrow(resOrdered[resOrdered$padj<0.05,])
#119 DE miRNAs
blocker_AvB<- resOrdered
blocker_sig_AvB<-blocker_AvB[blocker_AvB$padj<0.05,]
```


```{r}
#So that's a....
(119-89)/119
#25% increase in the number of DE genes!!
```

Plot these differences in DE miRNAs for 3 different cutoffs. Three graphs:

1) Number of DE species for blocked/unblocked at 3 sig cutoffs
2) Number of DE Species UNIQUE to that group for blocked/unblocked at 3 sig cutoffs
3) Median raw read count for these UNIQUE DE species at 3 sig cutoffs
```{r}
#padj<0.01
Sig_Counts_Blocked_0.01<- GetCounts_UniqueDEmiRNA(blocker_AvB, noblocker_AvB, 0.01, CL_Downsampled)[[1]]
Sig_Counts_UnBlocked_0.01<- GetCounts_UniqueDEmiRNA(blocker_AvB, noblocker_AvB, 0.01, CL_Downsampled)[[2]]
Median_Blocked_0.01<- median(rowMeans(Sig_Counts_Blocked_0.01[,1:6,])) #14.33333
Median_UnBlocked_0.01<-median(rowMeans(Sig_Counts_UnBlocked_0.01[,7:12,])) #651
NSig_Blocked_0.01<-nrow(Sig_Counts_Blocked_0.01) #30
NSig_UnBlocked_0.01<-nrow(Sig_Counts_UnBlocked_0.01) #16

#padj<0.05
Sig_Counts_Blocked_0.05<- GetCounts_UniqueDEmiRNA(blocker_AvB, noblocker_AvB, 0.05, CL_Downsampled)[[1]]
Sig_Counts_UnBlocked_0.05<- GetCounts_UniqueDEmiRNA(blocker_AvB, noblocker_AvB, 0.05, CL_Downsampled)[[2]]
Median_Blocked_0.05<-median(rowMeans(Sig_Counts_Blocked_0.05[,1:6,])) #12.33333
Median_UnBlocked_0.05<-median(rowMeans(Sig_Counts_UnBlocked_0.05[,7:12,])) #249.4167
NSig_Blocked_0.05<-nrow(Sig_Counts_Blocked_0.05) #56
NSig_UnBlocked_0.05<-nrow(Sig_Counts_UnBlocked_0.05) #30

#padj<0.1
Sig_Counts_Blocked_0.1<- GetCounts_UniqueDEmiRNA(blocker_AvB, noblocker_AvB, 0.1, CL_Downsampled)[[1]]
Sig_Counts_UnBlocked_0.1<- GetCounts_UniqueDEmiRNA(blocker_AvB, noblocker_AvB, 0.1, CL_Downsampled)[[2]]
Median_Blocked_0.1<-median(rowMeans(Sig_Counts_Blocked_0.1[,1:6,])) #8
Median_UnBlocked_0.1<-median(rowMeans(Sig_Counts_UnBlocked_0.1[,7:12,])) #254.6667
NSig_Blocked_0.1<-nrow(Sig_Counts_Blocked_0.1) #77
NSig_UnBlocked_0.1<-nrow(Sig_Counts_UnBlocked_0.1) #33

#Dataframe summarizing n sig miRNAs + median counts for sig miRNAs specific to blocked/unblocked, for the 3 different significance levels
Sig_BvNB<- data.frame(SignificanceLevel= as.factor(c(rep(0.01, 2), rep(0.05 ,2), rep(0.1, 2))),
                      SignificantSpecies=c(nrow(blocker_AvB[blocker_AvB$padj<0.01,]), nrow(noblocker_AvB[noblocker_AvB$padj<0.01,]),
                                           nrow(blocker_AvB[blocker_AvB$padj<0.05,]), nrow(noblocker_AvB[noblocker_AvB$padj<0.05,]),
                                           nrow(blocker_AvB[blocker_AvB$padj<0.1,]), nrow(noblocker_AvB[noblocker_AvB$padj<0.1,])),
                      UniqueSignificantSpecies=c(NSig_Blocked_0.01, NSig_UnBlocked_0.01,
                                           NSig_Blocked_0.05, NSig_UnBlocked_0.05,
                                           NSig_Blocked_0.1, NSig_UnBlocked_0.1),
                      MedianUniqueCounts= c(Median_Blocked_0.01, Median_UnBlocked_0.01,
                                            Median_Blocked_0.05, Median_UnBlocked_0.05,
                                            Median_Blocked_0.1, Median_UnBlocked_0.1),
                      BlockerStatus= c(rep(c("Blocked", "Unblocked"), 3)))

#Plot this data on two separate graphs
p_NSpecies<- ggplot(Sig_BvNB, aes(y=SignificantSpecies, x=SignificanceLevel, fill=BlockerStatus))+
  geom_bar(position="dodge", stat="identity")+
  theme_classic()+
  xlab("Significance Level Threshold")+
  ylab("Number of significantly DE species")+
  scale_fill_manual("legend", values=c("Blocked"="firebrick3", "Unblocked"="blue3"))+
  ylim(0,200)+
  theme(axis.text.y   = element_text(size=10, colour = "black"),
        axis.title.y  = element_text(size=12, colour = "black"),
        axis.text.x   = element_text(size=10, colour = "black"),
        axis.title.x  = element_text(size=12, colour = "black"),
        panel.border = element_rect(colour = "black", fill=NA, size=2),
        legend.title=element_blank(),
        legend.position = c(.85,.96),
        legend.background = element_rect(fill=NA, size=1, linetype="solid", color="black"),
        axis.text.x.bottom  =element_text(size=15),
        legend.key.size = unit(.4, "cm"),
        legend.spacing.y = unit(0, "mm"))
  
p_UniqueNSpecies<-  ggplot(Sig_BvNB, aes(y=UniqueSignificantSpecies, x=SignificanceLevel, fill=BlockerStatus))+
  geom_bar(position="dodge", stat="identity")+
  theme_classic()+
  xlab("Significance Level Threshold")+
  ylab("Number of significantly DE species (for species unique to specific sample group)")+
  scale_fill_manual("legend", values=c("Blocked"="firebrick3", "Unblocked"="blue3"))+
  ylim(0,200)+
  theme(axis.text.y   = element_text(size=10, colour = "black"),
        axis.title.y  = element_text(size=12, colour = "black"),
        axis.text.x   = element_text(size=10, colour = "black"),
        axis.title.x  = element_text(size=12, colour = "black"),
        panel.border = element_rect(colour = "black", fill=NA, size=2),
        legend.title=element_blank(),
        legend.position = c(.85,.96),
        legend.background = element_rect(fill=NA, size=1, linetype="solid", color="black"),
        axis.text.x.bottom  =element_text(size=15),
        legend.key.size = unit(.4, "cm"),
        legend.spacing.y = unit(0, "mm")) 

p_MedianCounts<- ggplot(Sig_BvNB, aes(y=MedianUniqueCounts, x=SignificanceLevel, fill=BlockerStatus))+
  geom_bar(position="dodge", stat="identity")+
  theme_classic()+
  xlab("Significance Level Threshold")+
  ylab("Median raw counts (for species unique to specific sample group)")+
  scale_fill_manual("legend", values=c("Blocked"="firebrick3", "Unblocked"="blue3"))+
  theme(axis.text.y   = element_text(size=10, colour = "black"),
        axis.title.y  = element_text(size=12, colour = "black"),
        axis.text.x   = element_text(size=10, colour = "black"),
        axis.title.x  = element_text(size=12, colour = "black"),
        panel.border = element_rect(colour = "black", fill=NA, size=2),
        legend.title=element_blank(),
        legend.position = c(.85,.96),
        legend.background = element_rect(fill=NA, size=1, linetype="solid", color="black"),
        axis.text.x.bottom  =element_text(size=15),
        legend.key.size = unit(.4, "cm"),
        legend.spacing.y = unit(0, "mm"))
  

grid.arrange(p_NSpecies+ggtitle(""),p_UniqueNSpecies, p_MedianCounts+ggtitle(""), nrow=1)

ggarrange(p_NSpecies, p_UniqueNSpecies, p_MedianCounts, nrow=1)

```


Overall, much higher total read counts for no blocker-specific sig DE miRNAs. Supports the idea that without blocker, can't get really good discrimination of lowly expressed miRNAs!


#Potential Off target effects: looking more closely

In original, targets-included CL analysis, there were 25 DE miRNAs at padj<0.01. How many of these are target or target families? How many were downregulated in blocker group? What were the mean raw counts for these?

```{r}
#Get 25 differentially expressed miRNAs from targets included analysis
OT<- as.data.frame(PotentialOffTargets[,c(2,6)])

#Add in raw count data for these miRNAs
RawCounts<- as.data.frame(rowMeans(CL_Downsampled[rownames(CL_Downsampled) %in% rownames(OT),]))
OT<- merge(OT, RawCounts, by=0)
colnames(OT)<- c("Species", "log2FoldChange", "padj", "MeanRawCounts")

#Annotate by blocker/blocker family status
OT$Target<- OT$Species %in% ActualTargets
OT$TargetFamily<- OT$Species %in% TargetFamilies
OT$EitherTarget_TargetFamily<- as.logical(rowSums(OT[,c(5,6)]))

#Annotate with padj information
OT$log2padj<- log2(OT$padj)
```

```{r}
#plot
p<-ggplot(OT,aes(x=log2FoldChange, y= padj, color=EitherTarget_TargetFamily))+
  geom_point(aes(size=MeanRawCounts))+
  scale_size_continuous(range = c(1.5, 35), guide=FALSE)+
  scale_color_manual(values=colors)+
  scale_x_continuous(limits = c(-6, 7))+
  scale_y_continuous(limits = c(-0.008, .05))+
  theme_classic()+
  xlab("log2 foldchange of differential expression")+
  ylab("Adjusted p value")+
  geom_vline(xintercept=0,linetype="dotted", color= "black", size=1.25)+
  guides(colour = guide_legend(override.aes = list(size=5)))+
  theme(axis.text.y   = element_text(size=10, colour = "black"),
        axis.title.y  = element_text(size=14, colour = "black"),
        axis.title.x  = element_text(size=14, colour = "black"),
        axis.text.x  = element_text(size=10, colour = "black"),
        legend.background = element_rect(fill=NA, size=1, 
                                         linetype="solid", color="black"),
         legend.title=element_blank(),
        legend.key.size = unit(.4, "cm"),
        legend.spacing.y = unit(0, "mm"),
        legend.position = c(.8,.93),
        panel.border = element_rect(colour = "black", fill=NA, size=2))
p


```

Write potential off targets to file
```{r}
write.csv(OT, paste(wd, "PotentialOffTargets_FromCL_TargetsIncludedAnalysis.csv", sep ="") )
```

