---
title: "R Notebook"
output: html_notebook
---

#Large cohort: impact of blockers on target detection, improvements with blockers

Only data from the large cohort is used here. Count matrices for this cohort were all generated using the Basespace small RNA app. All analyses using the large cohort are included here for clarity (Figure 1C, 2E, 4B, and supplemental figure 1B)

Import libraries
```{r}
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(reshape2))
suppressPackageStartupMessages(library(purrr))
suppressPackageStartupMessages(library(tibble))
suppressPackageStartupMessages(library(metaseqR))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(DESeq2))
suppressPackageStartupMessages(library(pheatmap))
suppressPackageStartupMessages(library(ggrepel))
suppressPackageStartupMessages(library(ggpubr))
windowsFonts("Arial" = windowsFont("Arial"))
```

Functions
```{r}
#For plotting log2 CPM values nicely
CPMPlot<- function(CPM, Sample){
  #Get subset of CPM for labeling points- points that are both sig DE and are higher in unblocked than   blocked (beneath the slope line)
  CPM$HigherInUnblocked<- (CPM$Blocker-CPM$NoBlocker)<0
  SubsetToLabel<- CPM[CPM$HigherInUnblocked=="TRUE" &CPM$Color!="Not DE",]
  
  #Plot
  p<- ggplot(CPM, aes(x=NoBlocker, y=Blocker, color=Color))+
  geom_point(size=3)+
  geom_abline(slope=1, color="red")+
  scale_colour_manual(values=c("red","black", "blue"))+
  theme_classic()+
  xlim(0,6)+
  ylim(0,6)+
  xlab(paste("Log of Unblocked Counts Per Million"))+
  ylab(paste("Log of Blocked Counts Per Million"))+
  geom_text_repel(data=SubsetToLabel, aes(x=NoBlocker, y=Blocker, label=rownames(SubsetToLabel)), color="black",  size=5, fontface="bold")+
  ggtitle(Sample)+
  theme(axis.text.y   = element_text(size=14, colour = "black", face="bold"),
        axis.title.y  = element_text(size=14, colour = "black", face="bold"),
        axis.title.x  = element_text(size=14, colour = "black", face="bold"),
        axis.text.x  = element_text(size=14, colour = "black", face="bold"),
        legend.background = element_rect(fill=NA, size=1, 
                                         linetype="solid", color="black"),
        legend.title=element_blank(),
        legend.key.size = unit(.4, "cm"),
        legend.spacing.y = unit(0, "mm"),
        legend.position = c(.15,.892),
        legend.text = element_text(size=18, face="bold"),
        panel.border = element_rect(colour = "black", fill=NA, size=2),
        plot.title=element_text(size=14, face="bold"))
  return(p)
  
}

#Function that takes a column of counts and returns the number of species that pass the given list of threhsolds
CountsPassingThreshold<- function(counts, threshold){
  Passthreshold<-c()
  for (i in 1:length(threshold)){
    Passthreshold[i]<- sum(counts>threshold[i])
    
  }
  return(Passthreshold)
}

```

#Overall steps:

1. Read in unblocked data-from multiple folders- and merge into 1. Filter based on read count. Write file out for later use, as this step is computationally intensive.
2. Read in blocked data-from multiple folders- and merge into 1. Filter based on read count. Write file out for later use, as this step is computationally intensive.
3. Read in filtered/merged unblocked data from step 1 --> Further filtering + plotting % map to targets
4. Read in filtered/merged blocked data from step 2 --> Further filtering + plotting % map to targets
5. For blocked  and unblocked  data, plot the % mapping to any target (Supplemental Figure 1B)
6. For blocked and unblocked datasets, plot percentage mapping to each of the 3 targets + all other non-heme miRs (not used as figure)
7. For blocked and unblocked datasets, plot percentage mapping to each of the 11 targets (mature and isomir from target groups) (Figure 1C)
8. Differential expression- unpaired, DESeq2. Used for coloring sig points in Log2 CPM plot 
9. Plot log2 CPM values for blocked and unblocked datasets, coloring based on sig (from previous step) (Figure 2E)
10. Plot the number of species passing count thresholds for blocked and unblocked datasets (Figure 4B)
11. Plotting off target effects (only used in Supplemental table 1)



#1: Read in unblocked data-from multiple folders- and merge into 1
```{r}
unblocked_readwd<-"Z:/Small_RNA/BaseSpace/"

#Get filenames needed
filenames1<-c("Hersh_PT8_Rerun_Small_RNA_11072017_counts",
             "Hersh-miRNA-PT9Reruns_Small_RNA_12182017_counts")

filenames2<-c("9256704449-2_Small_RNA_05252017_counts",
             "9416449660-3_Small_RNA_06282017_counts",
             "9597590455-6_Small_RNA_06292017_counts",
             "9959254099-1_Small_RNA_10062017_counts",
             "10272166057-5_Small_RNA_08212017_counts",
             "10284133418-9_Small_RNA_08212017_counts")

filenames3<- c("AERD_miRNA_Sequencing_Small_RNA_10062017_counts",
               "Kelan_Viva_miRNA_Small_RNA_09272017_counts"
        
)

filenames4<-c("Hersh_PT8_Rerun_Small_RNA_11072017_counts",
             "Hersh-miRNA-PT9Reruns_Small_RNA_12182017_counts",
             "9256704449-2_Small_RNA_05252017_counts",
             "9416449660-3_Small_RNA_06282017_counts",
             "9597590455-6_Small_RNA_06292017_counts",
             "10272166057-5_Small_RNA_08212017_counts",
             "10284133418-9_Small_RNA_08212017_counts"
            )

filenames<-filenames4
#Read in files
unblocked_input<-list()
for (i in 1:length(filenames)){
        df<-read.csv(paste(unblocked_readwd, filenames[i], ".csv", sep=""), stringsAsFactors = FALSE)
        df<-df[-1,]
        unblocked_input[[i]]<-df
}


```


Merge into one dataset, filter based on total reads per sample and per miRNA
```{r}
#Merge into one dataset
unblocked_all_input<- unblocked_input %>% purrr::reduce(full_join, by = "X")

#Convert to integer, remove NA
Rownames<-unblocked_all_input$X
rownames(unblocked_all_input)<-Rownames
unblocked_all<-unblocked_all_input[,-1]
unblocked_all<- as.data.frame(apply(unblocked_all, 2, as.integer))
rownames(unblocked_all)<-Rownames
unblocked_all<-na.omit(unblocked_all)

#Remove any samples with fewer than 2million total reads
unblocked_RemoveSample<- unblocked_all[,colSums(unblocked_all)>2000000]

#Downsample to common depth
unblocked_DS<- downsample.counts(unblocked_RemoveSample, set.seed(42)) 
#unblocked_DS$X<-Rownames
unblocked_DS$X<-rownames(unblocked_DS)

#sort and remove sequences from rownames, and convert to integer
unblocked_sort<-  unblocked_DS[order(unblocked_DS[,1], decreasing = TRUE),]
unblocked_sort$X<-rownames(unblocked_sort) 
unblocked_sort$X<- gsub(".*_", "", unblocked_sort$X)

#Collapse counts- multiple entries for a single species are collapsed into one
unblocked_collapse<- unblocked_sort %>% as.tibble() %>% group_by(X) %>% summarise_all(funs(if(is.numeric(.)) sum(.) else "Total")) %>% as.data.frame
rownames(unblocked_collapse)<-unblocked_collapse$X
unblocked_collapse<-unblocked_collapse[,-1]

#write to file for later use
writewd<-"C:/Users/Jenna/Documents/BlockerProject/"
write.csv(unblocked_collapse, paste(writewd, "LargeCohort_RawData_Unblocked.csv", sep=""))
```

#2: Read in blocked data-from multiple folders- and merge into 1. Filter based on read count. Write file out for later use, as this step is computationally intensive.
```{r}
blocked_readwd<-"Z:/COPD/miRNAseq/BaseSpace/"

#Get filenames needed
filenames<-c("21801069331-0_GECOPD_miRNASeq _Small_RNA_11222019_counts",
             "21801069331-0_GECOPD_miRNASeq _Small_RNA_12032019_counts",
             "21801069331-0_GECOPD_miRNASeq _Small_RNA_12042019_counts",
             "21801069331-0_GECOPD_miRNASeq _Small_RNA_12102019_counts",
             "21801069331-0_GECOPD_miRNASeq _Small_RNA_12112019_counts",
             "24200081892-6_GECOPD_P6_miRNASeq _Small_RNA_03092020_counts")

#Read in files
blocked_input<-list()
for (i in 1:length(filenames)){
        df<-read.csv(paste(blocked_readwd, filenames[i], ".csv", sep=""), stringsAsFactors = FALSE)
        df<-df[-1,]
        blocked_input[[i]]<-df
}

```

#Merge into one dataset, filter based on total reads per sample and per miRNA
```{r}
#Merge into one dataset
blocked_all_input<- blocked_input %>% purrr::reduce(full_join, by = "X")

#Convert to integer, remove NA
Rownames<-blocked_all_input$X
rownames(blocked_all_input)<-Rownames
blocked_all<-blocked_all_input[,-1]
blocked_all<- as.data.frame(apply(blocked_all, 2, as.integer))
rownames(blocked_all)<-Rownames
blocked_all<-na.omit(blocked_all)


#Remove any samples with fewer than 2million total reads
blocked_RemoveSample<- blocked_all[,colSums(blocked_all)>2000000]

#Downsample to common depth
blocked_DS<- downsample.counts(blocked_RemoveSample, set.seed(42)) 
#blocked_DS$X<-Rownames
blocked_DS$X<-rownames(blocked_DS)

#sort and remove sequences from rownames, and convert to integer
blocked_sort<-  blocked_DS[order(blocked_DS[,1], decreasing = TRUE),]
blocked_sort$X<-rownames(blocked_sort)
blocked_sort$X<- gsub(".*_", "", blocked_sort$X)


#Collapse counts- multiple entries for a single species are collapsed into one
blocked_collapse<- blocked_sort %>% as.tibble() %>% group_by(X) %>% summarise_all(funs(if(is.numeric(.)) sum(.) else "Total")) %>% as.data.frame
rownames(blocked_collapse)<-blocked_collapse$X
blocked_collapse<-blocked_collapse[,-1]

#write to csv for later use
write.csv(blocked_collapse, paste(writewd, "LargeCohort_RawData_blocked.csv", sep=""))
```


#3: Read in filtered/merged unblocked data from step 1 --> Further filtering + plotting % map to targets
```{r}
#read in counts- pre-processing in above chunk
writewd<-"C:/Users/Jenna/Documents/BlockerProject/"
unblocked_collapse<-read.csv(paste(writewd, "LargeCohort_RawData_Unblocked.csv"), row.names = "X")

#Remove any species with < 10 total counts-optional
unblocked_RemoveLow<- unblocked_collapse[rowSums(unblocked_collapse) >10,]
unblocked_RemoveLow<-unblocked_collapse

#Convert to CPM
unblocked_CPM<-unblocked_RemoveLow/colSums(unblocked_RemoveLow) *1000000

#Get total number of CPM for 3 targets-will be multiple lines, grep for targets.Sum together target CPMs
AllTargets<-c("hsa-miR-486", "hsa-miR-92a", "hsa-miR-451a")

TargetCPM_1<-lapply(AllTargets, function(x) {unblocked_CPM[grep(x, rownames(unblocked_CPM), ignore.case = TRUE),]}) #get targets
TargetCPM_2<-lapply(TargetCPM_1, function(x) {x[!(rownames(x) %in% c("hsa-miR-4511", "hsa-miR-4510","hsa-mir-4511", "hsa-mir-4510")),]}) #remove miR-4511/4510
TargetCPM_3<- lapply( TargetCPM_2, function(x) {colSums(x[,1:ncol(x)])}) #merge multiple species in a family into one
TargetCPM<- as.data.frame(do.call(rbind, TargetCPM_3))
rownames(TargetCPM)<-AllTargets
TargetCPM<- TargetCPM[,order(colnames(TargetCPM))]


#Get percentage of total CPMs that 3 targets make up
unblocked_CPM<-unblocked_CPM[,order(colnames(unblocked_CPM))]
unblocked_TargetPercentageCPM<-TargetCPM/colSums(unblocked_CPM) *100

#add row with percent mapping to non-target
unblocked_TargetPercentageCPM["AllOthers" ,] <- 100-colSums(unblocked_TargetPercentageCPM)

```


#4: Read in filtered/merged blocked data from step 2 --> Further filtering + plotting % map to targets
```{r}
#red in counts- pre-processing in above chunka
blocked_collapse<-read.csv( paste(writewd, "LargeCohort_RawData_blocked.csv", sep=""), row.names = "X")

#Remove any species with < 10 total counts  
blocked_RemoveLow<- blocked_collapse[rowSums(blocked_collapse) >10,]
blocked_RemoveLow<-blocked_collapse

#Convert to CPM
blocked_CPM<-blocked_RemoveLow/colSums(blocked_RemoveLow) *1000000

#Get total number of CPM for 3 targets-will be multiple lines, grep for targets.Sum together target CPMs
AllTargets<-c("hsa-miR-486", "hsa-miR-92a", "hsa-miR-451a")
TargetCPM_1<-lapply(AllTargets, function(x) {blocked_CPM[grep(x, rownames(blocked_CPM), ignore.case=TRUE),]}) #get targets
TargetCPM_2<-lapply(TargetCPM_1, function(x) {x[!(rownames(x) %in% c("hsa-miR-4511", "hsa-miR-4510","hsa-mir-4511", "hsa-mir-4510")),]}) #remove miR-4511/4510
TargetCPM_3<- lapply( TargetCPM_2, function(x) {colSums(x[,1:ncol(x)])}) #merge multiple species in a family into one
TargetCPM<- as.data.frame(do.call(rbind, TargetCPM_3))
rownames(TargetCPM)<-AllTargets
TargetCPM<- TargetCPM[,order(colnames(TargetCPM))]


#Get percentage of total CPMs that 3 targets make up
blocked_CPM<-blocked_CPM[,order(colnames(blocked_CPM))]
blocked_TargetPercentageCPM<-TargetCPM/colSums(blocked_CPM) *100

#Optional: add row with percent mapping to non-target
blocked_TargetPercentageCPM["AllOthers" ,] <- 100-colSums(blocked_TargetPercentageCPM)
```

#5: For blocked  and unblocked  data, plot the % mapping to any target (Supplemental Figure 1B)
```{r}
#sum percent mapping for the 3 targets, take mean across all samples
blocked_all<- colSums(blocked_TargetPercentageCPM[1:3,])
unblocked_all<- colSums(unblocked_TargetPercentageCPM[1:3,])
df<-data.frame(Group=c("Blocked", "Unblocked"), 
               PercentMapping=c(mean(blocked_all), mean(unblocked_all)),
               sd=c(sd(blocked_all), sd(unblocked_all)))

#plot: 042820_LargeCohort_PercentMapping_AnyTarget_2Groups
p<-ggplot(df, aes(y=PercentMapping, x=reorder(Group, - PercentMapping),fill=Group))+
  geom_bar(stat="identity", position="dodge", width=.7)+ 
  theme_classic()+
  ylab("Percent of reads mapping to targets species")+
  xlab("")+
  scale_fill_manual("", values=c("Blocked"="firebrick3", "Unblocked"="turquoise3"))+
  geom_errorbar(aes(ymin=PercentMapping-sd, ymax=PercentMapping+sd), width=.1, size=1, position=position_dodge(.9),color="black")+
  ylim(0,100)+
  theme(axis.text.y   = element_text(size=12, color="black", face= "bold"),
          axis.text.x   = element_text(size=14, color="black", face="bold"),
          axis.title.y  = element_text(size=16, color="black", face="bold"),
          panel.border = element_rect(colour = "black", fill=NA, size=2),
          legend.title=element_blank(),
          legend.position = "none",
          legend.background = element_rect(fill=NA, size=1, linetype="solid", color="black"),
          axis.text.x.bottom  =element_text(size=20))
```

Significance test comparing blocked and unblocked datasets:
```{r}
t.test(unname(blocked_all), unname(unblocked_all))
```

#6: For blocked and unblocked datasets, plot percentage mapping to each of the 3 targets + all other non-heme miRs (not used as figure)
```{r}
#Add column for blocked/unblocked to "TargetPercentageCPM" dataframes
unblocked_TargetPercentageCPM$BlockerStatus<-rep("Unblocked", nrow(unblocked_TargetPercentageCPM))
blocked_TargetPercentageCPM$BlockerStatus<-rep("Blocked", nrow(blocked_TargetPercentageCPM))

#reformat for plotting
df_unblocked<-unblocked_TargetPercentageCPM
df_unblocked$Species<- rownames(df_unblocked)
df_unblocked_melt<- melt(df_unblocked, by="Species")

df_blocked<-blocked_TargetPercentageCPM
df_blocked$Species<- rownames(df_blocked)
df_blocked_melt<- melt(df_blocked, by="Species")

#Add column for ordering
df_unblocked_melt$Order<- rep(1, nrow(df_unblocked_melt))
df_blocked_melt$Order<- rep(2, nrow(df_blocked_melt))

#rbind unblocked and blocked dataframes together
bind_TargetPercentageCPM<- rbind(df_unblocked_melt, df_blocked_melt)
bind_TargetPercentageCPM$Order<- rep(seq(1,length(unique(bind_TargetPercentageCPM$Species)),1),
                                     nrow(bind_TargetPercentageCPM)/length(unique(bind_TargetPercentageCPM$Species)))
bind_TargetPercentageCPM$BlockerStatus<-factor(bind_TargetPercentageCPM$BlockerStatus, levels=c("Unblocked", "Blocked"))


#Plot
p<-ggplot(bind_TargetPercentageCPM, aes(x = reorder(Species, Order), y =value, fill=BlockerStatus)) +
        geom_boxplot(size=0.3,color="black",outlier.size = .7)+
        ylab("Percent of total reads mapping")+
        xlab("")+
        theme_classic()+
        scale_fill_manual("", values=c("Blocked"="firebrick3", "Unblocked"="turquoise3"))+
        theme(axis.title.y = element_text(face="bold", size=16),
              axis.text.y = element_text(face="bold", size=12, color="black"),
              axis.text.x = element_text(face="bold", size=16, color="black", angle=45,hjust=1),
              legend.text = element_text(face="bold", size=16),
              legend.title=element_blank(),
              panel.border = element_rect(colour = "black", fill=NA, size=2),
              strip.background = element_rect(color = "black", size = 2),
              legend.position=c(.72,.9),
              legend.background = element_rect(fill=NA, size=1, linetype="solid", color="black"),
              axis.ticks = element_line(size = 1.5,color="black"),
              axis.ticks.length=unit(.2, "cm"))

```

Get significance values for comparing blocked/unblocked in each species
```{r}
#Run t.test for all species
SplitDFBymiR<-split(bind_TargetPercentageCPM, bind_TargetPercentageCPM$Species)
Ttest<- unlist(lapply(SplitDFBymiR, function(x) {t.test(x[x$BlockerStatus=="Unblocked",4],x[x$BlockerStatus=="Blocked", 4])$p.value}))
```


```{r}
GetMeans<-bind_TargetPercentageCPM
GetMeans$Grouping<-paste(GetMeans$BlockerStatus,"_", GetMeans$Species)
GetMeans<- GetMeans %>% group_by(Grouping)  %>% mutate(mean=mean(value)) %>% as.data.frame()
GetMeans[!duplicated(GetMeans$mean),]
```

#6: For blocked and unblocked datasets, plot percentage mapping to each of the 11 species (mature and isomiRs in target group) (Figure 1C)

Get uncollapsed unblocked dataset:
```{r}
#Convert to CPM
unblocked_CPM<-unblocked_RemoveLow/colSums(unblocked_RemoveLow) *1000000

#Get total number of CPM for 3 targets-will be multiple lines, grep for targets.Sum together target CPMs
AllTargets<-c("hsa-miR-486", "hsa-miR-92a", "hsa-miR-451a")

TargetCPM_1<-lapply(AllTargets, function(x) {unblocked_CPM[grep(x, rownames(unblocked_CPM), ignore.case = TRUE),]}) #get targets
TargetCPM_2<-lapply(TargetCPM_1, function(x) {x[!(rownames(x) %in% c("hsa-miR-4511", "hsa-miR-4510","hsa-mir-4511", "hsa-mir-4510")),]}) #remove miR-4511/4510
TargetCPM<- as.data.frame(do.call(rbind, TargetCPM_2))
TargetCPM<- TargetCPM[,order(colnames(TargetCPM))]

#Get percentage of total CPMs that 3 targets make up
unblocked_CPM<-unblocked_CPM[,order(colnames(unblocked_CPM))]
unblocked_TargetPercentageCPM<-TargetCPM/colSums(unblocked_CPM) *100

#add row with percent mapping to non-target
unblocked_TargetPercentageCPM["AllOthers" ,] <- 100-colSums(unblocked_TargetPercentageCPM)
```

Get uncollapsed blocked dataset:
```{r}
#Convert to CPM
blocked_CPM<-blocked_RemoveLow/colSums(blocked_RemoveLow) *1000000

#Get total number of CPM for 3 targets-will be multiple lines, grep for targets.Sum together target CPMs
AllTargets<-c("hsa-miR-486", "hsa-miR-92a", "hsa-miR-451a")

TargetCPM_1<-lapply(AllTargets, function(x) {blocked_CPM[grep(x, rownames(blocked_CPM), ignore.case = TRUE),]}) #get targets
TargetCPM_2<-lapply(TargetCPM_1, function(x) {x[!(rownames(x) %in% c("hsa-miR-4511", "hsa-miR-4510","hsa-mir-4511", "hsa-mir-4510")),]}) #remove miR-4511/4510
TargetCPM<- as.data.frame(do.call(rbind, TargetCPM_2))
TargetCPM<- TargetCPM[,order(colnames(TargetCPM))]

#Get percentage of total CPMs that 3 targets make up
blocked_CPM<-blocked_CPM[,order(colnames(blocked_CPM))]
blocked_TargetPercentageCPM<-TargetCPM/colSums(blocked_CPM) *100

#add row with percent mapping to non-target
blocked_TargetPercentageCPM["AllOthers" ,] <- 100-colSums(blocked_TargetPercentageCPM)
```

#Merge blocked/unblocked uncollapsed datasets --> get % mapping to each miR --> plot (Figure 1C)
```{r}
#Add column for blocked/unblocked to "TargetPercentageCPM" dataframes
unblocked_TargetPercentageCPM$BlockerStatus<-rep("Unblocked", nrow(unblocked_TargetPercentageCPM))
blocked_TargetPercentageCPM$BlockerStatus<-rep("Blocked", nrow(blocked_TargetPercentageCPM))

#reformat for plotting
df_unblocked<-unblocked_TargetPercentageCPM
df_unblocked$Species<- rownames(df_unblocked)
df_unblocked_melt<- melt(df_unblocked, by="Species")

df_blocked<-blocked_TargetPercentageCPM
df_blocked$Species<- rownames(df_blocked)
df_blocked_melt<- melt(df_blocked, by="Species")

#Add column for ordering
df_unblocked_melt$Order<- rep(1, nrow(df_unblocked_melt))
df_blocked_melt$Order<- rep(2, nrow(df_blocked_melt))

#rbind unblocked and blocked dataframes together
bind_TargetPercentageCPM<- rbind(df_unblocked_melt, df_blocked_melt)
bind_TargetPercentageCPM$Order<- rep(seq(1,length(unique(bind_TargetPercentageCPM$Species)),1),
                                     nrow(bind_TargetPercentageCPM)/length(unique(bind_TargetPercentageCPM$Species)))
bind_TargetPercentageCPM$BlockerStatus<-factor(bind_TargetPercentageCPM$BlockerStatus, levels=c("Unblocked", "Blocked"))
```


#Collapsing technical replicates: some samples occur multiple times in blocked/unblocked sets. Collapse into one by taking the average.
```{r}
#Remove control samples from counts table- only keeping samples named as "S."
bind_TargetPercentageCPM<-bind_TargetPercentageCPM[grepl("S.", bind_TargetPercentageCPM$variable),]

#Rename samples in counts- remove the Trimmed, -x -y, etc
bind_TargetPercentageCPM$variable<-gsub("Trimmed.x", "", bind_TargetPercentageCPM$variable)
bind_TargetPercentageCPM$variable<-gsub("Trimmed.y", "", bind_TargetPercentageCPM$variable)
bind_TargetPercentageCPM$variable<-gsub("Trimmed", "", bind_TargetPercentageCPM$variable)
bind_TargetPercentageCPM$variable<-gsub("trimmed.x", "", bind_TargetPercentageCPM$variable)
bind_TargetPercentageCPM$variable<-gsub("trimmed.y", "", bind_TargetPercentageCPM$variable)
bind_TargetPercentageCPM$variable<-gsub("trimmed", "", bind_TargetPercentageCPM$variable)

#Add new column- new sampleID (sampleID_Blocked/Unblocked_Species)
bind_TargetPercentageCPM$NewSampleID<- paste(bind_TargetPercentageCPM$variable, bind_TargetPercentageCPM$BlockerStatus, bind_TargetPercentageCPM$Species, sep="_")

#Collapse the same sample IDs together
sampleIDsCollapsed_bind_TargetPercentageCPM<- bind_TargetPercentageCPM %>% select(c(value, NewSampleID)) %>% group_by(NewSampleID) %>% dplyr::summarize(value = mean(value, na.rm=TRUE)) %>% as.data.frame()

#Split back out into separate columns
sampleIDsCollapsed_bind_TargetPercentageCPM<- separate(data = sampleIDsCollapsed_bind_TargetPercentageCPM, col = NewSampleID, into = c("Sample", "BlockerStatus", "Species"), sep = "_")

#Add order column for plotting
sampleIDsCollapsed_bind_TargetPercentageCPM$Order<- rep(c(12,6,5,3,4,1,2,9,10,11,8,7))

#rename for using same plotting chunk
bind_TargetPercentageCPM<-sampleIDsCollapsed_bind_TargetPercentageCPM
```


Get significance values for comparing blocked/unblocked in each species
```{r}
#Run t.test for all species
SplitDFBymiR<-split(bind_TargetPercentageCPM, bind_TargetPercentageCPM$Species)
Ttest<- unlist(lapply(SplitDFBymiR, function(x) {t.test(x[x$BlockerStatus=="Unblocked",4],x[x$BlockerStatus=="Blocked", 4])$p.value}))
```

#Optional: get only paired samples --> paired analysis
Use tech reps collapsed as starting point
```{r}
#Read in paired samples
pairedSamples<-read.csv(paste(wd, "Paired_LargeCohortSamples_BlockedUnblocked.csv", sep=""))

#Add column for Subject_Species
pairedSamples$SubjectSpecies<- paste(pairedSamples$Subject, pairedSamples$Species, sep="")

test<-pairedSamples %>% group_by(SubjectSpecies) %>% mutate(minus=Blocked_value-Unblocked_value)  %>%as.data.frame()

test2<-test %>% group_by(Species) %>% summarize(value=mean(minus))%>%as.data.frame

p<-ggplot(test2, aes(x=Species, y=value))+
        geom_bar(stat="identity", position="dodge")+
        theme(axis.text.x=element_text(angle=45, hjust=1))

p<-ggplot(test, aes(x=Species, y= minus))+
        geom_boxplot()
```


```{r}
#Plot
p<-ggplot(bind_TargetPercentageCPM, aes(x = reorder(Species, Order), y =value, fill=BlockerStatus)) +
        geom_boxplot(size=0.3,color="black",outlier.size = .7)+
        ylab("Percent of total reads mapping")+
        xlab("")+
        theme_classic()+
        scale_fill_manual("", values=c("Blocked"="firebrick3", "Unblocked"="turquoise3"))+
        theme(axis.title.y = element_text(face="bold", size=16),
              axis.text.y = element_text(face="bold", size=12, color="black"),
              axis.text.x = element_text(face="bold", size=16, color="black", angle=45,hjust=1),
              legend.text = element_text(face="bold", size=16),
              legend.title=element_blank(),
              panel.border = element_rect(colour = "black", fill=NA, size=2),
              strip.background = element_rect(color = "black", size = 2),
              legend.position=c(.72,.9),
              legend.background = element_rect(fill=NA, size=1, linetype="solid", color="black"),
              axis.ticks = element_line(size = 1.5,color="black"),
              axis.ticks.length=unit(.2, "cm"))
```

Get significance values for comparing blocked/unblocked in each species
```{r}
#Run t.test for all species
SplitDFBymiR<-split(bind_TargetPercentageCPM, bind_TargetPercentageCPM$Species)
Ttest<- unlist(lapply(SplitDFBymiR, function(x) {t.test(x[x$BlockerStatus=="Unblocked",4],x[x$BlockerStatus=="Blocked", 4])$p.value}))
```


```{r}
GetMeans<-bind_TargetPercentageCPM
GetMeans$Grouping<-paste(GetMeans$BlockerStatus,"_", GetMeans$Species)
GetMeans<- GetMeans %>% group_by(Grouping)  %>% mutate(mean=mean(value)) %>% as.data.frame()
GetMeans[!duplicated(GetMeans$mean),]
```

#8: Differential expression- unpaired, DESeq2. Used for coloring sig points in Log2 CPM plot 

Filter, merge and export count matrices
```{r}
#Read in pre-processed count matrices, convert to CPM
writewd<-"C:/Users/Jenna/Documents/BlockerProject/"
blocked_counts<-read.csv( paste(writewd, "LargeCohort_RawData_blocked.csv", sep=""), row.names = "X")
blocked_CPM<- blocked_counts/colSums(blocked_counts)*1000000

unblocked_counts<-read.csv( paste(writewd, "LargeCohort_RawData_Unblocked.csv", sep=""), row.names = "X")
unblocked_CPM<- unblocked_counts/colSums(unblocked_counts)*1000000

#Remove genes without > 1 CPM in 50% of unblocked/blocked samples- in CPM data
blocked_CPMPass<-blocked_CPM[rowSums(blocked_CPM>1)>(ncol(blocked_CPM)/2),]
unblocked_CPMPass<-unblocked_CPM[rowSums(unblocked_CPM>1)>(ncol(unblocked_CPM)/2),]


#For the genes that passsed the CPM threshold, get raw (downsampled) counts
blocked_CountsPass<-blocked_counts[rownames(blocked_counts) %in% rownames(blocked_CPMPass),]
unblocked_CountsPass<-unblocked_counts[rownames(unblocked_counts) %in% rownames(unblocked_CPMPass),]

#merge blocked and unblocked- only keep if found in both (i.e., >50% have CPM>1 in both blocked and unblocked)
counts<- merge(blocked_CountsPass, unblocked_CountsPass, all=FALSE,by=0 )
counts[is.na(counts)]<-0

#Create metadata, export
meta_blocked<-data.frame(Samples=colnames(blocked_CPMPass), Group="Blocked")
meta_unblocked<-data.frame(Samples=colnames(unblocked_CPMPass), Group="Unblocked")
meta<-rbind(meta_blocked,meta_unblocked)
```


Perform differential expression between two groups
```{r}
#Import raw counts
readwd<-"C:/Users/Jenna/Documents/BlockerProject/"
countsInput<-read.csv(paste(readwd, "LargeCohort_AllSamples_RawCounts_CPMFilter.csv", sep="" ))
counts<-countsInput
rownames(counts)<-counts$Row.names
counts<-counts[,3:ncol(counts)]

#Collapse counts-remove everything after 3rd "-" (-1, -5p, etc), then merge and sum together based on new miR IDs
counts$geneIDs<- sub("^([^-]*-[^-]*-[^-]*).*", "\\1", rownames(counts))
counts$geneIDs<- gsub("mir", "miR", counts$geneIDs)
counts_collapsed<-counts %>% group_by(geneIDs) %>% summarise_each(funs(sum)) %>% as.data.frame
rownames(counts_collapsed)<-counts_collapsed$geneIDs
counts_collapsed<- counts_collapsed[,2:ncol(counts_collapsed)]

#Import metadata
meta<- read.csv(paste(readwd,"LargeCohort_Metadata.csv",sep=""))
meta$Group<- factor(meta$Group, levels=c("Unblocked", "Blocked"))

#Run DESeq2
dds<-DESeqDataSetFromMatrix(counts_collapsed, meta, design=~Group)
dds<-DESeq(dds)

#Extract and order results
res<-na.omit(results(dds))
resOrdered<-res[order(res$padj),]
nrow(resOrdered[resOrdered$padj<0.01,]) #177
nrow(resOrdered[resOrdered$padj<0.05,]) # 180
sig<-resOrdered[resOrdered$padj<0.01,]
sig<- sig[abs(sig$log2FoldChange)>1,] #71


```

#9: Plot log2 CPM values for blocked and unblocked datasets, coloring based on sig (from previous step) (Figure 2E)
```{r}
#Convert to CPM
CPM<- counts_collapsed/colSums(counts_collapsed) *1000000

#Get mean CPM for blocked and unblocked for every species
blockedSamples<-meta[meta$Group=="Blocked",]
blocked_CPM<-CPM[,colnames(CPM) %in% as.character(blockedSamples$Samples)]

unblockedSamples<-meta[meta$Group=="Unblocked",]
unblocked_CPM<-CPM[,colnames(CPM) %in% as.character(unblockedSamples$Samples)]

CPM_Mean<-data.frame(Blocker=rowMeans(blocked_CPM), NoBlocker=rowMeans(unblocked_CPM))

#log2 of CPM
CPMLog<- log2(CPM_Mean+1) #+1 optional here

#Get miRNA info. Need: whether or not it was targeted + whether or not it was DE
#Add in target info- add column for whether or not the species was targeted
ActualTargets<- c('hsa-miR-486','hsa-miR-92a','hsa-miR-451a')
CPMLog$target<- rownames(CPMLog) %in% ActualTargets

#Add in DE info- using DESeq2 results from targets included analysis. Import data here.
CPMLog$Sig<- rownames(CPMLog) %in% rownames(sig)

#Add in third column, used to color points in graph. Add together target/sig columns. If it's a target and sig, it'll be 2, if it's just sig it'll be 1, if it's neither it'll be 0.
CPMLog$Color<- as.factor(rowSums(CPMLog[,3:4]))
CPMLog$Color<- gsub(1, "DE", CPMLog$Color)
CPMLog$Color<- gsub(0, "Not DE", CPMLog$Color)
CPMLog$Color<-gsub(2, "Target", CPMLog$Color)
CPMLog$Color<-factor(CPMLog$Color, levels=c("DE", "Not DE", "Target"))
CPMLog<-CPMLog[order(CPMLog$target, decreasing=TRUE),]

#plot
p<-CPMPlot(CPMLog, "Mean")+
        theme(axis.ticks = element_line(colour = "black", size =2), 
              axis.ticks.length = unit(2,"mm"))+
        xlim(0,20)+
        ylim(0,20)+
        theme(axis.title.x = element_text(size=20),
              axis.title.y = element_text(size=20))
```



#10: Plot the number of species passing count thresholds for blocked and unblocked datasets (Figure 4B)

Read in pre-processed count matrices- only processing is collapsing isomirs together + downsampling
```{r}
writewd<-"C:/Users/Jenna/Documents/BlockerProject/"
blocked_counts<-read.csv( paste(writewd, "LargeCohort_RawData_blocked.csv", sep=""), row.names = "X")
unblocked_counts<-read.csv( paste(writewd, "LargeCohort_RawData_Unblocked.csv", sep=""), row.names = "X")

#Collapse counts-remove everything after 3rd "-" (-1, -5p, etc), then merge and sum together based on new miR IDs
blocked_counts$geneIDs<- sub("^([^-]*-[^-]*-[^-]*).*", "\\1", rownames(blocked_counts))
blocked_counts$geneIDs<- gsub("mir", "miR", blocked_counts$geneIDs)
blocked_counts_collapsed<-blocked_counts %>% group_by(geneIDs) %>% summarise_each(funs(sum)) %>% as.data.frame
rownames(blocked_counts_collapsed)<-blocked_counts_collapsed$geneIDs
blocked_counts_collapsed<- blocked_counts_collapsed[,2:ncol(blocked_counts_collapsed)]

unblocked_counts$geneIDs<- sub("^([^-]*-[^-]*-[^-]*).*", "\\1", rownames(unblocked_counts))
unblocked_counts$geneIDs<- gsub("mir", "miR", unblocked_counts$geneIDs)
unblocked_counts_collapsed<-unblocked_counts %>% group_by(geneIDs) %>% summarise_each(funs(sum)) %>% as.data.frame
rownames(unblocked_counts_collapsed)<-unblocked_counts_collapsed$geneIDs
unblocked_counts_collapsed<- unblocked_counts_collapsed[,2:ncol(unblocked_counts_collapsed)]

blockedCounts<-blocked_counts_collapsed
unblockedCounts<-unblocked_counts_collapsed
```

Get number of species passing thresholds, then plot
```{r}
#Set desired thresholds
thresholds<-seq(0,500,50)
NumberInEachGroup<- length(thresholds)*5 #used for plotting later

#Get counts passing threshold for unblocked samples, then transform data- 4 columns: ID, threshold, #of species that passed that threshold, and blocker status
        unblockedPassThreshold<- as.data.frame(apply(unblockedCounts, 2, 
                                                     CountsPassingThreshold, thresholds))
        rownames(unblockedPassThreshold)<- thresholds
        
        #add 4th column: mean of all samples
        unblockedPassThreshold$Mean<- rowMeans(unblockedPassThreshold)
        unblockedPassThreshold<- as.data.frame(t(unblockedPassThreshold))
        unblockedPassThreshold$ID<- rownames(unblockedPassThreshold)
        
        #reformat- melt
        unblockedPassThreshold<- melt(unblockedPassThreshold)
        colnames(unblockedPassThreshold)<- c("ID", "threshold","speciesPassingThreshold")
        unblockedPassThreshold$status<- rep("unblocked", length(thresholds) * (530+1))
        unblockedPassThreshold$threshold<- as.integer(as.character((unblockedPassThreshold$threshold)))


#Get counts passing threshold for blocked samples, then transform data- 4 columns: ID, threshold, #of species that passed that threshold, and blocker status
        blockedPassThreshold<- as.data.frame(apply(blockedCounts, 2, 
                                                     CountsPassingThreshold, thresholds))
        rownames(blockedPassThreshold)<- thresholds
        
        #add 4th column: mean of all samples
        blockedPassThreshold$Mean<- rowMeans(blockedPassThreshold)
        blockedPassThreshold<- as.data.frame(t(blockedPassThreshold))
        blockedPassThreshold$ID<- rownames(blockedPassThreshold)
        
        #reformat- melt
        blockedPassThreshold<- melt(blockedPassThreshold)
        colnames(blockedPassThreshold)<- c("ID", "threshold","speciesPassingThreshold")
        blockedPassThreshold$status<- rep("blocked", length(thresholds) * (535+1))
        blockedPassThreshold$threshold<- as.integer(as.character((blockedPassThreshold$threshold)))

#combine into one
passThreshold<-rbind(unblockedPassThreshold,blockedPassThreshold)
passThreshold$status<-factor(passThreshold$status, levels=c("unblocked", "blocked"))
test<-passThreshold[passThreshold$status=="unblocked",]

#Plot
p<-ggplot(passThreshold, aes(x=threshold, y=speciesPassingThreshold, color=status))+
        geom_point(position="jitter",stat="identity",size=2,alpha=0.2)+
        scale_color_manual(values=c("turquoise3","firebrick3"))+
        theme_classic()+
        scale_x_continuous(breaks=seq(0,500,50))+
        scale_y_continuous(breaks=seq(0,600,100))+
        xlab("Counts threshold")+
        ylab("Number of miRs passing counts threshold")+
        theme(axis.text.x = element_text(size=12,face="bold",color="black"),
              axis.text.y = element_text(size=12,face="bold",color="black"),
              axis.title.x = element_text(size=14,face="bold",color="black"),
              axis.title.y = element_text(size=14,face="bold",color="black"),
              legend.title=element_blank(),
              legend.text = element_text(size=14, face="bold"),
              legend.position=c(.85,.85),
              legend.background = element_rect(fill=NA, size=1, 
                                         linetype="solid", color="black"),
              panel.border = element_rect(colour = "black", fill=NA, size=2))
        


means<- passThreshold[passThreshold$ID=="Mean",]
```


 
```{r}
#Get sig DE genes- p<0.01 and baseMean>50
sig<-resOrdered[resOrdered$padj<0.01,]
sig<-sig[sig$baseMean>=50,]
sig<- sig[abs(sig$log2FoldChange)>=0.5,]

#Add in CPM data for these miRNAs- across all 16 libraries

#Convert to CPM
counts_cpm<- counts_collapsed/colSums(counts_collapsed) *1000000

Unblocked_RawCounts<- as.data.frame(rowMeans(counts_cpm[rownames(counts_cpm) %in% rownames(sig),colnames(counts_cpm) %in% colnames(unblocked_counts)]))
Blocked_RawCounts<- as.data.frame(rowMeans(counts_cpm[rownames(counts_cpm) %in% rownames(sig),colnames(counts_cpm) %in% colnames(blocked_counts)]))
OT<- merge(as.data.frame(sig[,c(1,2,6)]), Unblocked_RawCounts, by=0)
rownames(OT)<-OT$Row.names
colnames(OT)<- c("Species", "baseMean", "log2FoldChange", "padj", "MeanCPMUnblockedCounts")
OT<- merge(OT, Blocked_RawCounts, by=0)
OT<-OT[,-1]
colnames(OT)<- c("Species", "baseMean", "log2FoldChange", "padj", "MeanCPMUnblockedCounts", "MeanCPMBlockedCounts")
OT$MeanRawCounts<- rowMeans(OT[,5:6])
OT$Species<-gsub("hsa-", "", OT$Species)

#Annotate by whether species is a target miR
ActualTargets<- c('miR-486-5p','miR-92a-3p','miR-451a') #CL
ActualTargets<- c('miR-486','miR-92a','miR-451a') #BS
OT$Target<- OT$Species %in% ActualTargets

#Entire list of all species in target family
Target_92<- rownames(counts_collapsed[grep("miR-92", rownames(counts_collapsed)),])

Target_451<- rownames(counts_collapsed[grep("miR-451", rownames(counts_collapsed)),]) 

Target_486<- rownames(counts_collapsed[grep("miR-486", rownames(counts_collapsed)),]) 

Target_25<- rownames(counts_collapsed[grep("miR-25", rownames(counts_collapsed)),]) #CL or BS

TargetFamilies<- gsub("hsa-", "", c(Target_451,Target_486,Target_92, Target_25))
OT$TargetFamily<- OT$Species %in% TargetFamilies
OT$Color<- as.factor(rowSums(OT[,c(8,9)]))
OT$Color<- gsub(0, "Non-Target miR", OT$Color)
OT$Color<- gsub(1, "miR in Target Family", OT$Color)
OT$Color<- gsub(2, "Target miR", OT$Color)


#set which miRs to be labeled on graph
SubsetToLabel<- OT[OT$Color %in% c("Target miR", "miR in Target Family"),]
SubsetToLabel$Species<- gsub("hsa-", "", SubsetToLabel$Species)
```

```{r}
write.csv(OT, paste(wd,"OffTargetStats_BaseSpacePilot.csv", sep=""))
```

Plot off target effects- same as in figure 3A, using pilot data. This plot not used as figure.
```{r}
#plot
p<-ggplot(OT,aes(x=log2FoldChange, y= log2(padj+1e-100), fill=Color))+
  geom_point(aes(size=MeanRawCounts), pch=21, color="black")+
  scale_size_continuous(range = c(1, 50), guide=FALSE)+
  #scale_x_continuous(limits=c(-6,6),n.breaks = 7)+
  #scale_y_continuous(limits = c(-0.008, .05))+
  theme_classic()+
  xlab("log2 foldchange of differential expression")+
  ylab("Log2(Adjusted p value)")+
  geom_vline(xintercept=0,linetype="dotted", color= "black", size=1.25)+
  guides(colour = guide_legend(override.aes = list(size=5)))+
  geom_text_repel(data=SubsetToLabel, aes(x=log2FoldChange, y=padj, 
                                          label=Species),
                                          size=3.5, fontface= "bold",
                                          nudge_y = 0.0035, point.padding = 1.7, 
                                          box.padding = 0.5)+
  scale_fill_manual(values=c("orange","black", "red"))+
  theme(axis.text.y   = element_text(size=12, face= "bold", colour = "black"),
        axis.title.y  = element_text(size=16, face= "bold", colour = "black"),
        axis.title.x  = element_text(size=16, face="bold", colour = "black"),
        axis.text.x  = element_text(size=12, face="bold", colour = "black"),
        legend.background = element_rect(fill=NA, size=1, 
                                         linetype="solid", color="black"),
        legend.title=element_blank(),
        legend.key.size = unit(.4, "cm"),
        legend.spacing.y = unit(0, "mm"),
        legend.position = c(.8,.93),
        panel.border = element_rect(colour = "black", fill=NA, size=2),
        legend.text = element_text(size=13, face="bold"))

p


```

```{r}
sessionInfo()
```

