---
title: "R Notebook"
output: html_notebook
---

#Large cohort: differential expression of paired subset. Plotting heatmaps and Log2FC Blocked vs Unblocked.

Import libraries
```{r}
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(reshape2))
suppressPackageStartupMessages(library(purrr))
suppressPackageStartupMessages(library(tibble))
suppressPackageStartupMessages(library(metaseqR))
suppressPackageStartupMessages(library(DESeq2))
suppressPackageStartupMessages(library(pheatmap))
suppressPackageStartupMessages(library(ggrepel))
suppressPackageStartupMessages(library(ggpubr))
suppressPackageStartupMessages(library(gsubfn))
windowsFonts("Arial" = windowsFont("Arial"))
```


Functions
```{r}
#For plotting log2 CPM values nicely
CPMPlot<- function(CPM, Sample){
  #Get subset of CPM for labeling points- points that are both sig DE and are higher in unblocked than   blocked (beneath the slope line)
  CPM$HigherInUnblocked<- (CPM$Blocker-CPM$NoBlocker)<0
  SubsetToLabel<- CPM[CPM$HigherInUnblocked=="TRUE" &CPM$Color!="Not DE",]
  SubsetToLabel<-CPM[rownames(CPM) %in% c("hsa-miR-486", "hsa-miR-92a", "hsa-miR-451a", "hsa-miR-92b", "hsa-miR-25"),]
  
  #Plot
  p<- ggplot(CPM, aes(x=NoBlocker, y=Blocker, color=Color))+
  geom_point(size=3)+
  geom_abline(slope=1, color="red")+
  scale_colour_manual(values=c("red","black", "blue"))+
  theme_classic()+
  xlim(0,6)+
  ylim(0,6)+
  xlab(paste("Log of Unblocked Counts Per Million"))+
  ylab(paste("Log of Blocked Counts Per Million"))+
  geom_text_repel(data=SubsetToLabel, aes(x=NoBlocker, y=Blocker, label=rownames(SubsetToLabel)), color="black",  size=5, fontface="bold")+
  ggtitle(Sample)+
  theme(axis.text.y   = element_text(size=14, colour = "black", face="bold"),
        axis.title.y  = element_text(size=14, colour = "black", face="bold"),
        axis.title.x  = element_text(size=14, colour = "black", face="bold"),
        axis.text.x  = element_text(size=14, colour = "black", face="bold"),
        legend.background = element_rect(fill=NA, size=1, 
                                         linetype="solid", color="black"),
        legend.title=element_blank(),
        legend.key.size = unit(.4, "cm"),
        legend.spacing.y = unit(0, "mm"),
        legend.position = c(.15,.892),
        legend.text = element_text(size=18, face="bold"),
        panel.border = element_rect(colour = "black", fill=NA, size=2),
        plot.title=element_text(size=14, face="bold"))
  return(p)
  
}

#Plot PCA nicely
MyPCA<- function(dds, metadata, PC1Variance, PC2Variance,miny, maxy, minx, maxx) {
  #Normalized (variance stabilized transformation), then run PCA
  vsd <- varianceStabilizingTransformation(dds, blind=TRUE)
  PCA<-plotPCA(vsd, intgroup="Batch",returnData=TRUE)
  PCA$blocker<- metadata$blocker
  
  #plot
  p<- ggplot(PCA, aes(x=PC1, y=PC2, color=Batch, shape=blocker))+
  geom_point(size=4)+
  #scale_color_manual(values=Groupcolors)+
  scale_shape_manual(values=c(16,18))+
  theme_classic()+
  ylim(miny,maxy)+
  xlim(minx,maxx)+
  ylab(paste("PC2: ", PC2Variance, "% Variance", sep =""))+
  xlab(paste("PC1: ", PC1Variance, "% Variance", sep =""))+
  #geom_text(aes(label=Sample),hjust=.5, size= 5,vjust=-1.2, color="black")+
  theme(axis.text.y   = element_text(size=12, colour = "black", face="bold"),
        axis.title.y  = element_text(size=14, colour = "black", face="bold"),
        axis.title.x  = element_text(size=14, colour = "black", face="bold"),
        axis.text.x  = element_text(size=12, colour = "black", face="bold"),
        legend.background = element_rect(fill=NA, size=1, 
                                         linetype="solid", color="black"),
        legend.title=element_blank(),
        legend.text = element_text(size=14, face="bold"),
        legend.key.size = unit(.4, "cm"),
        legend.spacing.y = unit(0, "mm"),
        legend.position = c(.9,.85),
        panel.border = element_rect(colour = "black", fill=NA, size=2)
        )
  return(p)
}

```
#Overall Steps:

1: Read in paired count matrixes and subject IDs of paired samples, merge counts together and collapse mature/isomiRs. Filter based on CPM.
2. PCA, coloring based on batch- to determine if there are any large batch effects
2: Create DESeq2 object, run pair-wise DE on all species and plot heatmap
3: Create DESeq2 object, run pair-wise DE on NON-TARGET species (miR-486/miR-451a/miR-92a/miR-92b/miR-25 removed priot to DE)
4: Get off targets (From targets included DESeq2, step 2), plot. Export potential off-targets to csv.

```{r}
#read in paired data, blocked and unblocked
wd<-"~/BlockerProject/"
unblocked_paired<-read.csv(paste(wd, "LargeCohort_PairedSubset_89Samples_Unblocked.csv", sep=""),row.names="X") 
blocked_paired<-read.csv(paste(wd, "LargeCohort_PairedSubset_89Samples_Blocked.csv", sep=""),row.names="X") 

#Read in paired samples- From BH. Add column for new sampleID- "sampleIDppm_Blocked/Unblocked". Needed since some sample IDs are replicated between blocked/unblocked. 
pairedSamples<-read.csv(paste(wd, "PairedSamples_LargeCohort_FromBH.csv", sep=""), row.names = "UniqueID")
pairedSamples<-pairedSamples[,-1]

#merge counts together
counts_paired<- merge(unblocked_paired, blocked_paired, by=0)
rownames(counts_paired)<-counts_paired$Row.names
counts_paired<-counts_paired[,-1]
counts_paired<-counts_paired[,order(colnames(counts_paired))]

#Collapse counts together for species in same family- -3p/5p/-1/-2
counts_paired$geneIDs<- sub("^([^-]*-[^-]*-[^-]*).*", "\\1", rownames(counts_paired))
counts_paired$geneIDs<- gsub("mir", "miR", counts_paired$geneIDs)
counts_collapsed_paired<-counts_paired %>% group_by(geneIDs) %>% summarise_each(funs(sum)) %>% as.data.frame
rownames(counts_collapsed_paired)<-counts_collapsed_paired$geneIDs
counts_collapsed_paired<- counts_collapsed_paired[,2:ncol(counts_collapsed_paired)]

#Filter to remove any species without >1 CPM in >=75% of samples =="detected"
CPM<-(counts_collapsed_paired/colSums(counts_collapsed_paired))*1000000
CPM_Above1<- CPM[apply(CPM, 1, function(x) {sum(x>=1)})>(ncol(CPM)*.75),]
counts_CPMPass<-counts_collapsed_paired[rownames(counts_collapsed_paired) %in% rownames(CPM_Above1),]
```

#2. PCA, coloring based on batch- to determine if there are any large batch effects
```{r}
#Read in batch info, merge with other metadata- for PCA plotting
batches<-read.csv(paste(wd, "PairedSample_Batches.csv", sep=""))
rownames(batches)<-paste(batches$Paired.samples, batches$BlockerStatus,sep="_")
pairedSamples<-merge(batches, pairedSamples, by=0)
pairedSamples<-pairedSamples[,c(1,5,4,6)]

#create DESeq2 object- just based on samplegroup (blockerstatus)
#will add samples to model later (i.e., pairwise analysis)
pairedSamples$blocker<-as.factor(pairedSamples$blocker)
pairedSamples$Batch<-as.factor(pairedSamples$Batch) #remove
pairedSamples$phenoId<-as.factor(pairedSamples$phenoId)

dds<- DESeqDataSetFromMatrix(countData=counts_CPMPass, 
                       colData=pairedSamples, 
                       design=~blocker)

colors<-c("darkgoldenrod4", "darkgoldenrod2","darkcyan", "indianred4", "indianred1","lightblue4", "mediumpurple4",
          "grey77", "grey35","navy", "red4","darkolivegreen4", "orchid4")
p<-MyPCA(dds, pairedSamples, 29,13,-8,25,-20,25)+  scale_color_manual(values=colors)
```

Samples do not cluster by batch EXCEPT for batch 12 (blocked, miRNA FC 5)

What's different about batch 12? Read in QC information, compare metrics for batch 12 vs the other blocked batches
```{r}
#Read in QC info
qc<-read.csv("~/BlockerProject/QCInfo_PairedBlockedSamples.csv")

#Subset down to metrics of interest
qc_plot<-qc[,c(1,4,7,10,12,14)]

#Plot all
apply(qc_plot[,3:6], 2, function(x){ggplot(qc_plot, aes(x=Group, y=x))+geom_boxplot()+ylab(x)})

t.test(qc_plot[qc_plot$Group=="Non-outliers (Batches 8-11)",4],qc_plot[qc_plot$Group=="Outliers (Batch 12)",4])
t.test(qc_plot[qc_plot$Group=="Non-outliers (Batches 8-11)",5],qc_plot[qc_plot$Group=="Outliers (Batch 12)",5])

#Only one metric is significantly different- %genome. Speaks to extraction quality, possibily. Plot and export for use as supplemental figure
p<-ggplot(qc_plot, aes(x=Group, y=PercentGenome))+ 
  geom_boxplot(fill="grey")+
  theme_classic()+
  xlab("")+
  ylab("Percent Genomic DNA")+
  scale_y_continuous(breaks=seq(0,15,2), labels=paste(seq(0,15,2), "%", sep=""))+
  theme(axis.title.y=element_text(size=16, face="bold"),
        axis.text.y=element_text(size=12, face="bold", color="black"),
        axis.text.x=element_text(size=14, face="bold", angle=30, hjust=.95, color="black"),
        panel.border = element_rect(colour = "black", fill=NA, size=2))
  
  
```
#Remove outliers (paired samples in batch 12) AND the paired sample of those outliers before proceeding with DE
```{r}
#Get list of samples to remove
outliers<-qc[qc$Group=="Outliers (Batch 12)",]
outliers$Row.names<-paste(outliers$Paired.samples, outliers$BlockerStatus, sep="_")
outliers<-merge(outliers, pairedSamples, by="Row.names")
AllOutliers<-pairedSamples[pairedSamples$phenoId %in% outliers$phenoId,1]

#Remove outliers + their unblocked pair from count matrix
counts_OutliersRemoved<- counts_CPMPass[,!(colnames(counts_CPMPass) %in% AllOutliers)]
```

#3: Create DESeq2 object, run pair-wise DE on all species

create metadata- need to remove all the outlier samples
```{r}
#Read in paired samples- From BH. Add column for new sampleID- "sampleIDppm_Blocked/Unblocked". Needed since some sample IDs are replicated between blocked/unblocked. 
pairedSamples<-read.csv(paste(wd, "PairedSamples_LargeCohort_FromBH.csv", sep=""), row.names = "UniqueID")
pairedSamples<-pairedSamples[,-1]

pairedSamples_OutliersRemoved<-pairedSamples[!(pairedSamples$phenoId %in% outliers$phenoId),]
pairedSamples_OutliersRemoved$blocker<-as.factor(pairedSamples_OutliersRemoved$blocker)
pairedSamples_OutliersRemoved$phenoId<-as.factor(pairedSamples_OutliersRemoved$phenoId)


```


```{r}
#create DESeq2 object- just based on samplegroup (blockerstatus)
#will add samples to model later (i.e., pairwise analysis)
dds<- DESeqDataSetFromMatrix(countData=counts_OutliersRemoved, 
                       colData=pairedSamples_OutliersRemoved, 
                       design=~blocker)

dds$blocker<- relevel(dds$blocker, ref = "Unblocked")

#Create model that takes into account sample differences
ddsMF<- dds
design(ddsMF)<- formula(~phenoId + blocker)

#run DESeq2
dds<- DESeq(ddsMF)

#extract and order results  
res_paired<-na.omit(results(dds))
resOrdered_paired<-res_paired[order(res_paired$padj),]

nrow(resOrdered_paired[resOrdered_paired$padj<0.05,]) #270 (385 if unpaired, 170 if paired+CPM 75% filter, 168 if outliers removed)
nrow(resOrdered_paired[resOrdered_paired$padj<0.01,]) #244 (295 if unpaired, 154 if paired+CPM 75% filter, 165 if outliers removed)
nrow(resOrdered_paired[resOrdered_paired$padj<0.01 & abs(resOrdered_paired$log2FoldChange)>1,]) #138 (206 if unpaired, 69 if paired+CPM 75% filter, 74 if outliers removed)
```

Export results
```{r}
write.csv(resOrdered_paired, paste(wd, "PairedCohort_DESeq2Results_Collapsed.csv",sep=""))
```

Plot heatmap
```{r}
rld<-varianceStabilizingTransformation(dds)

MinSubtract=-3
MaxSubtract=2
nspecies=50
fontsize=12 

#Select top sig DE miRNAs
topgenes<- rownames(resOrdered_paired[1:nspecies,])
mat<- assay(rld)[topgenes,]
mat<- mat - rowMeans(mat)

#Heatmap settings
mat_breaks<- seq(min(mat-MinSubtract), max(mat-MaxSubtract), length.out=75)

metadata<-pairedSamples_OutliersRemoved[,c(1,2)]
#rownames(metadata)<-metadata$Row.names

  #plot heatmap
  p<-pheatmap(mat, breaks = mat_breaks, 
            color =colorRampPalette( c("red", "black", "green"), space="rgb")(100),
            show_rownames = TRUE, show_colnames=FALSE,
            annotation = metadata,
            
            fontsize_row = fontsize,fontsize=12,treeheight_row=0,
            border_color= NA)
```



#Rerun pairwise DESeq2, removing 3 targets from analysis + removing 92b and 25 
Using count matrix that has been filtered to remove lowly expressed species (<50% of samples have >1 CPM)

```{r}
#Subset counts matrix, removing 3 targets
counts_collapsed_paired_RAT<- counts_OutliersRemoved[!(rownames(counts_OutliersRemoved) %in% c("hsa-miR-486",
                                                                                "hsa-miR-451a",
                                                                                "hsa-miR-92a",
                                                                                "hsa-miR-92b",
                                                                                "hsa-miR-25")),]

 
#create DESeq2 object- just based on samplegroup (blockerstatus)
#will add samples to model later (i.e., pairwise analysis)
dds<- DESeqDataSetFromMatrix(countData=counts_collapsed_paired_RAT, 
                       colData=pairedSamples_OutliersRemoved, 
                       design=~blocker)
dds$blocker<- relevel(dds$blocker, ref = "Unblocked")

#Create model that takes into account sample differences
ddsMF<- dds
design(ddsMF)<- formula(~phenoId + blocker)

#run DESeq2
dds<- DESeq(ddsMF)

#extract and order results  
res_paired_RAT<-na.omit(results(dds))
resOrdered_paired_RAT<-res_paired_RAT[order(res_paired_RAT$padj),]


nrow(resOrdered_paired_RAT[resOrdered_paired_RAT$padj<0.05,]) #275 ( 216 if CPM filtered) 
nrow(resOrdered_paired_RAT[resOrdered_paired_RAT$padj<0.01,]) #243 ( 206 if CPM filtered)
nrow(resOrdered_paired_RAT[resOrdered_paired_RAT$padj<0.01 & abs(resOrdered_paired_RAT$log2FoldChange)>1,]) #142 ( 106 if CPM filtered)
```

Export results
```{r}
write.csv(resOrdered_paired_RAT, paste(wd, "PairedCohort_DESeq2Results_ActualTargetsRemoved.csv",sep=""))
```


Plot heatmap
```{r}
#rld<-varianceStabilizingTransformation(dds)

MinSubtract=0
MaxSubtract=2
nspecies=100
  #Select top sig DE miRNAs
  topgenes<- rownames(resOrdered_paired_RAT[1:nspecies,])

  mat<- assay(rld)[topgenes,]
  mat<- mat - rowMeans(mat)
  
  #Heatmap settings
  mat_breaks<- seq(min(mat-MinSubtract), max(mat-MaxSubtract), length.out=75)

  metadata<-pairedSamples_OutliersRemoved

  #plot heatmap
  p<-pheatmap(mat, breaks = mat_breaks, 
            color =colorRampPalette( c("red", "black", "green"), space="rgb")(100),
            show_rownames = TRUE, show_colnames=FALSE,
            annotation = metadata,
            
            fontsize_row = fontsize,fontsize=12,treeheight_row=0,
            border_color= NA)
```


#4: Get off targets (From targets included DESeq2, step 2), plot. Export potential off-targets to csv.
 
```{r}
#Get raw counts- only downsampled + isomirs collapsed. Separate into blocked/unblocked
counts<-counts_OutliersRemoved
Unblocked_counts_collapsed_paired<-counts[,gsub(".*_", "", colnames(counts))=="Unblocked"]
Blocked_counts_collapsed_paired<-counts[,gsub(".*_", "", colnames(counts))=="Blocked"]

#Get sig DE genes- p<0.01 and baseMean>50
sig<-resOrdered_paired[resOrdered_paired$padj<0.01,]
sig<-sig[sig$baseMean>=50,]
sig<- sig[abs(sig$log2FoldChange)>=0,] #53, 101 if set at 0.5

#Add in CPM data for these miRNAs
        #Convert to CPM
        counts_cpm<- (counts/colSums(counts)) *1000000
        
        #Merge with DE data, reformat
        OT<- merge(as.data.frame(sig[,c(1,2,6)]), rowMeans(counts_cpm), by=0)
        colnames(OT)<- c("Species", "baseMean", "log2FoldChange", "padj", "MeanCPM")
        OT$Species<- gsub("hsa-", "", OT$Species)


#Annotate by whether species is a target miR
ActualTargets<- c('miR-486-5p','miR-92a-3p','miR-451a') #CL
ActualTargets<- c('miR-486','miR-92a','miR-451a') #BS
OT$Target<- OT$Species %in% ActualTargets

#Entire list of all species in target family
Target_92<- rownames(counts[grep("miR-92", rownames(counts)),])
Target_451<- rownames(counts[grep("miR-451", rownames(counts)),]) 
Target_486<- rownames(counts[grep("miR-486", rownames(counts)),]) 
Target_25<- rownames(counts[grep("miR-25", rownames(counts)),]) #CL or BS
TargetFamilies<- gsub("hsa-", "", c(Target_451,Target_486,Target_92, Target_25))

#Add family info in
OT$TargetFamily<- OT$Species %in% TargetFamilies
OT$Color<- as.factor(rowSums(OT[,c(6,7)]))
OT$Color<- gsub(0, "Non-Target miR", OT$Color)
OT$Color<- gsub(1, "miR in Target Family", OT$Color)
OT$Color<- gsub(2, "Target miR", OT$Color)


#set which miRs to be labeled on graph
SubsetToLabel<- OT[OT$Color %in% c("Target miR", "miR in Target Family"),]
SubsetToLabel$Species<- gsub("hsa-", "", SubsetToLabel$Species)
```

Write results to csv
```{r}
write.csv(OT, paste(wd,"OffTargetStats_PairedCohort_OutliersRemoved_PairwiseDESeq2_TargetsIncluded_75CPMFilter.csv", sep=""))
```

Plot off target effects- same as in figure 3A
```{r}
#plot
p<-ggplot(OT,aes(x=log2FoldChange, y= log2(padj+1e-100), fill=Color))+
  geom_point(aes(size=MeanCPM), pch=21, color="black")+
  scale_size_continuous(range = c(1, 50),guide=FALSE )+
  ylim(-350,50)+
  xlim(-6, 3)+
  theme_classic()+
  xlab("log2 foldchange of differential expression")+
  ylab("Log2(Adjusted p value)")+
  geom_vline(xintercept=0,linetype="dotted", color= "black", size=1.25)+
  guides(colour = guide_legend(override.aes = list(size=5)))+
  geom_text_repel(data=SubsetToLabel, aes(x=log2FoldChange, y=log2(padj+1e-100), 
                                          label=Species),
                                          size=5, fontface= "bold",
                                          nudge_y = 0.0035, point.padding = 1.7, 
                                          box.padding = 0.5)+
  scale_fill_manual(values=c("orange","black", "red"))+
  theme(axis.text.y   = element_text(size=15, face= "bold", colour = "black"),
        axis.title.y  = element_text(size=18, face= "bold", colour = "black"),
        axis.title.x  = element_text(size=18, face="bold", colour = "black"),
        axis.text.x  = element_text(size=15, face="bold", colour = "black"),
        legend.background = element_rect(fill=NA, size=1, 
                                         linetype="solid", color="black"),
        legend.title=element_blank(),
        legend.key.size = unit(.4, "cm"),
        legend.spacing.y = unit(0, "mm"),
        legend.position = c(.8,.93),
        panel.border = element_rect(colour = "black", fill=NA, size=2),
        legend.text = element_text(size=13, face="bold"))

p

```



#5: Plot log2 CPM values for blocked and unblocked datasets, coloring based on sig (from pairwise targets included analysis)
```{r}
#Convert to CPM
CPM<- counts_OutliersRemoved/colSums(counts_OutliersRemoved) *1000000

#Get mean CPM for blocked and unblocked for every species
blockedSamples<-pairedSamples[pairedSamples$blocker=="Blocked",]
blocked_CPM<-CPM[,colnames(CPM) %in% rownames(blockedSamples)]

unblockedSamples<-pairedSamples[pairedSamples$blocker=="Unblocked",]
unblocked_CPM<-CPM[,colnames(CPM) %in% rownames(unblockedSamples)]

CPM_Mean<-data.frame(Blocker=rowMeans(blocked_CPM), NoBlocker=rowMeans(unblocked_CPM))

#log2 of CPM
CPMLog<- log2(CPM_Mean+1) #+1 optional here

#Get miRNA info. Need: whether or not it was targeted + whether or not it was DE
#Add in target info- add column for whether or not the species was targeted
ActualTargets<- c('hsa-miR-486','hsa-miR-92a','hsa-miR-451a')
CPMLog$target<- rownames(CPMLog) %in% ActualTargets

#Add in DE info- using DESeq2 results from targets included analysis. Import data here.
CPMLog$Sig<- rownames(CPMLog) %in% rownames(sig)

#Add in third column, used to color points in graph. Add together target/sig columns. If it's a target and sig, it'll be 2, if it's just sig it'll be 1, if it's neither it'll be 0.
CPMLog$Color<- as.factor(rowSums(CPMLog[,3:4]))
CPMLog$Color<- gsub(1, "DE", CPMLog$Color)
CPMLog$Color<- gsub(0, "Not DE", CPMLog$Color)
CPMLog$Color<-gsub(2, "Target", CPMLog$Color)
CPMLog$Color<-factor(CPMLog$Color, levels=c("DE", "Not DE", "Target"))
CPMLog<-CPMLog[order(CPMLog$target, decreasing=TRUE),]

#plot
p<-CPMPlot(CPMLog, "Mean")+
        theme(axis.ticks = element_line(colour = "black", size =2), 
              axis.ticks.length = unit(2,"mm"))+
        xlim(0,20)+
        ylim(0,20)+
        theme(axis.title.x = element_text(size=20),
              axis.title.y = element_text(size=20))


```