---
title: "R Notebook"
output: html_notebook
---
Import libraries
```{r}
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(reshape2))
suppressPackageStartupMessages(library(purrr))
suppressPackageStartupMessages(library(tibble))
suppressPackageStartupMessages(library(metaseqR))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(DESeq2))
suppressPackageStartupMessages(library(pheatmap))
suppressPackageStartupMessages(library(ggrepel))
```
Functions
```{r}
PlotHeatmap<- function(OrderedResults, rld, metadata, nspecies, MinSubtract, MaxSubtract, fontsize){
  #Select top sig DE miRNAs
  topgenes<- rownames(OrderedResults[1:nspecies,])
  mat<- assay(rld)[topgenes,]
  mat<- mat - rowMeans(mat)
  colnames(mat)<-paste(metadata$Patient," (",metadata$sampleGroup, ")",sep="")
  
  #Heatmap settings
  mat_breaks<- seq(min(mat-MinSubtract), max(mat-MaxSubtract), length.out=75)

  #Create metadata- to be used to color code to show which group the sample belongs to
  metadata<- data.frame(Group=metadata$sampleGroup, 
                        row.names=paste(metadata$Patient," (",metadata$sampleGroup, ")",sep=""),
                        Sample=metadata$Patient)

  #plot heatmap
  p<-pheatmap(mat, breaks = mat_breaks, 
            color =colorRampPalette( c("red", "black", "green"), space="rgb")(100),
            show_rownames = TRUE, show_colnames=FALSE,
            annotation = metadata,
            annotation_colors= list(
              Group=c(Blocked="firebrick3", Unblocked="turquoise3"),
                      Sample=c(A="goldenrod1", 
                               B="violetred3", 
                               C="skyblue3", 
                               D="darkolivegreen4")),
            fontsize_row = fontsize,fontsize=12,treeheight_row=0,
            border_color= NA)
  return(p)
}

CPMPlot<- function(CPM, Sample){
  #Get subset of CPM for labeling points- points that are both sig DE and are higher in unblocked than   blocked (beneath the slope line)
  CPM$HigherInUnblocked<- (CPM$Blocker-CPM$NoBlocker)<0
  SubsetToLabel<- CPM[CPM$HigherInUnblocked=="TRUE" &CPM$Color!="Not DE",]
  
  #Plot
  p<- ggplot(CPM, aes(x=NoBlocker, y=Blocker, color=Color))+
  geom_point(size=3)+
  geom_abline(slope=1, color="red")+
  scale_colour_manual(values=c("red","black", "blue"))+
  theme_classic()+
  xlim(0,6)+
  ylim(0,6)+
  xlab(paste("Log of Unblocked Counts Per Million"))+
  ylab(paste("Log of Blocked Counts Per Million"))+
  geom_text_repel(data=SubsetToLabel, aes(x=NoBlocker, y=Blocker, label=rownames(SubsetToLabel)), color="black",  size=5, fontface="bold")+
  ggtitle(Sample)+
  theme(axis.text.y   = element_text(size=14, colour = "black", face="bold"),
        axis.title.y  = element_text(size=14, colour = "black", face="bold"),
        axis.title.x  = element_text(size=14, colour = "black", face="bold"),
        axis.text.x  = element_text(size=14, colour = "black", face="bold"),
        legend.background = element_rect(fill=NA, size=1, 
                                         linetype="solid", color="black"),
        legend.title=element_blank(),
        legend.key.size = unit(.4, "cm"),
        legend.spacing.y = unit(0, "mm"),
        legend.position = c(.15,.892),
        legend.text = element_text(size=18, face="bold"),
        panel.border = element_rect(colour = "black", fill=NA, size=2),
        plot.title=element_text(size=14, face="bold"))
  return(p)
  
}
```

#Read in unblocked data
```{r}
unblocked_readwd<-"Z:/Small_RNA/BaseSpace/"

#Get filenames needed
filenames1<-c("Hersh_PT8_Rerun_Small_RNA_11072017_counts",
             "Hersh-miRNA-PT9Reruns_Small_RNA_12182017_counts")

filenames2<-c("9256704449-2_Small_RNA_05252017_counts",
             "9416449660-3_Small_RNA_06282017_counts",
             "9597590455-6_Small_RNA_06292017_counts",
             "9959254099-1_Small_RNA_10062017_counts",
             "10272166057-5_Small_RNA_08212017_counts",
             "10284133418-9_Small_RNA_08212017_counts")

filenames3<- c("AERD_miRNA_Sequencing_Small_RNA_10062017_counts",
               "Kelan_Viva_miRNA_Small_RNA_09272017_counts"
        
)

filenames4<-c("Hersh_PT8_Rerun_Small_RNA_11072017_counts",
             "Hersh-miRNA-PT9Reruns_Small_RNA_12182017_counts",
             "9256704449-2_Small_RNA_05252017_counts",
             "9416449660-3_Small_RNA_06282017_counts",
             "9597590455-6_Small_RNA_06292017_counts",
             "10272166057-5_Small_RNA_08212017_counts",
             "10284133418-9_Small_RNA_08212017_counts"
            )

filenames<-filenames4
#Read in files
unblocked_input<-list()
for (i in 1:length(filenames)){
        df<-read.csv(paste(unblocked_readwd, filenames[i], ".csv", sep=""), stringsAsFactors = FALSE)
        df<-df[-1,]
        unblocked_input[[i]]<-df
}


```

#Can skip this: reading in pilot data as comparison
```{r}
wd<- "//Cifs2/rcdata$/Channing_miRNA_Blocker_Test/"
BaseSpaceDownSampled_Counts<- read.csv(paste(wd, "OtherResults/DownSampledCounts_BaseSpaceMatureIsomir_metaseqR_121119.csv", sep =""))
rownames(BaseSpaceDownSampled_Counts)<- BaseSpaceDownSampled_Counts$X
BaseSpaceDownSampled_Counts<- BaseSpaceDownSampled_Counts[,-1]

#remove "-1" "-5p" etc from rownames

BaseSpaceDownSampled_Counts$X<-sub("^([^-]*-[^-]*-[^-]*).*", "\\1", rownames(BaseSpaceDownSampled_Counts))

unblocked_all_input<-BaseSpaceDownSampled_Counts[,c(17,1:8)]
blocked_all_input<-BaseSpaceDownSampled_Counts[,c(17,9:16)]
```


#Merge into one dataset, filter based on total reads per sample and per miRNA
```{r}
#Merge into one dataset
unblocked_all_input<- unblocked_input %>% purrr::reduce(full_join, by = "X")

#Convert to integer, remove NA
Rownames<-unblocked_all_input$X
rownames(unblocked_all_input)<-Rownames
unblocked_all<-unblocked_all_input[,-1]
unblocked_all<- as.data.frame(apply(unblocked_all, 2, as.integer))
rownames(unblocked_all)<-Rownames
unblocked_all<-na.omit(unblocked_all)

#Remove any samples with fewer than 2million total reads
unblocked_RemoveSample<- unblocked_all[,colSums(unblocked_all)>2000000]

#Downsample to common depth
unblocked_DS<- downsample.counts(unblocked_RemoveSample, set.seed(42)) 
#unblocked_DS$X<-Rownames
unblocked_DS$X<-rownames(unblocked_DS)

#sort and remove sequences from rownames, and convert to integer
unblocked_sort<-  unblocked_DS[order(unblocked_DS[,1], decreasing = TRUE),]
unblocked_sort$X<-rownames(unblocked_sort) 
unblocked_sort$X<- gsub(".*_", "", unblocked_sort$X)

#Collapse counts- multiple entries for a single species are collapsed into one
unblocked_collapse<- unblocked_sort %>% as.tibble() %>% group_by(X) %>% summarise_all(funs(if(is.numeric(.)) sum(.) else "Total")) %>% as.data.frame
rownames(unblocked_collapse)<-unblocked_collapse$X
unblocked_collapse<-unblocked_collapse[,-1]

#write to file for later use
writewd<-"C:/Users/Jenna/Documents/BlockerProject/"
write.csv(unblocked_collapse, paste(writewd, "LargeCohort_RawData_Unblocked.csv", sep=""))
```


#Further filtering + plotting % map to targets
```{r}
#read in counts- pre-processing in above chunk
unblocked_collapse<-read.csv(paste(writewd, "LargeCohort_RawData_Unblocked.csv"), row.names = "X")

#Remove any species with < 10 total counts-optional
unblocked_RemoveLow<- unblocked_collapse[rowSums(unblocked_collapse) >10,]
unblocked_RemoveLow<-unblocked_collapse

#Convert to CPM- optional
unblocked_CPM<-unblocked_RemoveLow/colSums(unblocked_RemoveLow) *1000000
unblocked_CPM<-unblocked_RemoveLow

#Get total number of CPM for 3 targets-will be multiple lines, grep for targets.Sum together target CPMs
AllTargets<-c("hsa-miR-486", "hsa-miR-92", "hsa-miR-451")

TargetCPM_1<-lapply(AllTargets, function(x) {unblocked_CPM[grep(x, rownames(unblocked_CPM), ignore.case = TRUE),]}) #get targets
TargetCPM_2<-lapply(TargetCPM_1, function(x) {x[!(rownames(x) %in% c("hsa-miR-4511", "hsa-miR-4510","hsa-mir-4511", "hsa-mir-4510")),]}) #remove miR-4511/4510
TargetCPM_3<- lapply( TargetCPM_2, function(x) {colSums(x[,1:ncol(x)])}) #merge multiple species in a family into one
TargetCPM<- as.data.frame(do.call(rbind, TargetCPM_3))
rownames(TargetCPM)<-AllTargets
TargetCPM<- TargetCPM[,order(colnames(TargetCPM))]


#Get percentage of total CPMs that 3 targets make up
unblocked_CPM<-unblocked_CPM[,order(colnames(unblocked_CPM))]
unblocked_TargetPercentageCPM<-TargetCPM/colSums(unblocked_CPM) *100

#Optional: add row with percent mapping to non-target
unblocked_TargetPercentageCPM["AllOthers" ,] <- 100-colSums(unblocked_TargetPercentageCPM)

```


#Read in blocked data
```{r}
blocked_readwd<-"Z:/COPD/miRNAseq/BaseSpace/"

#Get filenames needed
filenames<-c("21801069331-0_GECOPD_miRNASeq _Small_RNA_11222019_counts",
             "21801069331-0_GECOPD_miRNASeq _Small_RNA_12032019_counts",
             "21801069331-0_GECOPD_miRNASeq _Small_RNA_12042019_counts",
             "21801069331-0_GECOPD_miRNASeq _Small_RNA_12102019_counts",
             "21801069331-0_GECOPD_miRNASeq _Small_RNA_12112019_counts",
             "24200081892-6_GECOPD_P6_miRNASeq _Small_RNA_03092020_counts")

#Read in files
blocked_input<-list()
for (i in 1:length(filenames)){
        df<-read.csv(paste(blocked_readwd, filenames[i], ".csv", sep=""), stringsAsFactors = FALSE)
        df<-df[-1,]
        blocked_input[[i]]<-df
}

```

#Merge into one dataset, filter based on total reads per sample and per miRNA
```{r}
#Merge into one dataset
blocked_all_input<- blocked_input %>% purrr::reduce(full_join, by = "X")

#Convert to integer, remove NA
Rownames<-blocked_all_input$X
rownames(blocked_all_input)<-Rownames
blocked_all<-blocked_all_input[,-1]
blocked_all<- as.data.frame(apply(blocked_all, 2, as.integer))
rownames(blocked_all)<-Rownames
blocked_all<-na.omit(blocked_all)


#Remove any samples with fewer than 2million total reads
blocked_RemoveSample<- blocked_all[,colSums(blocked_all)>2000000]

#Downsample to common depth
blocked_DS<- downsample.counts(blocked_RemoveSample, set.seed(42)) 
#blocked_DS$X<-Rownames
blocked_DS$X<-rownames(blocked_DS)

#sort and remove sequences from rownames, and convert to integer
blocked_sort<-  blocked_DS[order(blocked_DS[,1], decreasing = TRUE),]
blocked_sort$X<-rownames(blocked_sort)
blocked_sort$X<- gsub(".*_", "", blocked_sort$X)


#Collapse counts- multiple entries for a single species are collapsed into one
blocked_collapse<- blocked_sort %>% as.tibble() %>% group_by(X) %>% summarise_all(funs(if(is.numeric(.)) sum(.) else "Total")) %>% as.data.frame
rownames(blocked_collapse)<-blocked_collapse$X
blocked_collapse<-blocked_collapse[,-1]

#write to csv for later use
write.csv(blocked_collapse, paste(writewd, "LargeCohort_RawData_blocked.csv", sep=""))
```

#Further filtering + plotting % map to targets
```{r}
#read in counts- pre-processing in above chunk
blocked_collapse<-read.csv(paste(writewd, "LargeCohort_RawData_Unblocked.csv"), row.names = "X")

#Remove any species with < 10 total counts  
blocked_RemoveLow<- blocked_collapse[rowSums(blocked_collapse) >10,]
blocked_RemoveLow<-blocked_collapse

#Convert to CPM- optional
#blocked_CPM<-blocked_RemoveLow/colSums(blocked_RemoveLow) *1000000
blocked_CPM<-blocked_RemoveLow


#Get total number of CPM for 3 targets-will be multiple lines, grep for targets.Sum together target CPMs
AllTargets<-c("hsa-miR-486", "hsa-miR-92", "hsa-miR-451")
TargetCPM_1<-lapply(AllTargets, function(x) {blocked_CPM[grep(x, rownames(blocked_CPM), ignore.case=TRUE),]}) #get targets
TargetCPM_2<-lapply(TargetCPM_1, function(x) {x[!(rownames(x) %in% c("hsa-miR-4511", "hsa-miR-4510","hsa-mir-4511", "hsa-mir-4510")),]}) #remove miR-4511/4510
TargetCPM_3<- lapply( TargetCPM_2, function(x) {colSums(x[,1:ncol(x)])}) #merge multiple species in a family into one
TargetCPM<- as.data.frame(do.call(rbind, TargetCPM_3))
rownames(TargetCPM)<-AllTargets
TargetCPM<- TargetCPM[,order(colnames(TargetCPM))]


#Get percentage of total CPMs that 3 targets make up
blocked_CPM<-blocked_CPM[,order(colnames(blocked_CPM))]
blocked_TargetPercentageCPM<-TargetCPM/colSums(blocked_CPM) *100

#Optional: add row with percent mapping to non-target
blocked_TargetPercentageCPM["AllOthers" ,] <- 100-colSums(blocked_TargetPercentageCPM)
```


#Plot- 530 and 537 total samples
```{r}
#Add column for blocked/unblocked to "TargetPercentageCPM" dataframes
unblocked_TargetPercentageCPM$BlockerStatus<-rep("Unblocked", nrow(unblocked_TargetPercentageCPM))
blocked_TargetPercentageCPM$BlockerStatus<-rep("Blocked", nrow(blocked_TargetPercentageCPM))

#reformat for plotting
df_unblocked<-unblocked_TargetPercentageCPM
df_unblocked$Species<- rownames(df_unblocked)
df_unblocked_melt<- melt(df_unblocked, by="Species")

df_blocked<-blocked_TargetPercentageCPM
df_blocked$Species<- rownames(df_blocked)
df_blocked_melt<- melt(df_blocked, by="Species")

#Add column for ordering
df_unblocked_melt$Order<- rep(1, nrow(df_unblocked_melt))
df_blocked_melt$Order<- rep(2, nrow(df_blocked_melt))

#rbind unblocked and blocked dataframes together
bind_TargetPercentageCPM<- rbind(df_unblocked_melt, df_blocked_melt)
bind_TargetPercentageCPM$Order<- rep(c(1,2,3,4), nrow(bind_TargetPercentageCPM)/4)
bind_TargetPercentageCPM$BlockerStatus<-factor(bind_TargetPercentageCPM$BlockerStatus, levels=c("Unblocked", "Blocked"))




p<-ggplot(bind_TargetPercentageCPM, aes(x = reorder(Species, Order), y =value, fill=BlockerStatus)) +
        geom_boxplot()+
        ylab("Percent of total reads mapping")+
        xlab("")+
        theme_classic()+
        scale_fill_manual("", values=c("Blocked"="firebrick3", "Unblocked"="turquoise3"))+
        theme(axis.title.y = element_text(face="bold", size=14),
              axis.text.y = element_text(face="bold", size=12, color="black"),
              axis.text.x = element_text(face="bold", size=12, color="black"),
              legend.text = element_text(face="bold", size=10),
              legend.title=element_blank(),
              panel.border = element_rect(colour = "black", fill=NA, size=2),
              strip.background = element_rect(color = "black", size = 2),
              legend.position=c(.72,.9),
              legend.background = element_rect(fill=NA, size=1, linetype="solid", color="black"))


```

```{r}
#Tiny t test- blocked and unblocked dataset 4
miR486<- list(df_blocked_melt[grep("hsa-miR-486", df_blocked_melt$Species),c(1,2,4)],
               df_unblocked_melt[grep("hsa-miR-486", df_unblocked_melt$Species),c(1,2,4)])
t.test(miR486[[1]]$value, miR486[[2]]$value) #p<2.2e-16 #means: 19.19062  69.61180 

miR451<- list(df_blocked_melt[grep("hsa-miR-451", df_blocked_melt$Species),c(1,2,4)],
               df_unblocked_melt[grep("hsa-miR-451", df_unblocked_melt$Species),c(1,2,4)])
t.test(miR451[[1]]$value, miR451[[2]]$value) #p<2.2e-16 #means:  4.940708  1.462336

miR92<- list(df_blocked_melt[grep("hsa-miR-92", df_blocked_melt$Species),c(1,2,4)],
               df_unblocked_melt[grep("hsa-miR-92", df_unblocked_melt$Species),c(1,2,4)])
t.test(miR92[[1]]$value, miR92[[2]]$value) #p<2.2e-16  #means: 9.03835  22.05245 

AllOthers<- list(df_blocked_melt[grep("AllOthers", df_blocked_melt$Species),c(1,2,4)],
               df_unblocked_melt[grep("AllOthers", df_unblocked_melt$Species),c(1,2,4)])
t.test(AllOthers[[1]]$value, AllOthers[[2]]$value) #p<2.2e-16  #means: 66.830321  6.873419
```


#Differential expression- unpaired, DESeq2

Filter, merge and export count matrices
```{r}
#Remove genes without > 1 CPM in 50% of unblocked/blocked samples- in CPM data
blocked_CPMPass<-blocked_CPM[rowSums(blocked_CPM>1)>(ncol(blocked_CPM)/2),]
unblocked_CPMPass<-unblocked_CPM[rowSums(unblocked_CPM>1)>(ncol(unblocked_CPM)/2),]

#For the genes that passsed the CPM threshold, get raw (downsampled) counts
blocked_CountsPass<-blocked_RemoveLow[rownames(blocked_RemoveLow) %in% rownames(blocked_CPMPass),]
unblocked_CountsPass<-unblocked_RemoveLow[rownames(unblocked_RemoveLow) %in% rownames(unblocked_CPMPass),]

#merge blocked and unblocked, keeping all rows
blocked_CountsPass$geneID<- rownames(blocked_CountsPass)
unblocked_CountsPass$geneID<-rownames(unblocked_CountsPass)
counts<- merge(blocked_CountsPass, unblocked_CountsPass, all=TRUE,by="geneID" )
counts[is.na(counts)]<-0

#Export
write.csv(counts, paste(blocked_readwd, "LargeCohort_AllSamples_RawCounts_CPMFilter.csv", sep=""))

#Create metadata, export
meta_blocked<-data.frame(Samples=colnames(blocked_CPMPass), Group="Blocked")
meta_unblocked<-data.frame(Samples=colnames(unblocked_CPMPass), Group="Unblocked")
meta<-rbind(meta_blocked,meta_unblocked)
write.csv(meta, paste(blocked_readwd,"LargeCohort_Metadata.csv",sep=""))
```


Perform differential expression between two groups
```{r}
#Import raw counts
readwd<-"C:/Users/Jenna/Documents/BlockerProject/"
countsInput<-read.csv(paste(readwd, "LargeCohort_AllSamples_RawCounts_CPMFilter.csv", sep="" ))
counts<-countsInput
rownames(counts)<-counts$geneID
counts<-counts[,3:ncol(counts)]

#Collapse counts-remove everything after 3rd "-" (-1, -5p, etc), then merge and sum together based on new miR IDs
counts$geneIDs<- sub("^([^-]*-[^-]*-[^-]*).*", "\\1", rownames(counts))
counts$geneIDs<- gsub("mir", "miR", counts$geneIDs)
counts_collapsed<-counts %>% group_by(geneIDs) %>% summarise_each(funs(sum)) %>% as.data.frame
rownames(counts_collapsed)<-counts_collapsed$geneIDs
counts_collapsed<- counts_collapsed[,2:ncol(counts_collapsed)]

#Import metadata
meta<- read.csv(paste(readwd,"LargeCohort_Metadata.csv",sep=""))

#Run DESeq2
dds<-DESeqDataSetFromMatrix(counts_collapsed, meta, design=~Group)
dds<-DESeq(dds)

#Extract and order results
res<-na.omit(results(dds))
resOrdered<-res[order(res$padj),]
nrow(resOrdered[resOrdered$padj<0.01,]) #385
nrow(resOrdered[resOrdered$padj<0.05,]) #392
sig<-resOrdered[resOrdered$padj<0.01,]
sig<- sig[abs(sig$log2FoldChange)>1,] #281

#Export DE results
write.csv(resOrdered[resOrdered$padj<0.05,], paste(readwd, "LargeCohort_DESeq2Results_Collapsed_log2FC1.csv", sep=""))

```

```{r}
#Plot heatmap of DESeq2 results
rld<-varianceStabilizingTransformation(dds)
p<-PlotHeatmap(as.data.frame(resOrdered), rld, meta, 100, 0,0,12)

```

#Plot log2 CPM
```{r}
#Convert to CPM
CPM<- counts_collapsed/colSums(counts_collapsed) *1000000

#Get mean CPM for blocked and unblocked for every species
blockedSamples<-meta[meta$Group=="Blocked",]
blocked_CPM<-CPM[,colnames(CPM) %in% as.character(blockedSamples$Samples)]

unblockedSamples<-meta[meta$Group=="Unblocked",]
unblocked_CPM<-CPM[,colnames(CPM) %in% as.character(unblockedSamples$Samples)]

CPM_Mean<-data.frame(Blocker=rowMeans(blocked_CPM), NoBlocker=rowMeans(unblocked_CPM))


#log2 of CPM
CPMLog<- log10(CPM_Mean+1) #+1 optional here


#Get miRNA info. Need: whether or not it was targeted + whether or not it was DE
#Add in target info- add column for whether or not the species was targeted
ActualTargets<- c('hsa-miR-486','hsa-miR-92a','hsa-miR-451a')
CPMLog$target<- rownames(CPMLog) %in% ActualTargets

#Add in DE info- using DESeq2 results from targets included analysis. Import data here.
CPMLog$Sig<- rownames(CPMLog) %in% rownames(sig)

#Add in third column, used to color points in graph. Add together target/sig columns. If it's a target and sig, it'll be 2, if it's just sig it'll be 1, if it's neither it'll be 0.
CPMLog$Color<- as.factor(rowSums(CPMLog[,3:4]))
CPMLog$Color<- gsub(1, "DE", CPMLog$Color)
CPMLog$Color<- gsub(0, "Not DE", CPMLog$Color)
CPMLog$Color<-gsub(2, "Target", CPMLog$Color)
CPMLog$Color<-factor(CPMLog$Color, levels=c("DE", "Not DE", "Target"))
CPMLog<-CPMLog[order(CPMLog$target, decreasing=TRUE),]



p<-CPMPlot(CPMLog, "Mean")


```

```

