---
title: "R Notebook"
output: html_notebook
---

#Large cohort: impact of blockers on target detection, improvements with blockers

Only data from the large cohort is used here. Count matrices for this cohort were all generated using the Basespace small RNA app. All analyses using the large cohort are included here for clarity (Figure 1C, 2E, 4B, and supplemental figure 1B)

Import libraries
```{r}
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(reshape2))
suppressPackageStartupMessages(library(purrr))
suppressPackageStartupMessages(library(tibble))
suppressPackageStartupMessages(library(metaseqR))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(DESeq2))
suppressPackageStartupMessages(library(pheatmap))
suppressPackageStartupMessages(library(ggrepel))
suppressPackageStartupMessages(library(ggpubr))
suppressPackageStartupMessages(library(gsubfn))
windowsFonts("Arial" = windowsFont("Arial"))
```

Functions
```{r}
#For plotting log2 CPM values nicely
CPMPlot<- function(CPM, Sample){
  #Get subset of CPM for labeling points- points that are both sig DE and are higher in unblocked than   blocked (beneath the slope line)
  CPM$HigherInUnblocked<- (CPM$Blocker-CPM$NoBlocker)<0
  SubsetToLabel<- CPM[CPM$HigherInUnblocked=="TRUE" &CPM$Color!="Not DE",]
  
  #Plot
  p<- ggplot(CPM, aes(x=NoBlocker, y=Blocker, color=Color))+
  geom_point(size=3)+
  geom_abline(slope=1, color="red")+
  scale_colour_manual(values=c("red","black", "blue"))+
  theme_classic()+
  xlim(0,6)+
  ylim(0,6)+
  xlab(paste("Log of Unblocked Counts Per Million"))+
  ylab(paste("Log of Blocked Counts Per Million"))+
  geom_text_repel(data=SubsetToLabel, aes(x=NoBlocker, y=Blocker, label=rownames(SubsetToLabel)), color="black",  size=5, fontface="bold")+
  ggtitle(Sample)+
  theme(axis.text.y   = element_text(size=14, colour = "black", face="bold"),
        axis.title.y  = element_text(size=14, colour = "black", face="bold"),
        axis.title.x  = element_text(size=14, colour = "black", face="bold"),
        axis.text.x  = element_text(size=14, colour = "black", face="bold"),
        legend.background = element_rect(fill=NA, size=1, 
                                         linetype="solid", color="black"),
        legend.title=element_blank(),
        legend.key.size = unit(.4, "cm"),
        legend.spacing.y = unit(0, "mm"),
        legend.position = c(.15,.892),
        legend.text = element_text(size=18, face="bold"),
        panel.border = element_rect(colour = "black", fill=NA, size=2),
        plot.title=element_text(size=14, face="bold"))
  return(p)
  
}

#Function that takes a column of counts and returns the number of species that pass the given list of threhsolds
CountsPassingThreshold<- function(counts, threshold){
  Passthreshold<-c()
  for (i in 1:length(threshold)){
    Passthreshold[i]<- sum(counts>threshold[i])
    
  }
  return(Passthreshold)
}

Head<-function(x){x[1:10,1:10]}


#Create metadata and DESeq2 object, then run DESeq2
RunDESeq2<- function(counts,meta) {
  #create DESeq2 object- just based on samplegroup (blockerstatus)
  #will add samples to model later (i.e., pairwise analysis)
  dds<- DESeqDataSetFromMatrix(countData=counts, 
                               colData=meta, 
                               design=~sampleGroup)
  
  #Create model that takes into account sample differences
  ddsMF<- dds
  design(ddsMF)<- formula(~phenoId + sampleGroup)
  
  #run DESeq2
  dds<- DESeq(ddsMF)
  
  return(dds)
}
```

#Overall steps:


3. Read in filtered/merged unblocked data from step 1 --> Further filtering + plotting % map to targets
4. Read in filtered/merged blocked data from step 2 --> Further filtering + plotting % map to targets
5. For blocked  and unblocked  data, plot the % mapping to any target (Supplemental Figure 1B)
6. For blocked and unblocked datasets, plot percentage mapping to each of the 3 targets + all other non-heme miRs (not used as figure)
7. For blocked and unblocked datasets, plot percentage mapping to each of the 11 targets (mature and isomir from target groups) (Figure 1C)
8. Differential expression- unpaired, DESeq2. Used for coloring sig points in Log2 CPM plot 
9. Plot log2 CPM values for blocked and unblocked datasets, coloring based on sig (from previous step) (Figure 2E)
10. Plot the number of species passing count thresholds for blocked and unblocked datasets (Figure 4B)
11. Plotting off target effects (only used in Supplemental table 1)





#6: Read in unblocked tech collapsed (paired or unpaired) raw counts from above chunk --> Further filtering + plotting % map to targets
```{r}
#Read in tech collapsed raw counts
unblocked_RemoveLow<-read.csv(paste(wd, "UnblockedRawCounts_TechCollapsed.csv",sep=""),row.names="X") #unpaired full dataset
unblocked_RemoveLow<-read.csv(paste(wd, "LargeCohort_PairedSubset_89Samples_Unblocked.csv", sep=""),row.names="X") #paired subset

#Convert to CPM
unblocked_CPM<-unblocked_RemoveLow/colSums(unblocked_RemoveLow) *1000000

#Get total number of CPM for 3 targets-will be multiple lines, grep for targets.Sum together target CPMs
AllTargets<-c("hsa-miR-486", "hsa-miR-92a", "hsa-miR-451a")

TargetCPM_1<-lapply(AllTargets, function(x) {unblocked_CPM[grep(x, rownames(unblocked_CPM), ignore.case = TRUE),]}) #get targets
TargetCPM_2<-lapply(TargetCPM_1, function(x) {x[!(rownames(x) %in% c("hsa-miR-4511", "hsa-miR-4510","hsa-mir-4511", "hsa-mir-4510")),]}) #remove miR-4511/4510
TargetCPM_3<- lapply( TargetCPM_2, function(x) {colSums(x[,1:ncol(x)])}) #merge multiple species in a family into one
TargetCPM<- as.data.frame(do.call(rbind, TargetCPM_3))
rownames(TargetCPM)<-AllTargets
TargetCPM<- TargetCPM[,order(colnames(TargetCPM))]


#Get percentage of total CPMs that 3 targets make up
unblocked_CPM<-unblocked_CPM[,order(colnames(unblocked_CPM))]
unblocked_TargetPercentageCPM<-TargetCPM/colSums(unblocked_CPM) *100

#add row with percent mapping to non-target
unblocked_TargetPercentageCPM["AllOthers" ,] <- 100-colSums(unblocked_TargetPercentageCPM)
```



#7: Read in tech collapsed (paired or unpaired) raw counts from above chunk --> Further filtering + plotting % map to targets

```{r}
#Read in tech collapsed raw counts
blocked_RemoveLow<-read.csv(paste(wd, "BlockedRawCounts_TechCollapsed.csv",sep=""),row.names="X") #unpaired full dataset
blocked_RemoveLow<-read.csv(paste(wd, "LargeCohort_PairedSubset_89Samples_Blocked.csv", sep=""),row.names="X") #paired subset

#Convert to CPM
blocked_CPM<-blocked_RemoveLow/colSums(blocked_RemoveLow) *1000000

#Get total number of CPM for 3 targets-will be multiple lines, grep for targets.Sum together target CPMs
AllTargets<-c("hsa-miR-486", "hsa-miR-92a", "hsa-miR-451a")
TargetCPM_1<-lapply(AllTargets, function(x) {blocked_CPM[grep(x, rownames(blocked_CPM), ignore.case=TRUE),]}) #get targets
TargetCPM_2<-lapply(TargetCPM_1, function(x) {x[!(rownames(x) %in% c("hsa-miR-4511", "hsa-miR-4510","hsa-mir-4511", "hsa-mir-4510")),]}) #remove miR-4511/4510
TargetCPM_3<- lapply( TargetCPM_2, function(x) {colSums(x[,1:ncol(x)])}) #merge multiple species in a family into one
TargetCPM<- as.data.frame(do.call(rbind, TargetCPM_3))
rownames(TargetCPM)<-AllTargets
TargetCPM<- TargetCPM[,order(colnames(TargetCPM))]


#Get percentage of total CPMs that 3 targets make up
blocked_CPM<-blocked_CPM[,order(colnames(blocked_CPM))]
blocked_TargetPercentageCPM<-TargetCPM/colSums(blocked_CPM) *100

#Optional: add row with percent mapping to non-target
blocked_TargetPercentageCPM["AllOthers" ,] <- 100-colSums(blocked_TargetPercentageCPM)
```


#8: For blocked  and unblocked  data, plot the % mapping to any target (Supplemental Figure 1B)
```{r}
#sum percent mapping for the 3 targets, take mean across all samples
blocked_all<- colSums(blocked_TargetPercentageCPM[1:3,])
unblocked_all<- colSums(unblocked_TargetPercentageCPM[1:3,])
df<-data.frame(Group=c("Blocked", "Unblocked"), 
               PercentMapping=c(mean(blocked_all), mean(unblocked_all)),
               sd=c(sd(blocked_all), sd(unblocked_all)))

#plot: 042820_LargeCohort_PercentMapping_AnyTarget_2Groups
p<-ggplot(df, aes(y=PercentMapping, x=reorder(Group, - PercentMapping),fill=Group))+
  geom_bar(stat="identity", position="dodge", width=.7)+ 
  theme_classic()+
  ylab("Percent of reads mapping to targets species")+
  xlab("")+
  scale_fill_manual("", values=c("Blocked"="firebrick3", "Unblocked"="turquoise3"))+
  geom_errorbar(aes(ymin=PercentMapping-sd, ymax=PercentMapping+sd), width=.1, size=1, position=position_dodge(.9),color="black")+
  ylim(0,100)+
  theme(axis.text.y   = element_text(size=12, color="black", face= "bold"),
          axis.text.x   = element_text(size=14, color="black", face="bold"),
          axis.title.y  = element_text(size=16, color="black", face="bold"),
          panel.border = element_rect(colour = "black", fill=NA, size=2),
          legend.title=element_blank(),
          legend.position = "none",
          legend.background = element_rect(fill=NA, size=1, linetype="solid", color="black"),
          axis.text.x.bottom  =element_text(size=20))
```

Significance test comparing blocked and unblocked datasets:
```{r}
t.test(unname(blocked_all), unname(unblocked_all))
```

#9: For blocked and unblocked datasets, plot percentage mapping to each of the 3 targets + all other non-heme miRs (not used as figure)
```{r}
#Add column for blocked/unblocked to "TargetPercentageCPM" dataframes
unblocked_TargetPercentageCPM$BlockerStatus<-rep("Unblocked", nrow(unblocked_TargetPercentageCPM))
blocked_TargetPercentageCPM$BlockerStatus<-rep("Blocked", nrow(blocked_TargetPercentageCPM))

#reformat for plotting
df_unblocked<-unblocked_TargetPercentageCPM
df_unblocked$Species<- rownames(df_unblocked)
df_unblocked_melt<- melt(df_unblocked, by="Species")

df_blocked<-blocked_TargetPercentageCPM
df_blocked$Species<- rownames(df_blocked)
df_blocked_melt<- melt(df_blocked, by="Species")

#Add column for ordering
df_unblocked_melt$Order<- rep(1, nrow(df_unblocked_melt))
df_blocked_melt$Order<- rep(2, nrow(df_blocked_melt))

#rbind unblocked and blocked dataframes together
bind_TargetPercentageCPM<- rbind(df_unblocked_melt, df_blocked_melt)
bind_TargetPercentageCPM$Order<- rep(seq(1,length(unique(bind_TargetPercentageCPM$Species)),1),
                                     nrow(bind_TargetPercentageCPM)/length(unique(bind_TargetPercentageCPM$Species)))
bind_TargetPercentageCPM$BlockerStatus<-factor(bind_TargetPercentageCPM$BlockerStatus, levels=c("Unblocked", "Blocked"))


#Plot
p<-ggplot(bind_TargetPercentageCPM, aes(x = reorder(Species, Order), y =value, fill=BlockerStatus)) +
        geom_boxplot(size=0.3,color="black",outlier.size = .7)+
        ylab("Percent of total reads mapping")+
        xlab("")+
        theme_classic()+
        scale_fill_manual("", values=c("Blocked"="firebrick3", "Unblocked"="turquoise3"))+
        theme(axis.title.y = element_text(face="bold", size=16),
              axis.text.y = element_text(face="bold", size=12, color="black"),
              axis.text.x = element_text(face="bold", size=16, color="black", angle=45,hjust=1),
              legend.text = element_text(face="bold", size=16),
              legend.title=element_blank(),
              panel.border = element_rect(colour = "black", fill=NA, size=2),
              strip.background = element_rect(color = "black", size = 2),
              legend.position=c(.72,.9),
              legend.background = element_rect(fill=NA, size=1, linetype="solid", color="black"),
              axis.ticks = element_line(size = 1.5,color="black"),
              axis.ticks.length=unit(.2, "cm"))

```

Get significance values for comparing blocked/unblocked in each species
```{r}
#Run t.test for all species
SplitDFBymiR<-split(bind_TargetPercentageCPM, bind_TargetPercentageCPM$Species)
Ttest<- unlist(lapply(SplitDFBymiR, function(x) {t.test(x[x$BlockerStatus=="Unblocked",4],x[x$BlockerStatus=="Blocked", 4])$p.value}))
```


```{r}
GetMeans<-bind_TargetPercentageCPM
GetMeans$Grouping<-paste(GetMeans$BlockerStatus,"_", GetMeans$Species)
GetMeans<- GetMeans %>% group_by(Grouping)  %>% mutate(mean=mean(value)) %>% as.data.frame()
GetMeans[!duplicated(GetMeans$mean),]
```

#10: For blocked and unblocked datasets, plot percentage mapping to each of the 11 species (mature and isomiRs in target group) (Figure 1C)

Get uncollapsed unblocked dataset:
```{r}
#Convert to CPM
unblocked_CPM<-unblocked_RemoveLow/colSums(unblocked_RemoveLow) *1000000

#Get total number of CPM for 3 targets-will be multiple lines, grep for targets.Sum together target CPMs
AllTargets<-c("hsa-miR-486", "hsa-miR-92a", "hsa-miR-451a")

TargetCPM_1<-lapply(AllTargets, function(x) {unblocked_CPM[grep(x, rownames(unblocked_CPM), ignore.case = TRUE),]}) #get targets
TargetCPM_2<-lapply(TargetCPM_1, function(x) {x[!(rownames(x) %in% c("hsa-miR-4511", "hsa-miR-4510","hsa-mir-4511", "hsa-mir-4510")),]}) #remove miR-4511/4510
TargetCPM<- as.data.frame(do.call(rbind, TargetCPM_2))
TargetCPM<- TargetCPM[,order(colnames(TargetCPM))]

#Get percentage of total CPMs that 3 targets make up
unblocked_CPM<-unblocked_CPM[,order(colnames(unblocked_CPM))]
unblocked_TargetPercentageCPM<-TargetCPM/colSums(unblocked_CPM) *100

#add row with percent mapping to non-target
unblocked_TargetPercentageCPM["AllOthers" ,] <- 100-colSums(unblocked_TargetPercentageCPM)
```

Get uncollapsed blocked dataset:
```{r}
#Convert to CPM
blocked_CPM<-blocked_RemoveLow/colSums(blocked_RemoveLow) *1000000

#Get total number of CPM for 3 targets-will be multiple lines, grep for targets.Sum together target CPMs
AllTargets<-c("hsa-miR-486", "hsa-miR-92a", "hsa-miR-451a")

TargetCPM_1<-lapply(AllTargets, function(x) {blocked_CPM[grep(x, rownames(blocked_CPM), ignore.case = TRUE),]}) #get targets
TargetCPM_2<-lapply(TargetCPM_1, function(x) {x[!(rownames(x) %in% c("hsa-miR-4511", "hsa-miR-4510","hsa-mir-4511", "hsa-mir-4510")),]}) #remove miR-4511/4510
TargetCPM<- as.data.frame(do.call(rbind, TargetCPM_2))
TargetCPM<- TargetCPM[,order(colnames(TargetCPM))]

#Get percentage of total CPMs that 3 targets make up
blocked_CPM<-blocked_CPM[,order(colnames(blocked_CPM))]
blocked_TargetPercentageCPM<-TargetCPM/colSums(blocked_CPM) *100

#add row with percent mapping to non-target
blocked_TargetPercentageCPM["AllOthers" ,] <- 100-colSums(blocked_TargetPercentageCPM)
```

#Merge blocked/unblocked uncollapsed datasets --> get % mapping to each miR --> plot (Figure 1C)
```{r}
#Add column for blocked/unblocked to "TargetPercentageCPM" dataframes
unblocked_TargetPercentageCPM$BlockerStatus<-rep("Unblocked", nrow(unblocked_TargetPercentageCPM))
blocked_TargetPercentageCPM$BlockerStatus<-rep("Blocked", nrow(blocked_TargetPercentageCPM))

#reformat for plotting
df_unblocked<-unblocked_TargetPercentageCPM
df_unblocked$Species<- rownames(df_unblocked)
df_unblocked_melt<- melt(df_unblocked, by="Species")

df_blocked<-blocked_TargetPercentageCPM
df_blocked$Species<- rownames(df_blocked)
df_blocked_melt<- melt(df_blocked, by="Species")

#Add column for ordering
df_unblocked_melt$Order<- rep(1, nrow(df_unblocked_melt))
df_blocked_melt$Order<- rep(2, nrow(df_blocked_melt))

#rbind unblocked and blocked dataframes together
bind_TargetPercentageCPM<- rbind(df_unblocked_melt, df_blocked_melt)
bind_TargetPercentageCPM$Order<- rep(seq(1,length(unique(bind_TargetPercentageCPM$Species)),1),
                                     nrow(bind_TargetPercentageCPM)/length(unique(bind_TargetPercentageCPM$Species)))
bind_TargetPercentageCPM$BlockerStatus<-factor(bind_TargetPercentageCPM$BlockerStatus, levels=c("Unblocked", "Blocked"))
```


```{r}
#Plot
p<-ggplot(bind_TargetPercentageCPM, aes(x = reorder(Species, Order), y =value, fill=BlockerStatus)) +
        geom_boxplot(size=0.3,color="black",outlier.size = .7)+
        ylab("Percent of total reads mapping")+
        xlab("")+
        theme_classic()+
        scale_fill_manual("", values=c("Blocked"="firebrick3", "Unblocked"="turquoise3"))+
        theme(axis.title.y = element_text(face="bold", size=16),
              axis.text.y = element_text(face="bold", size=12, color="black"),
              axis.text.x = element_text(face="bold", size=16, color="black", angle=45,hjust=1),
              legend.text = element_text(face="bold", size=16),
              legend.title=element_blank(),
              panel.border = element_rect(colour = "black", fill=NA, size=2),
              strip.background = element_rect(color = "black", size = 2),
              legend.position=c(.72,.9),
              legend.background = element_rect(fill=NA, size=1, linetype="solid", color="black"),
              axis.ticks = element_line(size = 1.5,color="black"),
              axis.ticks.length=unit(.2, "cm"))
```

Get significance values for comparing blocked/unblocked in each species
```{r}
#Run t.test for all species
SplitDFBymiR<-split(bind_TargetPercentageCPM, bind_TargetPercentageCPM$Species)
Ttest<- unlist(lapply(SplitDFBymiR, function(x) {t.test(x[x$BlockerStatus=="Unblocked",4],x[x$BlockerStatus=="Blocked", 4])$p.value}))
```


```{r}
GetMeans<-bind_TargetPercentageCPM
GetMeans$Grouping<-paste(GetMeans$BlockerStatus,"_", GetMeans$Species)
GetMeans<- GetMeans %>% group_by(Grouping)  %>% mutate(mean=mean(value)) %>% as.data.frame()
GetMeans[!duplicated(GetMeans$mean),]
```

#11: Differential expression- unpaired, DESeq2. Used for coloring sig points in Log2 CPM plot 

Filter, merge and export count matrices
```{r}
#Read in pre-processed count matrices, convert sample ID to "SampleID_Blocked/Unblocked)", convert to CPM
writewd<-"C:/Users/Jenna/Documents/BlockerProject/"
blocked_counts<-read.csv(paste(wd, "BlockedRawCounts_TechCollapsed.csv",sep=""),row.names="X")
colnames(blocked_counts)<-paste(colnames(blocked_counts), "Blocked",sep="_")
blocked_CPM<- blocked_counts/colSums(blocked_counts)*1000000

unblocked_counts<-read.csv( paste(writewd, "UnblockedRawCounts_TechCollapsed.csv", sep=""), row.names = "X")
colnames(unblocked_counts)<-paste(colnames(unblocked_counts), "Unblocked", sep="_")
unblocked_CPM<- unblocked_counts/colSums(unblocked_counts)*1000000

#Remove genes without > 1 CPM in 50% of unblocked/blocked samples- in CPM data
blocked_CPMPass<-blocked_CPM[rowSums(blocked_CPM>1)>(ncol(blocked_CPM)/2),]
unblocked_CPMPass<-unblocked_CPM[rowSums(unblocked_CPM>1)>(ncol(unblocked_CPM)/2),]

#For the genes that passsed the CPM threshold, get raw (downsampled) counts
blocked_CountsPass<-blocked_counts[rownames(blocked_counts) %in% rownames(blocked_CPMPass),]
unblocked_CountsPass<-unblocked_counts[rownames(unblocked_counts) %in% rownames(unblocked_CPMPass),]

#merge blocked and unblocked- only keep if found in both (i.e., >50% have CPM>1 in both blocked and unblocked), export
counts<- merge(blocked_CountsPass, unblocked_CountsPass, all=FALSE,by=0 )
write.csv(counts, paste(wd, "LargeCohort_TechCollapsed_BlockedUnblockedMerged.csv",sep=""))

#Create metadata, export
meta_blocked<-data.frame(Samples=colnames(blocked_CPMPass), Group="Blocked")
meta_unblocked<-data.frame(Samples=colnames(unblocked_CPMPass), Group="Unblocked")
meta<-rbind(meta_blocked,meta_unblocked)
write.csv(meta, paste(wd, "LargeCohort_TechCollapsed_Metadata.csv", sep=""))
```

```{r}

targets<-c("hsa-miR-486" , "hsa-miR-92a" , "hsa-miR-92b" , "hsa-miR-451a", "hsa-miR-25")
counts_TR<-counts_CPMPass[!(rownames(counts_CPMPass) %in% targets),]
dds<-DESeqDataSetFromMatrix(counts_TR, pairedSamples, design=~blocker)
DESeq2::plotPCA(varianceStabilizingTransformation(dds), intgroup="blocker")
```

Perform differential expression between two groups
```{r}
#Import raw counts
readwd<-"C:/Users/Jenna/Documents/BlockerProject/"
countsInput<-read.csv(paste(readwd, "LargeCohort_TechCollapsed_BlockedUnblockedMerged.csv", sep="" ), row.names = "Row.names")
counts<-countsInput
counts<-counts[,-1]

#Collapse counts-remove everything after 3rd "-" (-1, -5p, etc), then merge and sum together based on new miR IDs
counts$geneIDs<- sub("^([^-]*-[^-]*-[^-]*).*", "\\1", rownames(counts))
counts$geneIDs<- gsub("mir", "miR", counts$geneIDs)
counts_collapsed<-counts %>% group_by(geneIDs) %>% summarise_each(funs(sum)) %>% as.data.frame
rownames(counts_collapsed)<-counts_collapsed$geneIDs
counts_collapsed<- counts_collapsed[,2:ncol(counts_collapsed)]

#Import metadata
meta<- read.csv(paste(readwd,"LargeCohort_TechCollapsed_Metadata.csv",sep=""))
meta$Group<- factor(meta$Group, levels=c("Unblocked", "Blocked"))

#Run DESeq2
dds<-DESeqDataSetFromMatrix(counts_collapsed, meta, design=~Group)
dds<-DESeq(dds)

#Extract and order results
res<-na.omit(results(dds))
resOrdered<-res[order(res$padj),]
nrow(resOrdered[resOrdered$padj<0.01,]) #177
nrow(resOrdered[resOrdered$padj<0.05,]) # 181
sig<-resOrdered[resOrdered$padj<0.01,]
sig<- sig[abs(sig$log2FoldChange)>1,] #89
```

#12: Plot log2 CPM values for blocked and unblocked datasets, coloring based on sig (from previous step) (Figure 2E)
```{r}
#Convert to CPM
CPM<- counts_collapsed/colSums(counts_collapsed) *1000000

#Get mean CPM for blocked and unblocked for every species
blockedSamples<-meta[meta$Group=="Blocked",]
blocked_CPM<-CPM[,colnames(CPM) %in% as.character(blockedSamples$Samples)]

unblockedSamples<-meta[meta$Group=="Unblocked",]
unblocked_CPM<-CPM[,colnames(CPM) %in% as.character(unblockedSamples$Samples)]

CPM_Mean<-data.frame(Blocker=rowMeans(blocked_CPM), NoBlocker=rowMeans(unblocked_CPM))

#log2 of CPM
CPMLog<- log2(CPM_Mean+1) #+1 optional here

#Get miRNA info. Need: whether or not it was targeted + whether or not it was DE
#Add in target info- add column for whether or not the species was targeted
ActualTargets<- c('hsa-miR-486','hsa-miR-92a','hsa-miR-451a')
CPMLog$target<- rownames(CPMLog) %in% ActualTargets

#Add in DE info- using DESeq2 results from targets included analysis. Import data here.
CPMLog$Sig<- rownames(CPMLog) %in% rownames(sig)

#Add in third column, used to color points in graph. Add together target/sig columns. If it's a target and sig, it'll be 2, if it's just sig it'll be 1, if it's neither it'll be 0.
CPMLog$Color<- as.factor(rowSums(CPMLog[,3:4]))
CPMLog$Color<- gsub(1, "DE", CPMLog$Color)
CPMLog$Color<- gsub(0, "Not DE", CPMLog$Color)
CPMLog$Color<-gsub(2, "Target", CPMLog$Color)
CPMLog$Color<-factor(CPMLog$Color, levels=c("DE", "Not DE", "Target"))
CPMLog<-CPMLog[order(CPMLog$target, decreasing=TRUE),]

#plot
p<-CPMPlot(CPMLog, "Mean")+
        theme(axis.ticks = element_line(colour = "black", size =2), 
              axis.ticks.length = unit(2,"mm"))+
        xlim(0,20)+
        ylim(0,20)+
        theme(axis.title.x = element_text(size=20),
              axis.title.y = element_text(size=20))
```

#13: Differential expression for paired cohort samples. Pair-wise DESeq2 used. Metadata includes sample information from BH.
```{r}
#read in paired data, blocked and unblocked
unblocked_paired<-read.csv(paste(wd, "LargeCohort_PairedSubset_89Samples_Unblocked.csv", sep=""),row.names="X") #paired subset
blocked_paired<-read.csv(paste(wd, "LargeCohort_PairedSubset_89Samples_Blocked.csv", sep=""),row.names="X") #paired subset


#merge counts together
counts_paired<- merge(unblocked_paired, blocked_paired, by=0)
rownames(counts_paired)<-counts_paired$Row.names
counts_paired<-counts_paired[,-1]
counts_paired<-counts_paired[,order(colnames(counts_paired))]

#Collapse counts together for species in same family- -3p/5p/-1/-2
counts_paired$geneIDs<- sub("^([^-]*-[^-]*-[^-]*).*", "\\1", rownames(counts_paired))
counts_paired$geneIDs<- gsub("mir", "miR", counts_paired$geneIDs)
counts_collapsed_paired<-counts_paired %>% group_by(geneIDs) %>% summarise_each(funs(sum)) %>% as.data.frame
rownames(counts_collapsed_paired)<-counts_collapsed_paired$geneIDs
counts_collapsed_paired<- counts_collapsed_paired[,2:ncol(counts_collapsed_paired)]

#Filter to remove any species without >1 CPM in >=50% of samples =="detected"
CPM<-(counts_collapsed_paired/colSums(counts_collapsed_paired))*1000000
CPM_Above1<- CPM[apply(CPM, 1, function(x) {sum(x>=1)})>(ncol(CPM)*.75),]
counts_CPMPass<-counts_collapsed_paired[rownames(counts_collapsed_paired) %in% rownames(CPM_Above1),]

#Read in paired samples- From BH. Add column for new sampleID- "sampleIDppm_Blocked/Unblocked". Needed since some sample IDs are replicated between blocked/unblocked. 
pairedSamples<-read.csv(paste(wd, "PairedSamples_LargeCohort_FromBH.csv", sep=""), row.names = "UniqueID")
pairedSamples<-pairedSamples[,-1]

```


Create DESeq2 object, run pair-wise DE on all species
```{r}
#create DESeq2 object- just based on samplegroup (blockerstatus)
#will add samples to model later (i.e., pairwise analysis)
dds<- DESeqDataSetFromMatrix(countData=counts_CPMPass, 
                       colData=pairedSamples, 
                       design=~blocker)

#Create model that takes into account sample differences
ddsMF<- dds
design(ddsMF)<- formula(~phenoId + blocker)

#run DESeq2
dds<- DESeq(ddsMF)

#extract and order results  
res_paired<-na.omit(results(dds))
resOrdered_paired<-res_paired[order(res_paired$padj),]

nrow(resOrdered_paired[resOrdered_paired$padj<0.05,]) #270 (385 if unpaired, 170 if paired+CPM 75% filter)
nrow(resOrdered_paired[resOrdered_paired$padj<0.01,]) #244 (295 if unpaired, 154 if paired+CPM 75% filter)
nrow(resOrdered_paired[resOrdered_paired$padj<0.01 & abs(resOrdered_paired$log2FoldChange)>1,]) #138 (206 if unpaired, 69 if paired+CPM 75% filter)

#write.csv(resOrdered_paired, paste(wd, "PairedCohort_DESeq2Results_Collapsed.csv",sep=""))
```


```{r}
#rld<-varianceStabilizingTransformation(dds)

MinSubtract=-3
MaxSubtract=2
nspecies=50
fontsize=12 

  #Select top sig DE miRNAs
  topgenes<- rownames(resOrdered_paired[1:nspecies,])
  mat<- assay(rld)[topgenes,]
  mat<- mat - rowMeans(mat)
  #colnames(mat)<-paste(metadata$Patient," (",metadata$sampleGroup, ")",sep="")
  
  #Heatmap settings
  mat_breaks<- seq(min(mat-MinSubtract), max(mat-MaxSubtract), length.out=75)

  #Create metadata- to be used to color code to show which group the sample belongs to
  #metadata<- data.frame(Group=pairedSamples$blocker, 
                       #row.names=rownames(pairedSam),
                        #Sample=metadata$Patient)
  metadata<-pairedSamples

  #plot heatmap
  p<-pheatmap(mat, breaks = mat_breaks, 
            color =colorRampPalette( c("red", "black", "green"), space="rgb")(100),
            show_rownames = TRUE, show_colnames=FALSE,
            annotation = metadata,
            
            fontsize_row = fontsize,fontsize=12,treeheight_row=0,
            border_color= NA)
```


#Rerun pairwise DESeq2, removing 3 targets from analysis + removing 92b and 25 
Using count matrix that has been filtered to remove lowly expressed species (<50% of samples have >1 CPM)
```{r}
#Subset counts matrix, removing 3 targets
counts_collapsed_paired_RAT<- counts_CPMPass[!(rownames(counts_CPMPass) %in% c("hsa-miR-486",
                                                                                           "hsa-miR-451a",
                                                                                           "hsa-miR-92a",
                                                                                           "hsa-miR-92b",
                                                                                           "hsa-miR-25")),]

 
#create DESeq2 object- just based on samplegroup (blockerstatus)
#will add samples to model later (i.e., pairwise analysis)
dds<- DESeqDataSetFromMatrix(countData=counts_collapsed_paired_RAT, 
                       colData=pairedSamples, 
                       design=~blocker)

#Create model that takes into account sample differences
ddsMF<- dds
design(ddsMF)<- formula(~phenoId + blocker)

#run DESeq2
dds<- DESeq(ddsMF)

#extract and order results  
res_paired_RAT<-na.omit(results(dds))
resOrdered_paired_RAT<-res_paired_RAT[order(res_paired_RAT$padj),]


#write.csv(resOrdered_paired_RAT, paste(wd, "PairedCohort_DESeq2Results_ActualTargetsRemoved.csv",sep=""))

nrow(resOrdered_paired_RAT[resOrdered_paired_RAT$padj<0.05,]) #275 ( 216 if CPM filtered) 
nrow(resOrdered_paired_RAT[resOrdered_paired_RAT$padj<0.01,]) #243 ( 206 if CPM filtered)
nrow(resOrdered_paired_RAT[resOrdered_paired_RAT$padj<0.01 & abs(resOrdered_paired_RAT$log2FoldChange)>1,]) #142 ( 106 if CPM filtered)
sig<-resOrdered_paired_RAT[resOrdered_paired_RAT$padj<0.01 & abs(resOrdered_paired_RAT$log2FoldChange)>1,]
```

```{r}
#rld<-varianceStabilizingTransformation(dds)

MinSubtract=0
MaxSubtract=2
nspecies=50
  #Select top sig DE miRNAs
  topgenes<- rownames(resOrdered_paired_RAT[1:nspecies,])
  #topgenes<-rownames(sig)
  mat<- assay(rld)[topgenes,]
  mat<- mat - rowMeans(mat)
  #colnames(mat)<-paste(metadata$Patient," (",metadata$sampleGroup, ")",sep="")
  
  #Heatmap settings
  mat_breaks<- seq(min(mat-MinSubtract), max(mat-MaxSubtract), length.out=75)

  #Create metadata- to be used to color code to show which group the sample belongs to
  #metadata<- data.frame(Group=pairedSamples$blocker, 
                       #row.names=rownames(pairedSam),
                        #Sample=metadata$Patient)
  metadata<-pairedSamples

  #plot heatmap
  p<-pheatmap(mat, breaks = mat_breaks, 
            color =colorRampPalette( c("red", "black", "green"), space="rgb")(100),
            show_rownames = TRUE, show_colnames=FALSE,
            annotation = metadata,
            
            fontsize_row = fontsize,fontsize=12,treeheight_row=0,
            border_color= NA)
```


#10: Plot the number of species passing count thresholds for blocked and unblocked datasets (Figure 4B)

Read in pre-processed count matrices- only processing is collapsing isomirs together + downsampling
```{r}
#read in paired data, blocked and unblocked
unblocked_counts<-read.csv(paste(wd, "LargeCohort_PairedSubset_89Samples_Unblocked.csv", sep=""),row.names="X") #paired subset
blocked_counts<-read.csv(paste(wd, "LargeCohort_PairedSubset_89Samples_Blocked.csv", sep=""),row.names="X") #paired subset

#Collapse counts-remove everything after 3rd "-" (-1, -5p, etc), then merge and sum together based on new miR IDs
blocked_counts$geneIDs<- sub("^([^-]*-[^-]*-[^-]*).*", "\\1", rownames(blocked_counts))
blocked_counts$geneIDs<- gsub("mir", "miR", blocked_counts$geneIDs)
blocked_counts_collapsed<-blocked_counts %>% group_by(geneIDs) %>% summarise_each(funs(sum)) %>% as.data.frame
rownames(blocked_counts_collapsed)<-blocked_counts_collapsed$geneIDs
blocked_counts_collapsed<- blocked_counts_collapsed[,2:ncol(blocked_counts_collapsed)]

unblocked_counts$geneIDs<- sub("^([^-]*-[^-]*-[^-]*).*", "\\1", rownames(unblocked_counts))
unblocked_counts$geneIDs<- gsub("mir", "miR", unblocked_counts$geneIDs)
unblocked_counts_collapsed<-unblocked_counts %>% group_by(geneIDs) %>% summarise_each(funs(sum)) %>% as.data.frame
rownames(unblocked_counts_collapsed)<-unblocked_counts_collapsed$geneIDs
unblocked_counts_collapsed<- unblocked_counts_collapsed[,2:ncol(unblocked_counts_collapsed)]

blockedCounts<-blocked_counts_collapsed
unblockedCounts<-unblocked_counts_collapsed
```

Get number of species passing thresholds, then plot
```{r}
#Set desired thresholds
thresholds<-seq(0,500,50)
NumberInEachGroup<- length(thresholds)*(89+1) #used for plotting later

#Get counts passing threshold for unblocked samples, then transform data- 4 columns: ID, threshold, #of species that passed that threshold, and blocker status
        unblockedPassThreshold<- as.data.frame(apply(unblockedCounts, 2, 
                                                     CountsPassingThreshold, thresholds))
        rownames(unblockedPassThreshold)<- thresholds
        
        #add 4th column: mean of all samples
        unblockedPassThreshold$Mean<- rowMeans(unblockedPassThreshold)
        unblockedPassThreshold<- as.data.frame(t(unblockedPassThreshold))
        unblockedPassThreshold$ID<- rownames(unblockedPassThreshold)
        
        #reformat- melt
        unblockedPassThreshold<- melt(unblockedPassThreshold)
        colnames(unblockedPassThreshold)<- c("ID", "threshold","speciesPassingThreshold")
        unblockedPassThreshold$status<- rep("unblocked", length(thresholds) * (ncol(unblockedCounts)+1))
        unblockedPassThreshold$threshold<- as.integer(as.character((unblockedPassThreshold$threshold)))


#Get counts passing threshold for blocked samples, then transform data- 4 columns: ID, threshold, #of species that passed that threshold, and blocker status
        blockedPassThreshold<- as.data.frame(apply(blockedCounts, 2, 
                                                     CountsPassingThreshold, thresholds))
        rownames(blockedPassThreshold)<- thresholds
        
        #add 4th column: mean of all samples
        blockedPassThreshold$Mean<- rowMeans(blockedPassThreshold)
        blockedPassThreshold<- as.data.frame(t(blockedPassThreshold))
        blockedPassThreshold$ID<- rownames(blockedPassThreshold)
        
        #reformat- melt
        blockedPassThreshold<- melt(blockedPassThreshold)
        colnames(blockedPassThreshold)<- c("ID", "threshold","speciesPassingThreshold")
        blockedPassThreshold$status<- rep("blocked", length(thresholds) * (ncol(blockedCounts)+1))
        blockedPassThreshold$threshold<- as.integer(as.character((blockedPassThreshold$threshold)))

#combine into one
passThreshold<-rbind(unblockedPassThreshold,blockedPassThreshold)
passThreshold$status<-factor(passThreshold$status, levels=c("unblocked", "blocked"))
test<-passThreshold[passThreshold$status=="unblocked",]

#Plot
p<-ggplot(passThreshold, aes(x=threshold, y=speciesPassingThreshold, color=status))+
        geom_point(position="jitter",stat="identity",size=2,alpha=0.2)+
        scale_color_manual(values=c("turquoise3","firebrick3"))+
        theme_classic()+
        scale_x_continuous(breaks=seq(0,500,50))+
        scale_y_continuous(breaks=seq(0,600,100))+
        xlab("Counts threshold")+
        ylab("Number of miRs passing counts threshold")+
        theme(axis.text.x = element_text(size=12,face="bold",color="black"),
              axis.text.y = element_text(size=12,face="bold",color="black"),
              axis.title.x = element_text(size=14,face="bold",color="black"),
              axis.title.y = element_text(size=14,face="bold",color="black"),
              legend.title=element_blank(),
              legend.text = element_text(size=14, face="bold"),
              legend.position=c(.85,.85),
              legend.background = element_rect(fill=NA, size=1, 
                                         linetype="solid", color="black"),
              panel.border = element_rect(colour = "black", fill=NA, size=2))
        


means<- passThreshold[passThreshold$ID=="Mean",]
```


#Number passing thresholds, plotted like pilot- for paired cohort
```{r}
#Set desired thresholds
thresholds<-seq(0,500,100)
NumberInEachGroup<- length(thresholds)*(89+1) #used for plotting later

#Read in paired samples- From BH. 
pairedSamples<-read.csv(paste(wd, "PairedSamples_LargeCohort_FromBH.csv", sep=""), row.names = "UniqueID")
pairedSamples<-pairedSamples[,-1]

#Change column names of counts matrix from sample --> subject. 
unblockedSubject<- as.data.frame(t(unblockedCounts))
blockedSubject<- as.data.frame(t(blockedCounts))

# Merge with subject info --> make that the rowname --> retranspose
unblockedSubject<-merge(unblockedSubject, pairedSamples, by=0)[,!(colnames(unblockedSubject)=="blocker"),]
blockedSubject<-merge(blockedSubject, pairedSamples, by=0)[,!(colnames(blockedSubject)=="blocker"),]

rownames(unblockedSubject)<-unblockedSubject$phenoId
rownames(blockedSubject)<-blockedSubject$phenoId

unblockedSubject<-unblockedSubject[,-1]
blockedSubject<-blockedSubject[,-1]

unblockedSubject<-as.data.frame(t(unblockedSubject))
blockedSubject<-as.data.frame(t(blockedSubject))

#Convert counts from factor -->integer
unblockedSubject_int<-as.data.frame(apply(unblockedSubject, 2, as.integer))
blockedSubject_int<-as.data.frame(apply(blockedSubject, 2, as.integer))
rownames(unblockedSubject_int)<- rownames(unblockedSubject)
rownames(blockedSubject_int)<- rownames(blockedSubject)
unblockedSubject<-na.omit(unblockedSubject_int)
blockedSubject<-na.omit(blockedSubject_int)

#Get counts passing threshold for unblocked samples, then transform data- 4 columns: ID (A/b/c/d), threshold, #of species that passed that threshold, and blocker status
        unblockedPassThreshold<- as.data.frame(apply(unblockedSubject, 2, 
                                                     CountsPassingThreshold, thresholds))
        rownames(unblockedPassThreshold)<- thresholds
        
        #add 4th column: mean of all samples
        unblockedPassThreshold$Mean<- rowMeans(unblockedPassThreshold)
        unblockedPassThreshold<- as.data.frame(t(unblockedPassThreshold))
        unblockedPassThreshold$ID<- rownames(unblockedPassThreshold)
        
        #reformat- melt
        unblockedPassThreshold<- melt(unblockedPassThreshold)
        colnames(unblockedPassThreshold)<- c("ID", "threshold","speciesPassingThreshold")
        unblockedPassThreshold$status<- rep("unblocked", NumberInEachGroup)
        unblockedPassThreshold$threshold<- as.integer(as.character((unblockedPassThreshold$threshold)))

#Get counts passing threshold for blocked samples, then transform data- 4 columns: ID (A/b/c/d), threshold, #of species that passed that threshold, and blocker status
        blockedPassThreshold<- as.data.frame(apply(blockedSubject, 2, 
                                                     CountsPassingThreshold, thresholds))
        rownames(blockedPassThreshold)<- thresholds
        
        #add 4th column: mean of all samples
        blockedPassThreshold$Mean<- rowMeans(blockedPassThreshold)
        blockedPassThreshold<- as.data.frame(t(blockedPassThreshold))
        blockedPassThreshold$ID<- rownames(blockedPassThreshold)
        
        #reformat- melt
        blockedPassThreshold<- melt(blockedPassThreshold)
        colnames(blockedPassThreshold)<- c("ID", "threshold","speciesPassingThreshold")
        blockedPassThreshold$status<- rep("blocked", NumberInEachGroup)
        blockedPassThreshold$threshold<- as.integer(as.character((blockedPassThreshold$threshold)))

#combine into one
passThreshold<-rbind(unblockedPassThreshold,blockedPassThreshold)

#reformat so that you have 4 columns: ID, threshold, #species passing in noblocker, #species passing in blocker
PT<-data.frame(passThreshold$ID[1:NumberInEachGroup],
               passThreshold$threshold[1:NumberInEachGroup],
               passThreshold$speciesPassingThreshold[1:NumberInEachGroup],
               passThreshold$speciesPassingThreshold[(NumberInEachGroup+1):nrow(passThreshold)])
colnames(PT)<- c("ID", "threshold", "SpeciesPT_noblocker", "SpeciesPT_blocker")

#now need to subtract each- i.e., sample X at threshold 1 with no blocker vs with blocker. get those values for all, then plot. 
df<- as_tibble(PT)
df<-df%>% group_by(ID) %>% mutate(BlockerMinusNoBlocker=SpeciesPT_blocker-SpeciesPT_noblocker) %>% as.data.frame()
df$Mean<- rep(c(rep("Sample",89),"Mean"),length(thresholds))
df$Mean<- as.factor(df$Mean)

Samplecolors<- c("goldenrod1","violetred3", "skyblue3", "darkolivegreen4")

#plot
p<- ggplot(df, aes(x=threshold, y=BlockerMinusNoBlocker, color=ID))+
  geom_line(aes(linetype = Mean, group=ID, size=Mean))+
  scale_linetype_manual(values= c("solid", "dashed"),guide=FALSE)+
  scale_size_manual(values=c("Sample"=1.2,"Mean"=2), guide=FALSE)+
  #geom_point(size=1.3)+
  xlab("Count threshold")+
  ylab("n additional species detected in blocked libraries")+
  #scale_y_continuous(breaks=seq(0, 250, 50), limits=c(0,250))+
  scale_x_continuous(breaks=seq(0, max(df$threshold), 50))+
  #scale_color_manual(values=c("goldenrod1","violetred3", "skyblue3", "darkolivegreen4", "black"))+
  theme_classic()+
  guides(color = guide_legend(override.aes = list(size = 2)))+
  theme(axis.text.y   = element_text(size=12, colour = "black", face= "bold"),
        axis.title.y  = element_text(size=16, colour = "black", face= "bold"),
        axis.text.x   = element_text(size=12, colour = "black", face= "bold"),
        axis.title.x  = element_text(size=16, colour = "black", face= "bold"),
        panel.border = element_rect(colour = "black", fill=NA, size=2),
        legend.title=element_blank(),
        legend.position = c(.9,.85),
        legend.background = element_rect(fill=NA, size=1, linetype="solid", color="black"),
        legend.key.size = unit(.4, "cm"),
        legend.spacing.y = unit(0, "mm"),
        legend.text=element_text(size=14, face="bold"))
p
```

#For paired cohort: Get change in mapping to targets for each pair --> average

```{r}
blocked_targets<-blockedSubject[rownames(blockedSubject) %in% c("hsa-miR-451a", "hsa-miR-486", "hsa-miR-92a"),]
unblocked_targets<-unblockedSubject[rownames(unblockedSubject) %in% c("hsa-miR-451a", "hsa-miR-486", "hsa-miR-92a"),]

blocked_PerMapTargets<-blocked_targets/colSums(blockedSubject)*100
unblocked_PerMapTargets<-unblocked_targets/colSums(unblockedSubject)*100

blocked_PerMapTargets<- blocked_PerMapTargets[,order(colnames(blocked_PerMapTargets))]
unblocked_PerMapTargets<- unblocked_PerMapTargets[,order(colnames(unblocked_PerMapTargets))]

test<-mapply(blocked_PerMapTargets, unblocked_PerMapTargets, 2,function(x,y){x-y})

```


#For paired cohort data: Get off targets, plot. Export potential off-targets to csv.
 
```{r}
#Get merged raw counts- only downsampled + isomirs collapsed
counts<-merge(unblockedCounts, blockedCounts, by=0)
rownames(counts)<-counts$Row.names
counts<-counts[,-1]

#Get sig DE genes- p<0.01 and baseMean>50
sig<-resOrdered_paired[resOrdered_paired$padj<0.01,]
sig<-sig[sig$baseMean>=50,]
sig<- sig[abs(sig$log2FoldChange)>=0.5,] #101

#Add in CPM data for these miRNAs- across all 16 libraries
        #Convert to CPM
        counts_cpm<- (counts/colSums(counts)) *1000000
        
        #Get CPM for blocked and unblocked for sig miRs
        Unblocked_SigCPM<- as.data.frame(rowMeans(unblockedCounts[rownames(unblockedCounts) %in% rownames(sig),]))
        Blocked_SigCPM<- as.data.frame(rowMeans(blockedCounts[rownames(blockedCounts) %in% rownames(sig),]))
        
        #Merge with DE data, reformat
        OT<- merge(as.data.frame(sig[,c(1,2,6)]), Unblocked_SigCPM, by=0)
        rownames(OT)<-OT$Species
        OT<- merge(OT, Blocked_SigCPM, by=0)
        OT<-OT[,-1]
        colnames(OT)<- c("Species", "baseMean", "log2FoldChange", "padj", "MeanCPMUnblockedCounts", "MeanCPMBlockedCounts")

#Get mean of CPM- used for plotting- and reformat species name
OT$MeanCPM<- rowMeans(OT[,5:6])
OT$Species<-gsub("hsa-", "", OT$Species)

#Annotate by whether species is a target miR
ActualTargets<- c('miR-486-5p','miR-92a-3p','miR-451a') #CL
ActualTargets<- c('miR-486','miR-92a','miR-451a') #BS
OT$Target<- OT$Species %in% ActualTargets

#Entire list of all species in target family
Target_92<- rownames(counts[grep("miR-92", rownames(counts)),])
Target_451<- rownames(counts[grep("miR-451", rownames(counts)),]) 
Target_486<- rownames(counts[grep("miR-486", rownames(counts)),]) 
Target_25<- rownames(counts[grep("miR-25", rownames(counts)),]) #CL or BS
TargetFamilies<- gsub("hsa-", "", c(Target_451,Target_486,Target_92, Target_25))

#Add family info in
OT$TargetFamily<- OT$Species %in% TargetFamilies
OT$Color<- as.factor(rowSums(OT[,c(8,9)]))
OT$Color<- gsub(0, "Non-Target miR", OT$Color)
OT$Color<- gsub(1, "miR in Target Family", OT$Color)
OT$Color<- gsub(2, "Target miR", OT$Color)


#set which miRs to be labeled on graph
SubsetToLabel<- OT[OT$Color %in% c("Target miR", "miR in Target Family"),]
SubsetToLabel$Species<- gsub("hsa-", "", SubsetToLabel$Species)
```

```{r}
write.csv(OT, paste(wd,"OffTargetStats_PairedCohort_PairwiseDESeq2_TargetsIncluded_75CPMFilter.csv", sep=""))
```

Plot off target effects- same as in figure 3A, using pilot data.
```{r}
#plot
p<-ggplot(OT,aes(x=log2FoldChange, y= log2(padj+1e-100), fill=Color))+
  geom_point(aes(size=MeanCPM), pch=21, color="black")+
  scale_size_continuous(range = c(1, 50), guide=FALSE)+
  #scale_x_continuous(limits=c(-6,6),n.breaks = 7)+
  #scale_y_continuous(limits = c(-0.008, .05))+
  theme_classic()+
  xlab("log2 foldchange of differential expression")+
  ylab("Log2(Adjusted p value)")+
  geom_vline(xintercept=0,linetype="dotted", color= "black", size=1.25)+
  guides(colour = guide_legend(override.aes = list(size=5)))+
  geom_text_repel(data=SubsetToLabel, aes(x=log2FoldChange, y=padj, 
                                          label=Species),
                                          size=3.5, fontface= "bold",
                                          nudge_y = 0.0035, point.padding = 1.7, 
                                          box.padding = 0.5)+
  scale_fill_manual(values=c("orange","black", "red"))+
  theme(axis.text.y   = element_text(size=12, face= "bold", colour = "black"),
        axis.title.y  = element_text(size=16, face= "bold", colour = "black"),
        axis.title.x  = element_text(size=16, face="bold", colour = "black"),
        axis.text.x  = element_text(size=12, face="bold", colour = "black"),
        legend.background = element_rect(fill=NA, size=1, 
                                         linetype="solid", color="black"),
        legend.title=element_blank(),
        legend.key.size = unit(.4, "cm"),
        legend.spacing.y = unit(0, "mm"),
        legend.position = c(.8,.93),
        panel.border = element_rect(colour = "black", fill=NA, size=2),
        legend.text = element_text(size=13, face="bold"))

p


```

```{r}
sessionInfo()
```

